---
title: "CHROM wins"
output: 
  html_document:
    code_folding: hide
    toc: true
date: "2023-04-07"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Each chromosome begins at the bottom of each heatmap and extends upward.
A matrix-like data structure was used so that all chromosomes have the same number of windows, but the upper windows are empty (dark purple) where each chromosome ends.
The yellow cells have the highest gene density where genes are from "*.primary_high_confidence.gff3.gz" files and include 'gene' in column 3.
There is a pair of barplots for select windows, the first barplot includes all samples, the second barplot is subset so that the sample name for the highest value is visable.
The chromosomes are not oriented, this is most evident on chromosome 8 which has a high density of genes at one end of the chromosome but not the other.
The X chromosome plot has 4 column s that are empty because those haplotypes have Y chromosomes instead.
The Y chromosome plot has only 4 columns because there were only 4 males.


```{r}
#release_dir <- "/media/knausb/E737-9B48/releases/scaffolded/"
release_dir <- "/media/knausb/E737-9B48/releases/oriented_beds/"
samps <- list.files(release_dir)
samps <- sub(".primary_high_confidence_oriented.bed.gz", "", samps)
samps <- samps[ grep("EDTAOutput", samps, invert = TRUE) ]
samps <- samps[ grep("orient_beds", samps, invert = TRUE) ]
samps <- samps[ grep("RStudio", samps, invert = TRUE) ]

# my_chroms <- as.data.frame(matrix( nrow = length(samps), ncol = 11 ))
# rownames(my_chroms) <- samps
# colnames(my_chroms) <- paste("chr", c(1:9, "X", "Y"), sep = "")
# my_chroms[1:3, 1:11]
```

Initialize window data structure.


```{r, eval = FALSE}
wins <- data.frame( start = seq(0, 120e6, by = 1e6) + 1, end = seq(0, 120e6, by = 1e6) + 1e6 )
wins[1:3, ]

chrom_l <- vector(mode = "list", length = 11)
names(chrom_l) <- paste("chr", c(1:9, "X", "Y"), sep = "")
for(i in 1:length(chrom_l)){
  chrom_l[[i]] <- matrix( nrow = nrow(wins), ncol = length(samps) )
  colnames(chrom_l[[i]]) <- samps
}
```


Populate GFF data structure

```{r, eval = FALSE}
# for( i in 1:length(samps) ){
#   tmp <- read.table( file = paste(release_dir, "/", samps[i], "/", samps[i], ".primary_high_confidence.gff3.gz", sep = ""),
#                      header = FALSE, sep = "\t", quote = "\"" )
#   tmp <- tmp[ tmp[, 3] == "gene", ]
#   tmp <- tmp[ grep("chr", tmp[,1]), ]
#   tmp[1:3, 1:8]
#   for(j in 1:11){
#     tmp2 <- tmp[ grep(names(chrom_l)[j], tmp[, 1]) , , drop = FALSE ]
#     for( k in 1:nrow(wins) ){
#       # Starts in column 4.
#       tmp3 <- tmp2[ tmp2[, 4] >= wins$start[k] & tmp2[, 4] <= wins$end[k], , drop = FALSE]
#       chrom_l[[j]][ k, samps[i]] <- nrow(tmp3)
#     }
#   }
# }
```


Populate BED data structure

```{r, eval = FALSE}
#i <- 1
# for( i in 1:length(samps) ){
#   tmp <- read.table( file = paste(release_dir, "/", samps[i], ".primary_high_confidence_oriented.bed.gz", sep = ""),
#                      header = FALSE, sep = "\t", quote = "\"" )
#   tmp <- tmp[ grep("chr", tmp[,1]), ]
# #  tmp[1:3, 1:6]
#   for(j in 1:11){
#     tmp2 <- tmp[ grep(names(chrom_l)[j], tmp[, 1]) , , drop = FALSE ]
#     for( k in 1:nrow(wins) ){
#       # GFF starts in column 4.
#       # BED starts in column 2.
#       tmp3 <- tmp2[ tmp2[, 2] >= wins$start[k] & tmp2[, 2] <= wins$end[k], , drop = FALSE]
#       chrom_l[[j]][ k, samps[i]] <- nrow(tmp3)
#     }
#   }
# }

```




```{r}
#save(chrom_l, file = "chrom_l.RData")
#
load(file = "chrom_l.RData")
#save(chrom_l, file = "chrom_l_obed.RData")
#load(file = "chrom_l_obed.RData")
```


```{r, fig.align='center', fig.width=10, fig.height=11, eval=FALSE}
library(vcfR)

for(i in 1:11){
  heatmap.bp(chrom_l[[i]])
  title( main = paste("Chromosome:", names(chrom_l)[i]) )
}
```


## Chromosome 1

```{r, fig.height=10, fig.width=10}
library(vcfR)
i <- 1
heatmap.bp(chrom_l[[i]])
title( main = paste("Chromosome:", names(chrom_l)[i]) )
```


```{r, fig.height=4, fig.width=10}
#heatmap.bp(chrom_l[[i]][56:58, ] )
#max(chrom_l[[i]])
win_num <- 57
barplot(chrom_l[[i]][win_num, ], las = 3)
title( ylab = "Genes")
title( main = paste("Chromosome:", names(chrom_l)[i], ", window", win_num - 1, "to", win_num , "Mbp" ) )
barplot(chrom_l[[i]][57, 30:50], las = 3)
title( ylab = "Genes")
title( main = paste("Chromosome:", names(chrom_l)[i], ", window", win_num - 1, "to", win_num , "Mbp" ) )
```


## Chromosome 2

```{r, fig.height=10, fig.width=10}
library(vcfR)
i <- 2
heatmap.bp(chrom_l[[i]])
title( main = paste("Chromosome:", names(chrom_l)[i]) )
```


```{r, fig.height=4, fig.width=10}
#heatmap.bp(chrom_l[[i]][46:50, ] )
#max(chrom_l[[i]])

win_num <- 48
barplot(chrom_l[[i]][win_num, ], las = 3)
title( ylab = "Genes")
title( main = paste("Chromosome:", names(chrom_l)[i], ", window", win_num - 1, "to", win_num , "Mbp" ) )
barplot(chrom_l[[i]][win_num, 30:50], las = 3)
title( ylab = "Genes")
title( main = paste("Chromosome:", names(chrom_l)[i], ", window", win_num - 1, "to", win_num , "Mbp" ) )
```


## Chromosome 3

```{r, fig.height=10, fig.width=10}
library(vcfR)
i <- 3
heatmap.bp(chrom_l[[i]])
title( main = paste("Chromosome:", names(chrom_l)[i]) )
```

```{r, fig.height=4, fig.width=10}
#heatmap.bp(chrom_l[[i]][56:78, ] )
#max(chrom_l[[i]])
win_num <- 78
barplot(chrom_l[[i]][win_num, ], las = 3)
title( ylab = "Genes")
title( main = paste("Chromosome:", names(chrom_l)[i], ", window", win_num - 1, "to", win_num , "Mbp" ) )
barplot(chrom_l[[i]][win_num, 40:60], las = 3)
title( ylab = "Genes")
title( main = paste("Chromosome:", names(chrom_l)[i], ", window", win_num - 1, "to", win_num , "Mbp" ) )
```


## Chromosome 4

```{r, fig.height=10, fig.width=10}
library(vcfR)
i <- 4
heatmap.bp(chrom_l[[i]])
title( main = paste("Chromosome:", names(chrom_l)[i]) )
#max(chrom_l[[i]])
```


## Chromosome 5

```{r, fig.height=10, fig.width=10}
library(vcfR)
i <- 5
heatmap.bp(chrom_l[[i]])
title( main = paste("Chromosome:", names(chrom_l)[i]) )
#max(chrom_l[[i]])
```


## Chromosome 6

```{r, fig.height=10, fig.width=10}
library(vcfR)
i <- 6
heatmap.bp(chrom_l[[i]])
title( main = paste("Chromosome:", names(chrom_l)[i]) )
```

```{r, fig.height=4, fig.width=10}
#heatmap.bp(chrom_l[[i]][56:78, ] )
#max(chrom_l[[i]])
win_num <- 1
barplot(chrom_l[[i]][win_num, ], las = 3)
title( ylab = "Genes")
title( main = paste("Chromosome:", names(chrom_l)[i], ", window", win_num - 1, "to", win_num , "Mbp" ) )
barplot(chrom_l[[i]][win_num, 10:40], las = 3)
title( ylab = "Genes")
title( main = paste("Chromosome:", names(chrom_l)[i], ", window", win_num - 1, "to", win_num , "Mbp" ) )
```


## Chromosome 7

```{r, fig.height=10, fig.width=10}
library(vcfR)
i <- 7
heatmap.bp(chrom_l[[i]])
title( main = paste("Chromosome:", names(chrom_l)[i]) )
```

```{r, fig.height=4, fig.width=10}
#heatmap.bp(chrom_l[[i]][56:78, ] )
#max(chrom_l[[i]])
win_num <- 1
barplot(chrom_l[[i]][win_num, ], las = 3)
title( ylab = "Genes")
title( main = paste("Chromosome:", names(chrom_l)[i], ", window", win_num - 1, "to", win_num , "Mbp" ) )
barplot(chrom_l[[i]][win_num, 40:70], las = 3)
title( ylab = "Genes")
title( main = paste("Chromosome:", names(chrom_l)[i], ", window", win_num - 1, "to", win_num , "Mbp" ) )
```


## Chromosome 8

```{r, fig.height=10, fig.width=10}
library(vcfR)
i <- 8
heatmap.bp(chrom_l[[i]])
title( main = paste("Chromosome:", names(chrom_l)[i]) )
```


## Chromosome 9

```{r, fig.height=10, fig.width=10}
library(vcfR)
i <- 9
heatmap.bp(chrom_l[[i]])
title( main = paste("Chromosome:", names(chrom_l)[i]) )
```


## Chromosome 10 (X)

```{r, fig.height=10, fig.width=10}
library(vcfR)
i <- 10
heatmap.bp(chrom_l[[i]])
title( main = paste("Chromosome:", names(chrom_l)[i]) )
```


## Chromosome 11 (Y)

```{r, fig.height=10, fig.width=10}
library(vcfR)
i <- 11
heatmap.bp(chrom_l[[i]])
title( main = paste("Chromosome:", names(chrom_l)[i]) )
```

