---
title: "SynChrom"
output: 
  html_document:
    toc: true
    code_folding: hide
date: "2023-03-14"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.align = "center")
```


## Methods


**Chrom plot methods**

The genetically female EH23 genome consists of 9 pairs of autosomes and a pair of sex chromosomes (XX).
Ideograms for each pair of chromosomes were created using ggplot2 [https://ggplot2.tidyverse.org] in R [https://www.R-project.org/].
One million base pair windows were created for each chromosome where the number of genes and the 'CpG' motif were counted for each window [unpublished script???].
The CpG count was converted into a rate by dividing by the window size, this also accommodated the last window of each chromosome which was less than one million base pairs in size.
Each window was scaled by subtracting the minimum and dividing by the maximum value for each window.
In order to visually emphasize the lack of the 'CpG' motif an inverse was taken by taking one and subtracting the scaled 'CpG' rate for each window.
This value was then multiplied by the scalar 1000 so that it would range from 0 - 1000.
The count of genes per window was converted to a rate by dividing by each window size.
This scaled, inverse 'CpG' rate was used for the width of windows while the color of each window was determined by the gene rate and colored using the viridis magma palette [10.5281/zenodo.4679424].

Structural variation among each pair of chromosomes was determined by using minimap2 [https://doi.org/10.1093/bioinformatics/bty191].
The minimap2 comparisons were annotated using SyRI [doi:10.1186/s13059-019-1911-0].
The syntenous and inverted regions were plotted using ggplot2::geom_polygon in a manner inspired by plotsr [https://doi.org/10.1093/bioinformatics/btac196].

The location of candidate loci was determined by using BLASTN [https://doi.org/10.1016/S0022-2836(05)80360-2].
BLASTN was called with the following options: -task megablast -evalue 0.001 -perc_identity 90 -qcov_hsp_perc 90.

Query sequences were as follows: cannabichromenic acid synthase (LY658671.1), cannabidiolic acid synthase (AB292682, AB292683, AB292684), tetrahydrocannabinolic acid synthase (AB212829, AB212830), and olivetolic acid cyclase (NC_044376.1:c4279947-4279296, NC_044376.1:c4272107-4271242).




grep "^>" todd_rrna/todd_rrna.fa

**Todd please review**

```
>45s_not_separated_AH3Ma.chr4
>26s_AH3Ma.chr4
>5.8s_AH3Ma.chr4
>18s_AH3Ma.chr4

>5S_AH3Ma.chr7:43382134-43382392

>AH3Ma_CS-237_satelite
>CsatSD_centromere_237bp
>CsatSD_centromere_370bp

```



FASTA2nuccomp

motif_counter [no citation???]

Gene density

minimmap2 [https://doi.org/10.1093/bioinformatics/bty191]
Syri [doi:10.1186/s13059-019-1911-0] inspired by plotsr [https://doi.org/10.1093/bioinformatics/btac196].

BLASTN
[https://doi.org/10.1016/S0022-2836(05)80360-2]

Onofri [https://doi.org/10.1016/j.phytochem.2015.03.006]


viridis [10.5281/zenodo.4679424]


**Allele frequency methods**

Genotype data in the VCF format [https://doi.org/10.1093/bioinformatics/btr330] was input into R [https://www.R-project.org/] using vcfR [https://doi.org/10.1111/1755-0998.12549].
Wright's *F~IS~* [https://doi.org/10.1038/166247a0] was calculated as (*H~S~* - *H~O~*)/*H~S~* where *H~O~* is the observed number of heterozygotes divided by their number and *H~S~* is the number of heterozygotes we expect based on the allele frequencies, calculated as the frequency of the first allele multiplied by the frequency of the second and then by two. 
Wright's *F~IS~* therefore gives us the deviation in heterozygosity from our random expectation.
Scatterplots were generated using ggplot2 [https://ggplot2.tidyverse.org].
Graphical panels were assembled into a single graphic using ggpubr [https://CRAN.R-project.org/package=ggpubr].





## Results


**Organization of the EH23 genome**

The EH23 *Cannabis sativa* genome is organized into 9 autosomes and a pair of sex chromosomes.
Each pair of chromosomes demonstrated differences in chromosome length but also demonstrated a large degree of synteny among each pair.
Genes were located predominantly near the ends of each chromosomes, as opposed to being evenly distributed throughout each chromosomes.
This pattern of gene location near the ends of the chromosomes is consistent with previous reports [https://doi.org/10.1101/gr.242594.118].
The abundance of the motif 'CpG' (cytosine, phosphate, guanine; a motif thought to be important in the context of methylation) varied throughout each pair of chromosomes.
This CpG abundance appeared greatest near the centromeres, suggesting it can be used to determine approximate centromere location.
Each of the 10 pairs of *Cannabis sativa* chromosomes appeared syntenic with an abundance of CpG motifs marking each centromere.


**Nonrandom inheritance of SNP markers**

Alleles inherited through a single generation in a population of plants grown in a common garden demonstrated random and nonrandom frequencies of inheritance.
The parental population consisted of a self-pollinated individual, so allele frequencies at variable (heterozygous) positions were 0.5.
Our random, or evolutionarily neutral, expectation is that allele frequencies will be maintained from generation to generation.
That is, the allele frequencies in the parental population (a single individual) should approximate the allele frequencies observed in the progeny population (270 progeny plants).
Similarly, our random, or evolutionarily neutral, expectation is that these alleles will combine to form genotypes randomly to meet our Hardy-Weinberg expectation (heterozygous genotypes at a rate of 0.5).
Variants in the EH23 *Cannabis sativa* genome demonstrated both neutral as well as non-random patterns of inheritance, typically at the chromosomal level.


Chromosome 5 exhibited our neutral assumptions where the frequency of each allele in the progeny population was at around 0.5, matching the frequencies observed in the parental generation.
The alleles on chromosome 5 combined to form genotypes at a random, or neutral, rate as well.
This is indicated by approximately half of the genotypes occurring as heterozygotes and an *F~IS~* near zero.
Chromosome 2 also exhibits our neutral assumptions where the frequencies of each allele was at 0.5.
However, these alleles combined to form genotypes in a manner where homozygotes are favored, resulting in a positive *F~IS~*.
Chromosome 3 demonstrated a deviation in allele frequency from our neutral expectation of 0.5, where one allele was over-represented in the progeny.
Despite this substantial deviation in allele frequency, these alleles joined together to form genotypes in a neutral manner, resulting in an *F~IS~* near zero and a heterozygote frequency near 0.5.
Chromosome 4 had a non-neutral deviation in allele frequency that was similar to that of chromosome 3.
However, on chromosome 4 the alleles joined to form genotypes in a manner where heterozygotes were favored, as indicated by a negative *F~IS~*.


Near the ends of the chromosomes, where most of the genes occur and recombination is observed, we tended to observe our neutral expectations.
For example, much of chromosome 1 demonstrated a deviation from our neutral allele frequency of 0.5 yet the ends of the chromosomes returned to our neutral expectation of 0.5.
This suggests that the predominantly non-recombining chromosomes behave as a single locus except at the ends where recombination may allow a return to neutrality.


Throughout the hemp genome we observed different patterns of evolution.
What is striking about these results is that different evolutionary scenarios were exhibited on different chromosomes.
For example, the positive *F~IS~* on chromosome may cause us to interpret it as a result of inbreeding, while the negative *F~IS~* on chromosome 4 may cause us to interpret it as a result of outbreeding.
These are phenomena we typically discuss as a property of a population or individuals within a population, where in the hemp genome we observe these phenomena to differentiate among chromosomes.


## Discussion

prezygotic versus postzygotic




