---
title: "SynChrom"
output: 
  html_document:
    toc: true
    code_folding: hide
date: "2023-03-14"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.align = "center")
```


## Methods


**Chromosome plot methods**

The genetically female EH23 genome consisted of 9 pairs of autosomes and a pair of sex chromosomes (XX).
Ideograms for each pair of chromosomes were created using ggplot2 [https://ggplot2.tidyverse.org] in R [https://www.R-project.org/].
The length of each chromosome was determined using a custom script [unpublished script???] and used with 'ggplot::geom_rect()' to initialize the plot.
One million base pair windows were created for each chromosome where the number of genes and the 'CpG' motif were counted for each window [unpublished script???].
The CpG count was converted into a rate by dividing by the window size, this also accommodated the last window of each chromosome which was less than one million base pairs in size.
In order to visually emphasize the lack of the 'CpG' motif in the centromeric region, an inverse of the CpG rate was taken by taking one and subtracting the inverted 'CpG' rate for each window.
Each window was scaled by subtracting the minimum and dividing by the maximum value for each window.
This value was then multiplied by the scalar 1000 so that it would range from 0 - 1000, in accordance with the BED file format that this information was output in.
The count of genes per window was converted to a rate by dividing by each window size.
This scaled, inverse 'CpG' rate was used for the width of windows while the color of each window was determined by the gene rate and colored using the viridis magma palette [10.5281/zenodo.4679424].

Structural variation among each pair of chromosomes was determined using minimap2 [https://doi.org/10.1093/bioinformatics/bty191].
The minimap2 comparisons were annotated using SyRI [doi:10.1186/s13059-019-1911-0].
The syntenous and inverted regions were plotted using ggplot2::geom_polygon in a manner inspired by plotsr [https://doi.org/10.1093/bioinformatics/btac196].


The location of candidate loci within EH23 haplotypes A and B were determined using BLASTN [https://doi.org/10.1016/S0022-2836(05)80360-2].
BLASTN was called with the following options: -task megablast -evalue 0.001 -perc_identity 90 -qcov_hsp_perc 90.
Query sequences were as follows: cannabichromenic acid synthase (LY658671.1), cannabidiolic acid synthase (AB292682, AB292683, AB292684), tetrahydrocannabinolic acid synthase (AB212829, AB212830), and olivetolic acid cyclase (NC_044376.1:c4279947-4279296, NC_044376.1:c4272107-4271242).


**Todd please review**

```
bash$ grep "^>" todd_rrna/todd_rrna.fa
>45s_not_separated_AH3Ma.chr4
>26s_AH3Ma.chr4
>5.8s_AH3Ma.chr4
>18s_AH3Ma.chr4

>5S_AH3Ma.chr7:43382134-43382392

>AH3Ma_CS-237_satelite

# Onofri et al.? [https://doi.org/10.1016/j.phytochem.2015.03.006]
>CsatSD_centromere_237bp
>CsatSD_centromere_370bp
```


Tabular results (subject chromosome, subject start of alignment, subject end of alignment) from BLASTN were read into R [https://www.R-project.org/] and plotted on ideograms with ggplot2::geom_rect() [https://ggplot2.tidyverse.org].



**Allele frequency methods**


Genotype data in the VCF format [https://doi.org/10.1093/bioinformatics/btr330] was input into R [https://www.R-project.org/] using vcfR [https://doi.org/10.1111/1755-0998.12549].
Allele and heterozygote counts were made with vcfR [https://doi.org/10.1111/1755-0998.12549].
Wright's *F~IS~* [https://doi.org/10.1038/166247a0] was calculated as (*H~S~* - *H~O~*)/*H~S~* where *H~O~* is the observed number of heterozygotes divided by their number and *H~S~* is the number of heterozygotes we expect based on the allele frequencies, calculated as the frequency of the first allele multiplied by the frequency of the second and then by two. 
Wright's *F~IS~* therefore gives us the deviation in heterozygosity from our random expectation.
Scatterplots were generated using ggplot2 [https://ggplot2.tidyverse.org].
Graphical panels were assembled into a single graphic using ggpubr [https://CRAN.R-project.org/package=ggpubr].



## Results


**Organization of the EH23 genome**

The EH23 *Cannabis sativa* genome is organized into 9 pairs of autosomes and a pair of sex chromosomes (Figure 1A).
Each pair of chromosomes were different in chromosome length and also demonstrated a large degree of synteny among each pair.
Genes were located predominantly near the ends of each chromosomes, as opposed to being distributed evenly throughout each chromosome.
This pattern of gene location near the ends of the chromosomes is consistent with previous reports [https://doi.org/10.1101/gr.242594.118].
The abundance of the motif 'CpG' (cytosine, phosphate, guanine; a motif thought to be important in the context of methylation) varied throughout each pair of chromosomes.
This CpG abundance was associated with centromeric repeats, suggesting it can be used to determine approximate centromeric region.
Each of the 10 pairs of *Cannabis sativa* chromosomes appeared largely syntenic with an abundance of CpG motifs associated with centromeric regions.


**Nonrandom inheritance of SNP markers**

Alleles inherited through a single generation in a population of plants grown in a common garden demonstrated random and nonrandom frequencies of inheritance.
The parental population consisted of a self-pollinated individual, so allele frequencies at variable (heterozygous) positions were 0.5.
Our random, or evolutionarily neutral, expectation is that allele frequencies will be maintained from generation to generation.
That is, the allele frequencies in the parental population (a single individual) should approximate the allele frequencies observed in the progeny population (270 progeny plants).
Similarly, our random, or evolutionarily neutral, expectation is that these alleles will combine to form genotypes randomly to meet our Hardy-Weinberg expectation (heterozygous genotypes at a rate of 0.5).
Variants in the EH23 *Cannabis sativa* genome demonstrated both neutral as well as non-random patterns of inheritance, typically at the chromosomal level.


A random, or evolutionarily neutral, pattern of inheritance was observed on chromosome 4.


Chromosome 5 exhibited our neutral assumptions where the frequency of each allele in the progeny population was at around 0.5, matching the frequencies observed in the parental generation.
The alleles on chromosome 5 combined to form genotypes at a random, or neutral, rate as well.
This is indicated by approximately half of the genotypes occurring as heterozygotes and an *F~IS~* near zero.
Chromosome 2 also exhibits our neutral assumptions where the frequencies of each allele was at 0.5.
However, these alleles combined to form genotypes in a manner where homozygotes are favored, resulting in a positive *F~IS~*.
Chromosome 3 demonstrated a deviation in allele frequency from our neutral expectation of 0.5, where one allele was over-represented in the progeny.
Despite this substantial deviation in allele frequency, these alleles joined together to form genotypes in a neutral manner, resulting in an *F~IS~* near zero and a heterozygote frequency near 0.5.
Chromosome 4 had a non-neutral deviation in allele frequency that was similar to that of chromosome 3.
However, on chromosome 4 the alleles joined to form genotypes in a manner where heterozygotes were favored, as indicated by a negative *F~IS~*.


Near the ends of the chromosomes, where most of the genes occur and recombination is observed [https://doi.org/10.1101/gr.242594.118], we tended to observe our neutral expectations.
For example, much of chromosome 1 demonstrated a deviation from our neutral allele frequency of 0.5 yet the ends of the chromosomes returned to our neutral expectation of 0.5.
This suggests that the predominantly non-recombining chromosomes behave as a single locus except at the ends where recombination may allow a return to neutrality.


Throughout the hemp genome we observed different patterns of evolution.
What is striking about these results is that different evolutionary scenarios were exhibited on different chromosomes.
For example, the positive *F~IS~* on chromosome may cause us to interpret it as a result of inbreeding, while the negative *F~IS~* on chromosome 4 may cause us to interpret it as a result of outbreeding.
These are phenomena we typically discuss as a property of a population or individuals within a population, where in the hemp genome we observe these phenomena to differentiate among chromosomes.


## Discussion

### Chromosomal organization

The phased diploid assembly of *Cannabis sativa* L. strain EH23 provides new perspectives on the hemp genome.
The EH23 assembly's nine pairs of automosomes and the pair of sex chromosomes (XX) were numbered and oriented according to the haploid CBDRx assembly (GCF_900626175.2)[https://doi.org/10.1111/nph.17243].
The ends of the chromosomes are enriched for gene presence, as reported previously [https://doi.org/10.1101/gr.242594.118], except for chromosome 8 which did not have a high density of genes at one end.

Length polymorphism was evident between each pair of chromosomes.
While some of this difference in length may be due to assembly efficiency, 




### Inheritance of alleles

prezygotic versus postzygotic


Inheritance of genetic material from generation to generation is expected to occur in a random, or evolutionarily neutral, manner.
This should be observed in a large population as equal allele frequencies that randomly combine to form genotypes that satisfy our Hardy–Weinberg proportions.
This was observed on chromosome 4 where each allele frequency was equal (*p* ≈ *q* ≈ 0.5) and no deviation in our random expectation of heterozygosity was observed (*F~IS~* ≈ 0).
Elsewhere in the genome, deviations from our neutral expectation were observed.


Chromosome 2 demonstrated allele frequencies that substantially deviated from equal (*p* ≈ 0.6, *q* ≈ 0.4). 
However, these alleles combined to form genotypes at a proportion that met our random expectation.
This suggests selection on the alleles but not the genotypes.
Because there was a single, self pollinated, parent, we expect the allele frequencies and the genotype frequencies to be equal in the parent, indicating that the selection event occurred after the parent, and before the formation of genotypes.
In the male lineage this conversation typically involves sperm competition.

In the female lineage this may be the result of meiotic drive.
Recall that in the formation of the female gamete, or megagamete, the cytoplasm is unequally shared among the cells, resulting in a single gamete as opposed to the four gametetes, microgametes, that are formed in the male lineage.
When certain alleles, combinations of alleles, or chromosomes, are preferentially observed in the population of megagametes we refer to this as meiotic drive.




Chromosome 7 exhibited equal allele frequencies (*p* ≈ *q* ≈ 0.5) but these alleles combined to form an excess of heterozygotes (*F~IS~* < 0).
This suggests a heterozygote advantage and selection on the genotype and not the individual alleles.
This high degree of heterozygosity may be a sign of genetic load, a condition where heterozygosity results in one allelic form being non-functional but the phenotype is unaffected because a single copy is all that is necessary for survival.
As long as the non-functional form is paired with a functional form, as a heterozygote, there is no affect.
But homozygotes for the non-functional form become deleterious.
Hemp is considered to not tolerate inbreeding, candidates for genes associated with this trait may be located on chromosome 7 as it attempts to harbor genetic load that is not present elsewhere in the genome.


Chromosome 3 exhibited both unequal allele frequencies (*p* ≈ 0.6, *q* ≈ 0.4) and these alleles combined to form an excess of heterozygotes (*F~IS~* < 0).


Note that while much of each chromosome behaved similar to a single locus, the ends of each chromosome demonstrated a return to the random expectation.
This is likely due to a low amount of recombination throughout much of the chromosome, resulting in each chromosome behaving as a single locus.
However the gene dense regions at the end of each chromosome are where recombination was observed and indicates a mechanism for this region of each chromosome to exhibit random segregation.






## References

FASTA2nuccomp

motif_counter [no citation???]

Onofri [https://doi.org/10.1016/j.phytochem.2015.03.006]


## Figure captions


**Figure 1** Genomic landscape of *Cannabis sativa* strain ERBxHO40.
**Panel A**
The nine pairs of autosomes and pair of sex chromosomes (XX).
One million base pair rectangular windows extend outward from each pair of haplotypes at a width proportion to the absence of the 'CpG' motif.
Each rectangular window is colored by gene density with yellow and orange indicating high gene density and blue and purple indicating low gene density.
Each pair of haplotypes is connected by polygons to indicate structural arrangement, with grey indicating syntenous regions and orange connecting inversions.
Rectangles along each haplotype indicate select loci with 
light blue indicating centromeric and subtelomeric repeats, 
firebrick red indicating 45S, 26S, 5.8S, 18S RNA arrays, black indicating 5S RNA arrays,
and forest green indicating cannabinoids (cannabichromenic acid synthase, cannabidiolic acid synthase, tetrahydrocannabinolic acid synthase, and olivetolic acid cyclase).
**Panel B** The inheritance of alleles throughout the genome.
The upper panel presents the frequency of each allele throughout the genome.
This population (n = 270 plants) is the product of a selfed parent.
The allele frequencies for heterozygous positions in thre parent should be 0.5 and our random, or evolutionarily neutral, expectation is that this frequency of 0.5 will be observed in the progeny population.
The lower panel presents *F~IS~* or the deviation from our random, or evolutionarily neutral, expectation of heterozygosity.
A positive *F~IS~* indicates a deficiency of heterozygotes, a negative *F~IS~* indicates an excess of heterozygotes and an *F~IS~* of zero indicates that our allele frequencies predict heterozygosity.


