---
title: "SynChrom"
output: 
  html_document:
    toc: true
    code_folding: hide
date: "2023-03-14"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.align = "center")
```



## Outline


- Technical accomplishment
  - 39 diploid annotated assemblies or 78 scaffolded haplotypes
  - ~100 unscaffolded, annotated, halploid assemblies


- Structure of a population of genomes / Pangenomic applications
  - EH23 assembly: within genome variation, genic organization, chromosome lengths, structural variation, chrom 7 cannabinoid synthases, chrom 8 acrocentric
  - Segregation distortion: F2 vs scaffolded panel
  - TEs
  - Structural Variation


- Case studies / Gene classes
  - R-genes
  - Terpenes
  - Cannabinoid synthases
  - Cannabinoid pathway


- Natural and Artificial history
  - Plant classes and differentiation



## PAG abstract

Hemp genome history.

EH23

Organization

Rgenes

PK > PK F > JL > CBDRx > CB2
EH23 genome structure: genes, recomb, centromeres
Cann Syn
Rgenes

250 words???

Brian J. Knaus^1^, Kelly J. Vining^1^, Others^2^

^1^Department of Horticulture, Oregon State University, Corvallis, OR 97331, USA    
^2^The Plant Molecular and Cellular Biology Laboratory, The Salk Institute for Biological Studies, La Jolla, CA 92037, USA

Genomics of *Cannabis* L. strives to advance our understanding of how this plant contributes as a grain, fiber, and oil (CBCA, CBDA, THCA, and others) crop, as well as a part of the development of human culture.
We present pangenomic work that has resulted in 78 chromosome scale haplotypes assembled from 39 plants.
A focal point of these assemblies is the EH23 assembly that illustrated length polymorphism among pairs of chromosomes as well as inversions, demonstrating substantial among haplotype diversity.
Genotyping of a selfed population exhibited segregation distortion where alleles on some chromosomes were non-randomly inherited and alleles combined to form genotypes non-randomly as well, resulting in a heterozygote deficit or excess that differentiated by chromosome.
Near chromosomal level assemblies for chromosome 7 provided a perspective on CBDAS and THCAS alleles that associate with plant chemotype where THCAS type plants are THCAS:THCAS, CBDAS type plants are CBDAS:CBDAS, and balanced chemotype plants are CBDAS:THCAS.
These haplotypes provide perspectives that facilitate comparison among assemblies to help validate and provide inference on structural organization within each *Cannabis* chromosome.




Genome assembly within *Cannabis* has proceeded from the assembly of the THC strain Purple Kush, with an N50 of 16.2 Kb (van Bakel et al. 2011), to the resequencing of Purple Kush (N50 of 61.0 Mbp) accompanied by the food and fiber type Finola (N50 of 77.1 Mbp; Laverty et al. 2018), the putatively wild plant from Asia ('jl', N50 of 83.0 Mbp; Gao et al. 2020), the CBD strain Cannbio-2 (N50 of 91.5 Mbp; Braich et al. 2020), the CBD strain CBDRx (N50 of 91.9 Mbp; Grassa et al. 2021) and other assemblies.












## Methods


**Ideogram methods**

The genetically female EH23 genome was expected to consist of 9 pairs of autosomes and a pair of sex chromosomes (XX).
Ideograms for each pair of chromosomes were created using ggplot2 [https://ggplot2.tidyverse.org] in R [https://www.R-project.org/].
The length of each chromosome was determined using 'nuccomp.py' [https://github.com/knausb/nuccomp] and used with 'ggplot::geom_rect()' to initialize the plot.
One million base pair windows were created for each chromosome where the number of 'CpG' motifs were counted for each window with the program 'motif_counter.py' [https://github.com/knausb/nuccomp].
The CpG count was converted into a rate by dividing by the window size, this also accommodated the last window of each chromosome which was less than one million base pairs in size.
These rates were scaled by subtracting the minimum rate and then dividing by the maximum rate (the maximum rate after subtracting the minimum rate), on a per chromosome basis.
In order to visually emphasize the enrichment of the 'CpG' motif in the centromeric region, an inverse of the CpG rate was taken by taking one and subtracting the 'CpG' rate for each window.
This scaled, inverse 'CpG' rate was used for the width of each one mbp window and colored based on gene density using the viridis magma palette [https://doi.org/10.5281/zenodo.4679424].


Structural variation among each pair of chromosomes was determined using minimap2 [https://doi.org/10.1093/bioinformatics/bty191].
The minimap2 comparisons were annotated using SyRI [https://doi.org/10.1186/s13059-019-1911-0].
The syntenous and inverted regions were plotted using ggplot2::geom_polygon() in a manner inspired by plotsr [https://doi.org/10.1093/bioinformatics/btac196] but implemented in R [https://github.com/ViningLab/CannabisPangenome].


The location of candidate loci within EH23 haplotypes A and B were determined using BLASTN [https://doi.org/10.1016/S0022-2836(05)80360-2].
Query sequences were as follows: cannabichromenic acid synthase (LY658671.1), cannabidiolic acid synthase (AB292682, AB292683, AB292684), tetrahydrocannabinolic acid synthase (AB212829, AB212830), and olivetolic acid cyclase (NC_044376.1:c4279947-4279296, NC_044376.1:c4272107-4271242).
These sequences were combined with centromeric, telomeric and rRNA sequences in the file 'blastn_queries_rrna_cann.fasta' [https://github.com/ViningLab/CannabisPangenome].
BLASTN was called with the following options: -task megablast -evalue 0.001 -perc_identity 90 -qcov_hsp_perc 90.
Tabular results (subject chromosome, subject start of alignment, subject end of alignment) from BLASTN were read into R [https://www.R-project.org/] and plotted on ideograms with ggplot2::geom_rect() [https://ggplot2.tidyverse.org].


**Allele frequency methods**


Genotype data in the VCF format [https://doi.org/10.1093/bioinformatics/btr330] was input into R [https://www.R-project.org/] using vcfR [https://doi.org/10.1111/1755-0998.12549].
Allele and heterozygote counts were made with vcfR [https://doi.org/10.1111/1755-0998.12549].
Wright's *F~IS~* [https://doi.org/10.1038/166247a0] was calculated to provide the deviation in heterozygosity from our random expectation.
Wright's *F~IS~* was calculated as (*H~S~* - *H~O~*)/*H~S~* where *H~O~* is the observed number of heterozygotes divided by their number and *H~S~* is the number of heterozygotes we expect based on the allele frequencies, calculated as the frequency of the first allele multiplied by the frequency of the second multiplied by two and divided by their number. 
Scatterplots were generated using ggplot2 [https://ggplot2.tidyverse.org].
Graphical panels were assembled into a single graphic using ggpubr [https://CRAN.R-project.org/package=ggpubr].



## Results


**Organization of the EH23 genome**

The EH23 *Cannabis sativa* genome was organized into 9 pairs of autosomes and a pair of sex chromosomes (Figure 1A).
Each pair of chromosomes were different in chromosome length but also demonstrated a large degree of synteny among each pair.
Large inversions were present on chromosome 7 with smaller inversions distributed throughout the remaining chromosomes.
Genes were located predominantly near the ends of each chromosomes, as opposed to being distributed evenly throughout each chromosome.
This pattern of gene location near the ends of the chromosomes is consistent with previous reports [https://doi.org/10.1101/gr.242594.118].
The abundance of the motif 'CpG' (cytosine, phosphate, guanine; a motif thought to be important in the context of methylation) varied throughout each pair of chromosomes.
This CpG abundance was associated with centromeric repeats, suggesting it can be used to determine approximate centromeric region.
Each of the 10 pairs of *Cannabis sativa* chromosomes appeared largely syntenic with an abundance of CpG motifs associated with centromeric regions.


**Nonrandom inheritance of SNP markers**


Alleles inherited through a single generation, resulting from a controlled cross, in a population of plants grown in a common garden, demonstrated both patterns of random and nonrandom inheritance.
The parental population consisted of a self-pollinated individual, so allele frequencies at variable (heterozygous) positions were expected to be 0.5.
Our random, or evolutionarily neutral, expectation is that allele frequencies will be maintained from generation to generation.
That is, the allele frequencies in the parental population (a single individual) should approximate the allele frequencies observed in the progeny population (270 progeny plants).
Similarly, our random, or evolutionarily neutral, expectation is that these alleles will combine to form genotypes randomly to meet our Hardy-Weinberg expectation (heterozygous genotypes at a rate of 2*pq* = 0.5; *F~IS~* = 0).
Variants in the EH23 *Cannabis sativa* genome demonstrated both neutral as well as non-random patterns of inheritance, typically at the chromosomal level.


A random, or evolutionarily neutral, pattern of inheritance was observed on throughout nuch of chromosome 4 (Figure 1B).
The allele frequencies were near 0.5 for both the major and minor allele, and these alleles combined into genotypes at a rate near our neutral expectation (*F~IS~* = 0).
Loci on chromosome 2 demonstrated deviations from our neutral expectation.
The major allele on chromosome 2 was observed at a rate near 0.6 and the minor allele was observed to be near 0.4.
While the alleles were unequally inherited, they combined to form genotypes in a neutral, or random, manner (*F~IS~* $\approx$ 0).
The alleles on chromosome 5 were inherited at a rate near our neutral expectation (major and minor allele frequencies near 0.5), but these alleles combined to form an excess of homozygotes relative to our neutral expectation (*F~IS~* > 0).
The alleles on chromosome 3 were inherted with a major allele frequency of approximately 0.6 and a minor allele near 0.4.
The alleles on chromosome 3 did not combine to form genotypes in a neutral manner but instead exhibited a heterozygote excess (*F~IS~* < 0).
Deviations from our neutral expectations were observed throughout each chromosome, where observed, with the exception of the very ends of each chromosome.



Chromosome 2 also exhibited our neutral assumptions where the frequencies of each allele was at 0.5.
However, these alleles combined to form genotypes in a manner where homozygotes are favored, resulting in a positive *F~IS~*.
Chromosome 3 demonstrated a deviation in allele frequency from our neutral expectation of 0.5, where one allele was over-represented in the progeny.
Despite this substantial deviation in allele frequency, these alleles joined together to form genotypes in a neutral manner, resulting in an *F~IS~* near zero and a heterozygote frequency near 0.5.
Chromosome 4 had a non-neutral deviation in allele frequency that was similar to that of chromosome 3.
However, on chromosome 4 the alleles joined to form genotypes in a manner where heterozygotes were favored, as indicated by a negative *F~IS~*.


Near the very ends of the chromosomes, where most of the genes occur and recombination is observed [https://doi.org/10.1101/gr.242594.118], we tended to observe our neutral expectations on all chromosomes.
For example, much of chromosome 1 demonstrated a deviation from our neutral allele frequency of 0.5 yet the ends of the chromosomes returned to our neutral expectation of 0.5.
This suggests that the predominantly non-recombining chromosomes behave as a single locus except at the ends where recombination may allow a return to neutrality.


Throughout the hemp genome we observed different patterns of evolution.
What is striking about these results is that different evolutionary scenarios were exhibited on different chromosomes.
For example, the positive *F~IS~* on chromosome may cause us to interpret it as a result of inbreeding, while the negative *F~IS~* on chromosome 4 may cause us to interpret it as a result of outbreeding.
These are phenomena we typically discuss as a property of a population or individuals within a population, where in the hemp genome we observe these phenomena to differentiate among chromosomes.


## Discussion





**History of Cannabis genomics**

The genomics age for *Cannabis* was initiated with a short read assembly of the THC-type strain Purple Kush [https://doi.org/10.1186/gb-2011-12-10-r102].


**Purple Kush** N50 = 16.2 Kb, 0.0162 Mbp
A combination of Illumina short read sequencing and 454 pyrosequencing was used to create an assembly consisting of 786.6 Mbp organized into 136,290 scaffolds with an N50 of 16.2 Kbp.

GCA_000230575.1


**Purple Kush and Finola** N50 = 61.0 Mbp (PK) and 77.1 Mbp (Finola)
Laverty and colleagues presented a comparative genomics perspective based on the genomes of a THC-type strain (Purple Kush) and a CBD-type strain (Finola) [https://doi.org/10.1101/gr.242594.118].

**jl_Jilong** N50= 83.0 Mbp
Gao and colleagues contributed a genomic assembly from a plant from Southern China that is putatively wild [https://doi.org/10.1038/s41438-020-0295-3].


**CBDRx** N50= 91.9 Mbp
Grassa and colleagues contributed a genomic assembly from the CBD-type strain CBDRx [https://doi.org/10.1111/nph.17243].


**Cannbio-2** N50 = 91.5 Mbp





### Chromosomal organization

The phased diploid assembly of *Cannabis sativa* L. strain EH23 provides new perspectives on the hemp genome.
The EH23 assembly's nine pairs of automosomes and the pair of sex chromosomes (XX) were numbered and oriented according to the haploid CBDRx assembly (GCF_900626175.2)[https://doi.org/10.1111/nph.17243].
The ends of the chromosomes are enriched for gene presence, as reported previously [https://doi.org/10.1101/gr.242594.118], except for chromosome 8 which did not have a high density of genes at one end.

Length polymorphism was evident between each pair of chromosomes.
While some of this difference in length may be due to assembly efficiency, 




### Inheritance of alleles

prezygotic versus postzygotic


Inheritance of genetic material from generation to generation is expected to occur in a random, or evolutionarily neutral, manner.
This should be observed in a large population as equal allele frequencies that randomly combine to form genotypes that satisfy our Hardy–Weinberg proportions.
This was observed on chromosome 4 where each allele frequency was equal (*p* ≈ *q* ≈ 0.5) and no deviation in our random expectation of heterozygosity was observed (*F~IS~* ≈ 0).
Elsewhere in the genome, deviations from our neutral expectation were observed.


Chromosome 2 demonstrated allele frequencies that substantially deviated from equal (*p* ≈ 0.6, *q* ≈ 0.4). 
However, these alleles combined to form genotypes at a proportion that met our random expectation.
This suggests selection on the alleles but not the genotypes.
Because there was a single, self pollinated, parent, we expect the allele frequencies and the genotype frequencies to be equal in the parent, indicating that the selection event occurred after the parent, and before the formation of genotypes.
In the male lineage this conversation typically involves sperm competition.

In the female lineage this may be the result of meiotic drive.
Recall that in the formation of the female gamete, or megagamete, the cytoplasm is unequally shared among the cells, resulting in a single gamete as opposed to the four gametetes, microgametes, that are formed in the male lineage.
When certain alleles, combinations of alleles, or chromosomes, are preferentially observed in the population of megagametes we refer to this as meiotic drive.




Chromosome 7 exhibited equal allele frequencies (*p* ≈ *q* ≈ 0.5) but these alleles combined to form an excess of heterozygotes (*F~IS~* < 0).
This suggests a heterozygote advantage and selection on the genotype and not the individual alleles.
This high degree of heterozygosity may be a sign of genetic load, a condition where heterozygosity results in one allelic form being non-functional but the phenotype is unaffected because a single copy is all that is necessary for survival.
As long as the non-functional form is paired with a functional form, as a heterozygote, there is no affect.
But homozygotes for the non-functional form become deleterious.
Hemp is considered to not tolerate inbreeding, candidates for genes associated with this trait may be located on chromosome 7 as it attempts to harbor genetic load that is not present elsewhere in the genome.


Chromosome 3 exhibited both unequal allele frequencies (*p* ≈ 0.6, *q* ≈ 0.4) and these alleles combined to form an excess of heterozygotes (*F~IS~* < 0).


Note that while much of each chromosome behaved similar to a single locus, the ends of each chromosome demonstrated a return to the random expectation.
This is likely due to a low amount of recombination throughout much of the chromosome, resulting in each chromosome behaving as a single locus.
However the gene dense regions at the end of each chromosome are where recombination was observed and indicates a mechanism for this region of each chromosome to exhibit random segregation.






## References

nuccomp [https://github.com/knausb/nuccomp]

motif_counter [no citation???]

Onofri [https://doi.org/10.1016/j.phytochem.2015.03.006]


## Figure captions


**Figure 1** Genomic landscape of *Cannabis sativa* strain ERBxHO40.
**Panel A**
The nine pairs of autosomes and pair of sex chromosomes (XX).
One million base pair rectangular windows extend outward from each pair of haplotypes at a width proportion to the absence of the 'CpG' motif.
Each rectangular window is colored by gene density with yellow and orange indicating high gene density and blue and purple indicating low gene density.
Each pair of haplotypes is connected by polygons to indicate structural arrangement, with grey indicating syntenous regions and orange connecting inversions.
Rectangles along each haplotype indicate select loci with 
light blue indicating centromeric and subtelomeric repeats, 
firebrick red indicating 45S, 26S, 5.8S, 18S RNA arrays, black indicating 5S RNA arrays,
and forest green indicating cannabinoids (cannabichromenic acid synthase, cannabidiolic acid synthase, tetrahydrocannabinolic acid synthase, and olivetolic acid cyclase).
**Panel B** The inheritance of alleles throughout the genome.
The upper panel presents the frequency of each allele throughout the genome.
This population (n = 270 plants) is the product of a selfed parent.
The allele frequencies for heterozygous positions in thre parent should be 0.5 and our random, or evolutionarily neutral, expectation is that this frequency of 0.5 will be observed in the progeny population.
The lower panel presents *F~IS~* or the deviation from our random, or evolutionarily neutral, expectation of heterozygosity.
A positive *F~IS~* indicates a deficiency of heterozygotes, a negative *F~IS~* indicates an excess of heterozygotes and an *F~IS~* of zero indicates that our allele frequencies predict heterozygosity.




**Figure X** Cannabinoid synthase pathway genes of *Cannabis sativa* strain ERBxHO40.
**Panel A**

Chromosome 7 included a single high stringency match for either tetrahydrocannabinolic acid synthase or cannabidiolic acid synthase, associated with chemotype, as well as a population of low stringency matches.
For each diploid plant's chromosome (x-axis) colored points indicate the location of BUSCO (Benchmarking Universal Single-Copy Orthologs) genes for chromosome 7 (elevation for each point is in million base pairs).
Lines connect identical BUSCO genes among neighboring samples.
Changes in the elevation and spacing among BUSCO genes reflects structural variation within samples, while the lines draw attention to variability among each sample's immediate neighbor.
Triangles mark the location of matches for three cannabinoid synthases (CBCAS, CBDAS, and THCAS).
Red triangles indicate high stringency matches to THCAS (BLASTN query of NCBI accession "AB212829.1", percent identity >= 0.98, number of gaps = 0).
Blue triangles indicate high stringency matches to CBDAS (BLASTN query of NCBI accession "AB292682.1", percent identity >= 0.98, number of gaps = 0).
Yellow triangles indicate high stringency matches to CBCAS (BLASTN query of NCBI accession "LY658671.1", percent identity >= 0.98, number of gaps = 0).
Gray triangles indicate low stringency matches to cannabinoid synthases (BLASTN query of cannabichromenic acid synthase (LY658671.1), cannabidiolic acid synthase (AB292682, AB292683, AB292684), and tetrahydrocannabinolic acid synthase (AB212829, AB212830), percent identity >= 0.90 and < 0.98, number of gaps > 0).
Each plant that was characterized as predominantly producing THCA (Chemotype I) consisted of a pair of chromosomes (a and b) that each had a single match to tetrahydrocannabinolic acid synthase (AB212829.1), indicated by a single red triangle.
Each plant that was characterized as predominantly producing CBDA (Chemotype III) consisted of a pair of chromosomes (a and b) that each had a single match to cannabidiolic acid synthase (AB292682.1), indicated by a single blue triangle.
Each plant that was characterized as producing both THCA and CBDA (Chemotype II) consisted of a 
pair of chromosomes where one chromosome included a single match to tetrahydrocannabinolic acid synthase (AB212829.1), indicated by a single red triangle, and the other chromosome included a single match to cannabidiolic acid synthase (AB292682.1), indicated by a single blue triangle.
Plants that were characterized as being cannabigerolic acid accumulators (Chemotype IV) included chromosomes that may have had a single match to tetrahydrocannabinolic acid synthase (AB212829.1) or no high stringency matches to any of the three assayed cannabinoid synthases.
Gray triangles indicate that a substantial population of low stringency matches were observed on all of the chromosome 7s.




