---
title: "Varin GWAS"
author: "Brian J. Knaus"
date: "`r format(Sys.time(), '%Y, %B %d')`"
output: 
  html_document:
    toc: true
editor_options: 
  chunk_output_type: console
---



https://www.ncbi.nlm.nih.gov/nuccore/?term=Cannabis%20ALT4

https://www.ncbi.nlm.nih.gov/search/all/?term=Cannabis%20BKR



```{r setup, include=FALSE}
#knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(fig.align = "center")
knitr::opts_chunk$set(fig.width = 12)
knitr::opts_chunk$set(results = "hide")

```


## Methods





## GAPIT Blink


```{r}
t1 <- Sys.time()
#library(vcfR)
```


```{r}
gwas <- read.csv("GWAS_blink_varin.csv.gz")
gwas[1:3, 1:5]
```


```{r}
nucs <- read.csv("nucs.csv.gz")
nucs[1:3, 1:5]
```



```{r}
my_pal <- c("#420A6866", "#D94D3D66", "#FCAF1366", "#BF395266", "#81206C66", 
"#62146E66", "#ED692566", "#A12A6366", "#F98B0B66", "#F6D64566"
)
```


```{r, echo=FALSE, fig.width=12, fig.cap="**Figure X** Manhattan plot."}
# png(filename = "varin_total_gwas.png", width = 6.5, height = 4, units = "in",
#    res = 300)


palette(my_pal)

plot(x = gwas$POS, y = -log10(gwas$P.value), 
#     col = vat$Chromosome,
     col = gwas$chrom_num,
     pch = 20, xlab = "", ylab = "", xaxt = "n")

abline(h = seq(0, 30, by = 1), lty = 3, col = "#C0C0C088")
abline(h = seq(0, 30, by = 5), lty = 1, col = "#C0C0C088")

axis(side = 1, at = nucs$mids, labels = c(1:9, "X"))
#axis(side = 1, at = nuc$lpos, labels = c(1:9, "X"))
#title( xlab = "Chromosome (CBDRx; GCF_900626175.2)", line = 2 )
title( xlab = "Chromosome (EH23b)", line = 2 )
title( ylab = expression("-Log"["10"](italic(p)-value)), line = 1.8 )
title( main = "Varin content (Blink model; Varin_Total)" )


# abline( v = c(69653599 + nucs$starts[4], 69653748 + nucs$starts[4]),
#         col = "#1E90FF",
# #        col = "#B22222",
#         lty = 3, lwd = 2)


#palette("default")
#dev.off()
```




```{r, echo=FALSE, fig.width=12, fig.cap="**Figure X** Manhattan plot. Points at the top of the panel are large and not transparent, points at the bottom of the panel are small and transparent."}
# png(filename = "varin_total_gwas.png", width = 6.5, height = 4, units = "in",
#    res = 300)


palette(my_pal)
gwas$col <- my_pal[gwas$chrom_num]
gwas$col <- substr(gwas$col, start = 1, stop = 7)
# #max(-log10(gwas$P.value))
# #round(-log10(gwas$P.value)/24 * 10)

 
# my_ps <- -log10(gwas$P.value)
# 
# #as.integer(as.hexmode(my_ps))
# 
# my_ps <- my_ps/max(my_ps) * 59
# my_ps <- my_ps + 40
# my_ps <- round(my_ps)
# range(my_ps)
# #hist(my_ps)
# 
# #format(my_ps, width = 2)
# 
# gwas$col <- paste(gwas$col, formatC(my_ps, width = 2, format = "d", flag = "0"), sep = "")



my_ps <- -log10(gwas$P.value)
#as.integer(as.hexmode(my_ps))
my_ps <- my_ps/max(my_ps) * 235
my_ps <- my_ps + 20
my_ps <- round(my_ps)
range(my_ps)
my_ps <- as.hexmode(my_ps)
#my_ps %% 16
#table(nchar(my_ps))

#my_ps[nchar(my_ps) > 2]

my_ps <- format(my_ps, width = 2)

#gwas$col <- paste(gwas$col, formatC(my_ps, width = 2, format = "d", flag = "0"), sep = "")



gwas$col <- paste(gwas$col, my_ps, sep = "")
table(nchar(gwas$col))

#gwas$col[nchar(gwas$col) > 9]



plot(x = gwas$POS, y = -log10(gwas$P.value), 
#     col = vat$Chromosome,
     #col = gwas$chrom_num,
     col = gwas$col,
     pch = 20, xlab = "", ylab = "", xaxt = "n",
     cex = -log10(gwas$P.value) * 0.20)
     #cex = -log10(gwas$P.value) * 0.14)
#     cex = 1-abs(gwas$P.value) * 1.1)
#     cex = abs(gwas$Effect) * 0.1)

abline(h = seq(0, 30, by = 1), lty = 3, col = "#C0C0C088")
abline(h = seq(0, 30, by = 5), lty = 1, col = "#C0C0C088")

axis(side = 1, at = nucs$mids, labels = c(1:9, "X"))
#axis(side = 1, at = nuc$lpos, labels = c(1:9, "X"))
#title( xlab = "Chromosome (CBDRx; GCF_900626175.2)", line = 2 )
title( xlab = "Chromosome (EH23b)", line = 2 )
title( ylab = expression("-Log"["10"](italic(p)-value)), line = 1.8 )
title( main = "Varin content (Blink model; Varin_Total)" )


# abline( v = c(69653599 + nucs$starts[4], 69653748 + nucs$starts[4]),
#         col = "#1E90FF",
# #        col = "#B22222",
#         lty = 3, lwd = 2)


#palette("default")
#dev.off()
```


Coordinates can be checked at the Michael Lab genome browser (JBrowse).
https://resources.michael.salk.edu/resources/cannabis_genomes/index.html


```{r, results='asis'}
#plot(gwas$P.value, gwas$H.B.P.Value)
#gwas[1:3, ]
gwas2 <- gwas[sort.int(gwas$P.value, decreasing = FALSE, index.return = TRUE)$ix, ]
gwas3 <- data.frame( gwas2[, c(1, 4)],
                     '-log10(P)' = -log10(gwas2$P.value),
                     gwas2[, c(5, 6, 7, 8, 9, 3)]
                     )
names(gwas3)[3] <- "-log10(P)"
gwas3 <- gwas3[1:10, ]
#gwas3[1:3, ]


knitr::kable(gwas3, 
             caption = "**Table X.** Select results from GAPIT GWAS using the Blink model, sorted by P.value.",
             row.names = FALSE)
```




## Epistasis


```{r}
library(vcfR)
```


```{r, results='hide'}
vcf <- read.vcfR("varin_Blink.vcf.gz")
```


```{r}
vcf
gt <- extract.gt(vcf)
#is_het(gt)
gt2 <- matrix( nrow = nrow(gt), ncol = ncol(gt), dimnames = dimnames(gt) )

gt2[ gt == "0/0" ] <- 0
gt2[ gt == "0/1" ] <- 1
gt2[ gt == "1/1" ] <- 2
#gt <- apply(gt, MARGIN = 2, as.numeric)

```



```{r}
pheno <- read.csv('ERBxHO40_Phenotypes.csv')
pheno[1:3,]
```


### Synchronize


```{r, results='asis'}
pheno <- pheno[colnames(gt), ]
all(pheno$ID == colnames(gt))
```


### Phenotype

```{r, fig.width=6}
hist(pheno$Varin_Total, breaks = seq(0, 65, by = 2))
```


### lm 4_68382150

```{r}
# #gt[, 1:4]
# lm1 <- lm( pheno$Varin_Total ~ gt2["4_20005812", ] )
# #lm1 <- lm( pheno$Varin_Total ~ gt[1, ] : gt[2, ] )
# summary(lm1)
```


```{r, fig.width=6}
# plot(gt2["4_20005812", ], pheno$Varin_Total)
# abline(lm1)
```


```{r}
my_dat <- data.frame(
  Varin_Total = pheno$Varin_Total,
  gt = gt2["4_68382150", ],
#  gt = gt2["4_20005812", ],
  pcol = "#000000"
#  gt_4_20005812 = gt2["4_20005812", ]
#  gt_7_290913 = gt2["7_290913", ]
)
my_dat$pcol[ my_dat$gt == 2 ] <- "#B22222"
# my_dat$gt <- my_dat$gt + 1

my_dat2 <- my_dat
#my_dat2 <- my_dat[ my_dat$gt != 2, ]

lm1 <- lm( Varin_Total ~ gt, data = my_dat2 )
# lm11 <- lm( Varin_Total ~ gt + I(gt^2), data = my_dat2 )
lm11 <- lm( Varin_Total ~ gt + I(gt^2), data = my_dat )


summary(lm1)

aggregate(Varin_Total ~ gt, data = my_dat, mean)

# x <- 1
# x <- 2
# lm11$coefficients[1] + x * lm11$coefficients[2] + x^2 * lm11$coefficients[3]

my_vals <- data.frame(
  x = seq(0, 2, length.out=10)
)
my_vals$y <- lm11$coefficients[1] + my_vals$x * lm11$coefficients[2] + my_vals$x^2 * lm11$coefficients[3]

```



```{r, results='markup'}
library(pander)
#pander(lm1)
pander(summary(lm1))
pander(summary(lm11))
```


```{r, fig.width=6}
# plot( jitter(my_dat$gt, factor  = 0.6), my_dat$Varin_Total)
# abline(lm1)
```


```{r, fig.width=6, fig.cap="**Figure X.** Ability of the genotype to predict the phenotype."}
library(ggplot2)
#my_dat$gt <- as.factor(my_dat$gt)
p <- ggplot( data = my_dat, mapping = aes( x = gt, y = Varin_Total, fill = gt, group = gt ))
p <- p + theme_bw()
p <- p + geom_violin()
p <- p + scale_x_continuous( 
  breaks = c(0, 1, 2),
  labels = c("0/0", "0/1", "1/1")
)

#p <- p + geom_violin( scale = "count" )
#p <- p + scale_fill_manual(values=c("#999999", "#E69F00", "#56B4E9"))
#p <- p + scale_fill_continuous(values=c("#999999", "#E69F00", "#56B4E9"))
# 
p <- p + theme(legend.position="none")
p <- p + scale_fill_continuous(type = "viridis", begin = 0.2)
#p <- p + geom_jitter(shape=16, position=position_jitter(0.1), col = "#B22222")
p <- p + geom_jitter(shape=16, position=position_jitter(0.15), col = my_dat$pcol, size = 2)
p <- p + ylab("Varin Total (units???)")
#p <- p + xlab("Genotype at chromosome 4, position 20,005,812 bp")
p <- p + xlab("Genotype")
#p <- p + ggtitle("Chromosome 4, position 20,005,812 bp")
p <- p + ggtitle("Chromosome 4, position 68,382,150 bp")
#
p <- p + geom_abline( slope = lm1$coefficients[2], intercept = lm1$coefficients[1], linewidth = 2, col = "#808080")
p

p + geom_line( data = my_vals, mapping = aes( x = x, y = y ), inherit.aes = F, linewidth = 2, col = "#808080" )

```


### lm 7_290913


```{r}
#gt[, 1:4]
#lm2 <- lm( pheno$Varin_Total ~ gt2["7_290913", ] )
#summary(lm2)
```


```{r}
my_dat <- data.frame(
  Varin_Total = pheno$Varin_Total,
  gt = gt2["7_290913", ],
#  gt = gt2["4_20005812", ],
  pcol = "#000000"
#  gt_4_20005812 = gt2["4_20005812", ]
#  gt_7_290913 = gt2["7_290913", ]
)
my_dat$pcol[ my_dat$gt == 2 ] <- "#B22222"
#
my_dat2 <- my_dat
#my_dat2 <- my_dat[ my_dat$gt != 2, ]
lm2 <- lm( Varin_Total ~ gt, data = my_dat2 )
lm2 <- lm( Varin_Total ~ gt * gt, data = my_dat2 )
lm2 <- lm( Varin_Total ~ gt:gt, data = my_dat2 )


summary(lm2)
```


```{r, results='markup'}
library(pander)
pander(summary(lm2))
```



```{r, fig.width=6, fig.cap="**Figure X.** Ability of the genotype to predict the phenotype."}
library(ggplot2)
#my_dat$gt <- as.factor(my_dat$gt)
p <- ggplot( data = my_dat, mapping = aes( x = gt, y = Varin_Total, fill = gt, group = gt ))
p <- p + theme_bw()
p <- p + geom_violin()
p <- p + scale_x_continuous( 
  breaks = c(0, 1, 2),
  labels = c("0/0", "0/1", "1/1")
)

p <- p + theme(legend.position="none")
p <- p + scale_fill_continuous(type = "viridis", begin = 0.2)
p <- p + geom_jitter(shape=16, position=position_jitter(0.15), col = my_dat$pcol, size = 2)
p <- p + ylab("Varin Total (units???)")
p <- p + xlab("Genotype")
#p <- p + ggtitle("Chromosome 4, position 20,005,812 bp")
#p <- p + ggtitle("Chromosome 4, position 68,382,150 bp")
p <- p + ggtitle("Chromosome 7, position 290,913 bp")
#
p <- p + geom_abline( slope = lm2$coefficients[2], 
                      intercept = lm2$coefficients[1],
                      linewidth = 2, col = "#808080")
p
```


### Epistasis: interaction among loci


```{r, results='markup'}
#lm2 <- lm( pheno$Varin_Total ~ gt2["4_20005812", ] )
# lm3 <- lm( pheno$Varin_Total ~ gt2["4_20005812", ] * gt2["4_20005812", ] )
# lm3 <- lm( pheno$Varin_Total ~ poly(gt2["4_20005812", ], degree = 1, raw = T) )
# 
# lm3 <- lm( pheno$Varin_Total ~ gt2["4_20005812", ] + I(gt2["4_20005812", ]^2) )
# lm3 <- lm( pheno$Varin_Total ~ gt2["4_20005812", ] + gt2["4_20005812", ]:gt2["4_20005812", ] )
# lm3 <- lm( pheno$Varin_Total ~ gt2["4_20005812", ] * gt2["7_290913", ] )

lm3 <- lm( pheno$Varin_Total ~ gt2["4_68382150", ] + gt2["7_290913", ] )
#summary(lm3)
lm4 <- lm( pheno$Varin_Total ~ gt2["4_68382150", ] * gt2["7_290913", ] )
#lm4 <- lm( pheno$Varin_Total ~ gt2["4_68382150", ] * gt2["7_290913", ] + gt2["4_68382150", ]:gt2["4_68382150", ])

lm5 <- lm( pheno$Varin_Total ~ gt2["4_68382150", ] * gt2["7_290913", ] + I(gt2["4_68382150", ]^2) )

```


```{r, results='markup'}
library(pander)
pander(summary(lm3))
pander(summary(lm4))
pander(summary(lm5))
```



```{r, fig.width=6}
# plot(gt2["7_290913", ], pheno$Varin_Total)
# coefficients(lm3)
# abline(lm3)
```








```{r}
#lm2 <- lm( pheno$Varin_Total ~ gt2[1, ] + I(gt2[1, ]^2) )
#lm2 <- lm( pheno$Varin_Total ~ poly(gt2[1, ], 2, raw = FALSE) )
#summary(lm2)
```


## Model building


```{r, results='markup'}
lm1 <- lm( pheno$Varin_Total ~ gt2["4_68382150", ] + I(gt2["4_68382150", ]^2) )
# lm2 <- lm( pheno$Varin_Total ~ gt2["4_68382150", ] + I(gt2["4_68382150", ]^2) + 
#              gt2["4_36256098", ] + I(gt2["4_36256098", ]^2) + 
#              gt2["4_68382150", ]:gt2["4_36256098", ] )
lm2 <- lm( pheno$Varin_Total ~ gt2["4_68382150", ] + I(gt2["4_68382150", ]^2) + 
             gt2["4_36256098", ] + I(gt2["4_36256098", ]^2) )
pander(summary(lm2))

lm3 <- lm( pheno$Varin_Total ~ gt2["4_68382150", ] + I(gt2["4_68382150", ]^2) + 
             gt2["4_36256098", ] + I(gt2["4_36256098", ]^2) + 
             gt2["X_12488664", ] + I(gt2["X_12488664", ]^2) )
pander(summary(lm3))

# lm4 <- lm( pheno$Varin_Total ~ gt2["4_68382150", ] + I(gt2["4_68382150", ]^2) + 
#              gt2["4_36256098", ] + I(gt2["4_36256098", ]^2) + 
#              gt2["7_290913", ] + I(gt2["7_290913", ]^2) )
lm4 <- lm( pheno$Varin_Total ~ gt2["4_68382150", ] + I(gt2["4_68382150", ]^2) + 
             gt2["4_36256098", ] + I(gt2["4_36256098", ]^2) + 
             gt2["7_290913", ] + I(gt2["7_290913", ]^2) +
             gt2["4_68382150", ]:gt2["4_36256098", ] +
             gt2["4_68382150", ]:gt2["7_290913", ]
             )
pander(summary(lm4))


#lm5 <- lm( pheno$Varin_Total ~ gt2["4_68382150", ] * gt2["7_290913", ] + I(gt2["4_68382150", ]^2) )

```


### 5 loci

```{r, results='markup'}
lm1 <- lm( pheno$Varin_Total ~ gt2["4_68382150", ] + 
             gt2["4_36256098", ] + 
             gt2["7_290913", ] +
             gt2["X_44482355", ] +
             gt2["X_12488664", ] +
             I(gt2["4_68382150", ]^2) +
             I(gt2["4_36256098", ]^2) + 
             I(gt2["7_290913", ]^2) +
             I(gt2["X_44482355", ]^2) +
             I(gt2["X_12488664", ]^2) +
             gt2["4_68382150", ]:gt2["4_36256098", ] +
             gt2["4_68382150", ]:gt2["7_290913", ] +
             gt2["4_68382150", ]:gt2["X_44482355", ] +
             gt2["4_68382150", ]:gt2["X_12488664", ] +
             gt2["4_36256098", ]:gt2["7_290913", ] +
             gt2["4_36256098", ]:gt2["X_44482355", ] +
             gt2["4_36256098", ]:gt2["X_12488664", ] +
             gt2["7_290913", ]:gt2["X_44482355", ] +
             gt2["7_290913", ]:gt2["X_12488664", ] +
             gt2["X_44482355", ]:gt2["X_12488664", ]
             )
pander(summary(lm1))


var_df <- data.frame( var = diag(vcov(lm1)))
var_df$var_p <- var_df$var/sum(var_df$var)
#var_df[sort.int(var_df$var_p, decreasing = TRUE, index.return = TRUE)$ix, ]
knitr::kable(var_df[sort.int(var_df$var, decreasing = TRUE, index.return = TRUE)$ix, ])

# library("lmerTest")
# VCrandom <- VarCorr(lm1)
# print(VCrandom, comp = "Variance")

#summary(lm1)$sigma

```




```{r}
#summary(lm( pheno$Varin_Total ~ polym( gt2["4_68382150", ], gt2["4_36256098", ], degree=2, raw=T)))
```


### 5 loci, ordered


```{r, results='markup'}
lm1 <- lm( pheno$Varin_Total ~ 
             gt2["4_68382150", ] + 
             gt2["4_36256098", ] + 
             gt2["X_12488664", ] +
             gt2["X_44482355", ] +
             gt2["7_290913", ] +
             I(gt2["4_68382150", ]^2) +
             I(gt2["X_44482355", ]^2) +
             I(gt2["4_36256098", ]^2) + 
             I(gt2["X_12488664", ]^2) +
             I(gt2["7_290913", ]^2) +
             gt2["4_68382150", ]:gt2["X_12488664", ] +
             gt2["4_36256098", ]:gt2["X_12488664", ] +
             gt2["4_68382150", ]:gt2["X_44482355", ] +
             gt2["4_68382150", ]:gt2["4_36256098", ] +
             gt2["X_44482355", ]:gt2["X_12488664", ] +
             gt2["4_68382150", ]:gt2["7_290913", ] +
             gt2["4_36256098", ]:gt2["7_290913", ] +
             gt2["4_36256098", ]:gt2["X_44482355", ] +
             gt2["7_290913", ]:gt2["X_44482355", ] +
             gt2["7_290913", ]:gt2["X_12488664", ] 
             )
pander(summary(lm1))

lm1s <- summary(lm1)

#grep("Intercept", rownames(lm1s$coefficients))

lm1_add <- lm1s$coefficients[grep("\\^2|:", rownames(lm1s$coefficients), invert = TRUE), ]
lm1_dom <- lm1s$coefficients[grep("\\^2", rownames(lm1s$coefficients), invert = FALSE), ]
lm1_epi <- lm1s$coefficients[grep(":", rownames(lm1s$coefficients), invert = FALSE), ]

lm1_add <- lm1_add[sort.int(lm1_add[, "Pr(>|t|)"], decreasing = FALSE, index.return = TRUE)$ix, ]
lm1_dom <- lm1_dom[sort.int(lm1_dom[, "Pr(>|t|)"], decreasing = FALSE, index.return = TRUE)$ix, ]
lm1_epi <- lm1_epi[sort.int(lm1_epi[, "Pr(>|t|)"], decreasing = FALSE, index.return = TRUE)$ix, ]

lm1_epi <- rbind(lm1_add, lm1_dom, lm1_epi)

library(kableExtra)
knitr::kable(lm1_epi) %>%
  kable_styling( full_width = FALSE ) %>%
  row_spec(c(7:11), background = "#C0C0C0") %>%
#  row_spec(x, c(7, 12:15), bold = TRUE, italic = TRUE)
  row_spec(c(7, 12:15), background = "#FFA07A")



#knitr::kable(lm1s$coefficients[ sort.int(lm1s$coefficients[, 'Estimate'], decreasing = F, index.return = T)$ix, ])
#knitr::kable(lm1s$coefficients[ sort.int(lm1s$coefficients[, 'Pr(>|t|)'], decreasing = F, index.return = T)$ix, ])



var_df <- data.frame( var = diag(vcov(lm1)))
var_df$var_p <- var_df$var/sum(var_df$var)
#var_df[sort.int(var_df$var_p, decreasing = TRUE, index.return = TRUE)$ix, ]
#knitr::kable(var_df[sort.int(var_df$var, decreasing = TRUE, index.return = TRUE)$ix, ])


```




## More single loci

### X_44482355


```{r}
#gt2[, 1:3]

# my_locus <- "4_68382150"
my_locus <- "X_44482355"

my_dat <- data.frame(
  Varin_Total = pheno$Varin_Total,
  gt = gt2[ my_locus, ],
#  gt = gt2["4_20005812", ],
  pcol = "#000000"
#  gt_4_20005812 = gt2["4_20005812", ]
#  gt_7_290913 = gt2["7_290913", ]
)
#table(my_dat$gt)
my_dat$pcol[ my_dat$gt == 2 ] <- "#B22222"
my_dat2 <- my_dat
#my_dat2 <- my_dat[ my_dat$gt != 2, ]
lm1 <- lm( Varin_Total ~ gt, data = my_dat2 )
summary(lm1)
```


```{r, results='markup'}
library(pander)
#pander(lm1)
pander(summary(lm1))
```


```{r, fig.width=6, fig.cap="**Figure X.** Ability of the genotype to predict the phenotype."}
library(ggplot2)
#my_dat$gt <- as.factor(my_dat$gt)
p <- ggplot( data = my_dat, mapping = aes( x = gt, y = Varin_Total, fill = gt, group = gt ))
p <- p + theme_bw()
p <- p + geom_violin()
p <- p + scale_x_continuous( 
  breaks = c(0, 1, 2),
  labels = c("0/0", "0/1", "1/1")
)

p <- p + theme(legend.position="none")
p <- p + scale_fill_continuous(type = "viridis", begin = 0.2)
p <- p + geom_jitter(shape=16, position=position_jitter(0.15), col = my_dat$pcol, size = 2)
p <- p + ylab("Varin Total (units???)")
p <- p + xlab("Genotype")
#p <- p + ggtitle("Chromosome 4, position 20,005,812 bp")
#p <- p + ggtitle("Chromosome 4, position 68,382,150 bp")
p <- p + ggtitle(my_locus)
#
p <- p + geom_abline( slope = lm1$coefficients[2], intercept = lm1$coefficients[1], linewidth = 2, col = "#808080")
p
```


### 4_36256098

```{r}
#gt2[, 1:3]

# my_locus <- "4_68382150"
# my_locus <- "X_44482355"
my_locus <- "4_36256098"

my_dat <- data.frame(
  Varin_Total = pheno$Varin_Total,
  gt = gt2[ my_locus, ],
#  gt = gt2["4_20005812", ],
  pcol = "#000000"
#  gt_4_20005812 = gt2["4_20005812", ]
#  gt_7_290913 = gt2["7_290913", ]
)
#table(my_dat$gt)
my_dat$pcol[ my_dat$gt == 2 ] <- "#B22222"
my_dat2 <- my_dat
#my_dat2 <- my_dat[ my_dat$gt != 2, ]
lm1 <- lm( Varin_Total ~ gt, data = my_dat2 )
summary(lm1)
lm2 <- lm( Varin_Total ~ gt + I(gt^2), data = my_dat )
```


```{r, results='markup'}
library(pander)
#pander(lm1)
pander(summary(lm1))
pander(summary(lm2))
```


```{r, fig.width=6, fig.cap="**Figure X.** Ability of the genotype to predict the phenotype."}
library(ggplot2)
#my_dat$gt <- as.factor(my_dat$gt)
p <- ggplot( data = my_dat, mapping = aes( x = gt, y = Varin_Total, fill = gt, group = gt ))
p <- p + theme_bw()
p <- p + geom_violin()
p <- p + scale_x_continuous( 
  breaks = c(0, 1, 2),
  labels = c("0/0", "0/1", "1/1")
)

p <- p + theme(legend.position="none")
p <- p + scale_fill_continuous(type = "viridis", begin = 0.2)
p <- p + geom_jitter(shape=16, position=position_jitter(0.15), col = my_dat$pcol, size = 2)
p <- p + ylab("Varin Total (units???)")
p <- p + xlab("Genotype")
p <- p + ggtitle(my_locus)
#
p <- p + geom_abline( slope = lm1$coefficients[2], intercept = lm1$coefficients[1], linewidth = 2, col = "#808080")
p
```


### 10_12488664

```{r}
#gt2[, 1:3]

# my_locus <- "4_68382150"
# my_locus <- "X_44482355"
# my_locus <- "4_36256098"
my_locus <- "X_12488664"


my_dat <- data.frame(
  Varin_Total = pheno$Varin_Total,
  gt = gt2[ my_locus, ],
#  gt = gt2["4_20005812", ],
  pcol = "#000000"
#  gt_4_20005812 = gt2["4_20005812", ]
#  gt_7_290913 = gt2["7_290913", ]
)
#table(my_dat$gt)
my_dat$pcol[ my_dat$gt == 2 ] <- "#B22222"
my_dat2 <- my_dat
#my_dat2 <- my_dat[ my_dat$gt != 2, ]
lm1 <- lm( Varin_Total ~ gt, data = my_dat2 )
summary(lm1)
```


```{r, results='markup'}
library(pander)
#pander(lm1)
pander(summary(lm1))
```


```{r, fig.width=6, fig.cap="**Figure X.** Ability of the genotype to predict the phenotype."}
library(ggplot2)
#my_dat$gt <- as.factor(my_dat$gt)
p <- ggplot( data = my_dat, mapping = aes( x = gt, y = Varin_Total, fill = gt, group = gt ))
p <- p + theme_bw()
p <- p + geom_violin()
p <- p + scale_x_continuous( 
  breaks = c(0, 1, 2),
  labels = c("0/0", "0/1", "1/1")
)

p <- p + theme(legend.position="none")
p <- p + scale_fill_continuous(type = "viridis", begin = 0.2)
p <- p + geom_jitter(shape=16, position=position_jitter(0.15), col = my_dat$pcol, size = 2)
p <- p + ylab("Varin Total (units???)")
p <- p + xlab("Genotype")
p <- p + ggtitle(my_locus)
#
p <- p + geom_abline( slope = lm1$coefficients[2], intercept = lm1$coefficients[1], linewidth = 2, col = "#808080")
p
```







```{r, results='asis'}
t99 <- Sys.time()
t99 - t1
```


