---
title: "SynChrom v3 chrom7"
output: 
  html_document:
    toc: true
    code_folding: hide
date: "2023-04-20"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.align = "center")
```




```{r}

#my_samp <- "DPSV"
#my_samp <- "NLv1"
#my_samp <- "SAN2"
#my_samp <- "SODL"
#
my_samp <- "WHW"


my_samp_a <- paste(my_samp, "a", sep = "")
my_samp_b <- paste(my_samp, "b", sep = "")

my_chr <- "chr7"

#my_sample <- "/media/knausb/Vining_lab/knausb/mm2_maps/salk_chr7/syri/DPSVa_chr7_DPSVb_chr7syri.out"
#
my_sample <- "/media/knausb/Vining_lab/knausb/mm2_maps/salk_chr7/syri/"

```



## Chroms (nuccomp)


```{r, results='hide'}
t1 <- Sys.time()

NUCSa <- read.csv(paste("../FigureSideo/", my_samp_a, ".softmasked_nuccomp.csv", sep = ""))
NUCSa <- NUCSa[grep("chr", NUCSa$Id), ]
NUCSa$chrom_num <- sub(".+chr", "", NUCSa$Id)
NUCSa$chrom_num[ NUCSa$chrom_num == "X" ] <- 10
NUCSa$chrom_num <- as.numeric(NUCSa$chrom_num)
NUCSa[1:3, ]
NUCSa[, 1:2]

NUCSb <- read.csv(paste("../FigureSideo/", my_samp_b, ".softmasked_nuccomp.csv", sep = ""))
NUCSb <- NUCSb[grep("chr", NUCSb$Id), ]
NUCSb$chrom_num <- sub(".+chr", "", NUCSb$Id)
NUCSb$chrom_num[ NUCSb$chrom_num == "X" ] <- 10
NUCSb$chrom_num <- as.numeric(NUCSb$chrom_num)
NUCSb[1:3, ]
NUCSb[, 1:2]
```


## Syri

```{r, results='hide'}

my_infile <- paste("/media/knausb/Vining_lab/knausb/mm2_maps/salk_chr7/syri/", 
                   my_samp_a, "_", my_chr, "_", my_samp_b, "_", my_chr, "syri.out", sep = "")
syri <- read.table( my_infile, sep = "\t", na.strings = "-")
# syri <- read.table("/media/knausb/E737-9B48/knausb/mm2_maps/EH23a_EH23b/EH23a_EH23brcsyri.out",
#                    sep = "\t", na.strings = "-")
names(syri) <- c("ref_chrom", "ref_start", "ref_end", "ref_seq", "query_seq", "query_chrom", "query_start", "query_end", "unique_ID", "parent_ID", "Annotation_Type", "Copy_Status")
#syri$ref_start[is.na(as.numeric(syri$ref_start))]
syri$ref_start <- as.numeric(syri$ref_start)
syri$ref_end <- as.numeric(syri$ref_end)
syri$query_start <- as.numeric(syri$query_start)
syri$query_end <- as.numeric(syri$query_end)

syri[1:3, ]
sort(table(syri$Annotation_Type), decreasing = TRUE)

syn <- syri[syri[, 11] == "SYN", ]
syn[1:3, ]
nrow(syn)
last_win <- 1
#
coalesce <- 1e0
#coalesce <- 5e4
#coalesce <- 1e5

#coalesce <- 4e5

#coalesce <- 1e6
for( j in 2:nrow(syn) ){
  if( syn$ref_start[j] - syn$ref_end[last_win] < coalesce &
      syn$query_start[j] - syn$query_end[last_win] < coalesce &
      syn$ref_chrom[j] == syn$ref_chrom[last_win] ){
        syn$ref_end[ last_win ] <- syn$ref_end[j]
        syn$query_end[ last_win ] <- syn$query_end[j]
        #syn$ref_end[j] <- NA
        syn$unique_ID[j] <- NA
  } else {
    last_win <- j
  }
}

#syn <- syn[ !is.na( syn$ref_end ), ]
syn <- syn[ !is.na( syn$unique_ID ), ]
syn[1:3, ]
nrow(syn)
min(syn$ref_end - syn$ref_start)
#hist(syn$ref_end - syn$ref_start, main = "Length of Syntenous regions (bp)")
sort(syn$ref_end - syn$ref_start, decreasing = T)[1:10]

syn[1:3, ]

inv <- syri[syri$Annotation_Type == "INV", ]

dup <- syri[syri$Annotation_Type == "DUP", ]
dup[1:3, ]
nrow(dup)
trans <- syri[syri$Annotation_Type == "TRANS", ]
trans[1:3, ]
nrow(trans)
```



### gribbon


```{r}
# nrow(syn)
# x <- syn[1:1000, ]
# nsmooth <- 10
# min_len <- 1e5

#gribbon <- function(x, nsmooth = 50, min_len = 1e5){
gribbon <- function(
    x, 
    nsmooth = 10, 
    min_len = 1e5, 
    coalesce = 0, 
    invert = FALSE){
  
  x <- x[(x$ref_end - x$ref_start) >= min_len, ]
  x <- x[(x$query_end - x$query_start) >= min_len, ]
  if( nrow(x) <= 0 ){
    return("No rows")
  }
  
  # if( coalesce > 0 ){
  #   last_win <- 1
  #   for( j in 2:nrow(x) ){
  #     if( x$ref_start[j] - x$ref_end[last_win] < coalesce &
  #         x$query_start[j] - x$query_end[last_win] < coalesce ){
  #       x$ref_end[ last_win ] <- x$ref_end[j]
  #       x$query_end[ last_win ] <- x$query_end[j]
  #       x$ref_end[j] <- NA
  #     } else {
  #       last_win <- j
  #     }
  #   }
  #   x <- x[ !is.na( x$ref_end ), ]
  # }
  
  if( invert ){
    tmp <- x$query_start
    x$query_start <- x$query_end
    x$query_end <- tmp
  }
  
  # Set a chromosome number.
  x$ref_chrom_num <- sub(".+chr", "", x$ref_chrom)
  x$ref_chrom_num[ x$ref_chrom_num == "X" ] <- 10
  x$ref_chrom_num[ x$ref_chrom_num == "Y" ] <- 11
  x$ref_chrom_num <- as.numeric(x$ref_chrom_num)
  x$ref_chrom_num <- x$ref_chrom_num - 0.2
  x$query_chrom_num <- sub(".+chr", "", x$query_chrom)
  x$query_chrom_num[ x$query_chrom_num == "X" ] <- 10
  x$query_chrom_num[ x$query_chrom_num == "Y" ] <- 11
  x$query_chrom_num <- as.numeric(x$query_chrom_num)
  x$query_chrom_num <- x$query_chrom_num + 0.2
  
#  nrow(x)
  my_x <- seq( from = -4, to = 4, length.out = nsmooth)
  my_y <- 1/( 1 + exp(1)^-my_x)
  my_x <- (my_x + 4)/8
  # plot(my_x, my_y)
  
#  my_x  <- c(my_x, rev(my_x))
  my_polys <- data.frame( 
    ID = rep(x$unique_ID, each = nsmooth * 2),
    ref_chrom_num = rep(x$ref_chrom_num, each = nsmooth * 2),
#    ref_start = rep(NA, each = nsmooth * 2),
#    ref_end = rep(NA, each = nsmooth * 2),
    query_chrom_num = rep(x$query_chrom_num, each = nsmooth * 2),
#    query_start = rep(NA, each = nsmooth * 2),
#    query_end = rep(NA, each = nsmooth * 2)
     poly_xs = rep(NA, each = nsmooth * 2),
     poly_ys = rep(NA, each = nsmooth * 2)
  )
  my_polys[1:3, ]
  
  my_vs <- seq(1, nrow(my_polys) + 1, by = nsmooth * 2)
#  for( i in seq(1, nrow(my_polys), by = nsmooth * 2) ){
  for( i in 1:nrow(x) ){
    my_index <- my_vs[i]:(my_vs[i+1] - 1)
    xmins <- seq( x$ref_chrom_num[i], x$query_chrom_num[i], length.out = nsmooth)
    
#    my_y * (x$query_end[i] - x$query_start[i]) + x$query_start[i]
    
    ymins <- my_y * (x$query_start[i] - x$ref_start[i]) + x$ref_start[i]
    ymaxs <- my_y * (x$query_end[i] - x$ref_end[i]) + x$ref_end[i]
    
    my_polys$poly_xs[my_index]    <- c(xmins, rev(xmins))
    my_polys$poly_ys[my_index] <- c(ymins, rev(ymaxs))
    #x$ref_start[ my_index ] <- seq(x$ref_chrom[i], x$query_chrom[i], length.out = nsmooth)
    
  }
#  my_polys[1:3, ]
  return( my_polys )  
}

```



```{r}
#poly_coords <- gribbon(syn[1:1000, ])
#poly_coords <- gribbon(syn, min_len = 1e5)
#
nrow(syn)


poly_coords <- gribbon(syn, min_len = 1e1, coalesce = 0)

#poly_coords <- gribbon(syn, min_len = 1e3)
# poly_coords <- gribbon(syn, nsmooth = 40, min_len = 1e3)
# poly_coords <- gribbon(syn, min_len = 1e1)
length(unique(poly_coords$ID))

poly_coords_inv <- gribbon(inv, min_len = 1e4, invert = TRUE)
poly_coords_dup <- gribbon(dup, min_len = 1e4)
poly_coords_trans <- gribbon(trans, min_len = 1e4)

```


## Call motif counter

```{r, eval=FALSE}
#cmd <- "~/gits/hempy/bin/fast_win.py /media/knausb/Vining_lab/knausb/mm2_maps/EH23a_EH23b_v2/EH23b.rc458X.softmasked.fasta.gz --win_size 1000000"
#system(cmd)

```


## Motif windows and gene gff


```{r}
# sampn <- "EH23b"
# sampn <- "DPSVa"
# chrom_lengths <- NUCSa[ , c("Id", "Length")]

lu <- read.table("/media/knausb/E737-9B48/releases/scaffolded_csat_orientations.tsv", 
                 sep = "\t", header = TRUE)
#lu[1:3, ]

get_wins <- function( sampn, lookup_table, chrom_lengths ){

  wins_dir <- "../FigureSideo/"
  gffs_dir <- "/media/knausb/E737-9B48/releases/scaffolded/"
  my_lu <- lookup_table[ lu$Sample == sampn, ]

  # Input windows file.
  wins <- read.csv( paste( wins_dir, sampn, ".softmasked_wins.csv", sep = "") )
  wins <- wins[grep("chr", wins$Id), ]
  wins$chrn <- sub(".+chr", "", wins$Id)
  wins$chrn[ wins$chrn == "X" ] <- 10
  wins$chrn[ wins$chrn == "Y" ] <- 11
  wins$chrn <- as.numeric(wins$chrn)
  #wins[1:3, ]
  
  # Input GFF with genes.
  genes <- read.table( 
    paste(gffs_dir, sampn, "/", sampn, ".primary_high_confidence.gff3.gz", sep = "" ), 
    sep = "\t" )
  genes <- genes[genes[, 3] == "gene", ]
  genes <- genes[ grep("chr", genes$V1), ]
  
  # Orient windows and GFF data.
  for( i in 1:nrow( my_lu ) ){
    if( my_lu$Flip[i] == "True" ){
      flip_chrom <- paste( my_lu$Sample[i], my_lu$Chromosome[i], sep = "." )
      
      
      # Windows
      tmp <- wins$Start[wins$Id == flip_chrom]
      wins$Start[wins$Id == flip_chrom] <- rev(tmp)
      tmp <- wins$End[wins$Id == flip_chrom]
      wins$End[wins$Id == flip_chrom] <- rev(tmp)
      
      # GFF
      genes[1:3, ]
      # Start must be lower than stop.
      new_stop <- abs(genes[ genes[, 1] == flip_chrom, 4] - chrom_lengths$Length[ chrom_lengths$Id == flip_chrom ])
      new_start  <- abs(genes[ genes[, 1] == flip_chrom, 5] - chrom_lengths$Length[ chrom_lengths$Id == flip_chrom ])
      genes[ genes[, 1] == flip_chrom, 4] <- new_start
      genes[ genes[, 1] == flip_chrom, 5] <- new_stop

    }
  }
  

  # Windowize and scale CG by genome.
  #  wins$CGs <- 0
  wins$CGs <- wins$CG / wins$Win_length * 100
  wins$CGs <- wins$CGs - min(wins$CGs, na.rm = TRUE)
  wins$CGs <- wins$CGs / max(wins$CGs, na.rm = TRUE)
  # for( j in unique(wins$Id) ){
  #   wins$CGs[ wins$Id == j] <- wins$CGs[ wins$Id == j] - min(wins$CGs[ wins$Id == j], na.rm = TRUE)
  #   wins$CGs[ wins$Id == j] <- wins$CGs[ wins$Id == j] / max(wins$CGs[ wins$Id == j], na.rm = TRUE)
  # }
  wins$iCGs <- 1 - wins$CGs
  

  # Windowize genes from GFF.
  wins$gcnt <- 0
  for(i in 1:nrow(wins)){
    # if( sampn == "EH23b" ){
    #   wins$Id <- sub("EH23a", "EH23b", wins$Id)
    # }
    tmp <- genes[genes$V1 == wins$Id[i] & genes$V4 >= wins$Start[i] & genes$V5 < wins$End[i], ]
    wins$gcnt[i] <- nrow(tmp)
  }
  
  # Scaling of gene count windows is on a per assembly basis.
  wins$gcntsc <- wins$gcnt - min(wins$gcnt)
  wins$gcntsc <- wins$gcntsc / max(wins$gcntsc)
  my_index <- round( wins$gcntsc * 100 )
  my_index[ my_index <= 0] <- 1
  
  # Color ramp.
  #wins$gcol <- heat.colors(n=100)[ my_index ]
  #wins$gcol <- colorRampPalette(c("yellow", "orange", "red"))( 100 )[ my_index ]
  #wins$gcol <- colorRampPalette(c("red", "orange", "yellow"))( 100 )[ my_index ]
  #wins$gcol <- colorRampPalette(c("#0000FF", "#228B22", "#A0522D"))( 100 )[ my_index ]
  #wins$gcol <- colorRampPalette(c("#87CEEB", "#3CB371", "#228B22", "#A0522D"))( 100 )[ my_index ]
  
  #wins$gcol <- viridisLite::plasma(n = 100, alpha = 1, begin = 0.1, end = 1)[ my_index ]
  wins$gcol <- viridisLite::magma(n = 100, alpha = 1, begin = 0.2, end = 1.00)[ my_index ]
  
  return(wins)
}

```


```{r, results='hide'}
chrom_lengths <- NUCSa[ , c("Id", "Length")]

# my_samp_a_wins <- get_wins(my_samp_a, chrom_lengths)
my_samp_a_wins <- get_wins( sampn = my_samp_a, lookup_table = lu, chrom_lengths = chrom_lengths )
#my_samp_a_wins[1:3, ]

chrom_lengths <- NUCSb[ , c("Id", "Length")]
my_samp_b_wins <- get_wins( sampn = my_samp_b, lookup_table = lu, chrom_lengths = chrom_lengths )
my_samp_b_wins[1:3, ]

```



Windowize SNPs


```{r win_snps}
# EH23a_wins[1:3, ]
# 
# snps <- syri[syri$Annotation_Type == "SNP", ]
# nrow(snps)
# snps[1:3, ]
# 
# EH23a_wins$snps <- 0
# 
# for(i in 1:nrow(EH23a_wins)){
#   tmp <- snps[ snps$ref_chrom == EH23a_wins$Id[i] & snps$ref_start >= EH23a_wins$Start[i] & snps$ref_end <= EH23a_wins$End[i], ]
#   EH23a_wins$snps[i] <- nrow(tmp)
# }
# EH23a_wins$snpssc <- (EH23a_wins$snps - min(EH23a_wins$snps))/max(EH23a_wins$snps)
# 
# hist(EH23a_wins$snps)
# hist(EH23a_wins$snps[ EH23a_wins$Id == "EH23a.chr4" ])
# my_hist <- hist(EH23a_wins$snpssc * 9 + 1, plot = F)
# 
# #barplot(my_hist$density, space = 0, col = gray.colors(n=10))
# #axis(side=1)
# barplot(my_hist$density ~ my_hist$breaks[-10], space = 0, col = gray.colors(n=10))
# 
# 

```




## BLAST


```{r}
#sampn <- "EH23a"
sampn <- my_samp_a
ablstn <- read.csv(
    paste("/media/knausb/Vining_lab/knausb/blast_projects/todd_rrna/blastn_uc/", sampn, "_blastn.csv", sep = ""),
    header = FALSE)
colnames(ablstn) <- c("qseqid","qlen","sseqid","slen","qstart","qend",
                     "sstart","send","evalue","bitscore","score","length",
                     "pident","nident","mismatch","positive","gapopen",
                     "gaps","ppos","sstrand")
ablstn <- ablstn[grep("chr", ablstn$sseqid), ]
ablstn$chrn <- sub(".*chr", "", ablstn$sseqid)
ablstn$chrn[ablstn$chrn == "X"] <- 10
ablstn$chrn[ablstn$chrn == "Y"] <- 11
ablstn$chrn <- as.numeric(ablstn$chrn)


#sampn <- "EH23b"
sampn <- my_samp_b
bblstn <- read.csv(
    paste("/media/knausb/Vining_lab/knausb/blast_projects/todd_rrna/blastn_uc/", sampn, "_blastn.csv", sep = ""),
    header = FALSE)
colnames(bblstn) <- c("qseqid","qlen","sseqid","slen","qstart","qend",
                     "sstart","send","evalue","bitscore","score","length",
                     "pident","nident","mismatch","positive","gapopen",
                     "gaps","ppos","sstrand")
bblstn <- bblstn[grep("chr", bblstn$sseqid), ]
bblstn$chrn <- sub(".*chr", "", bblstn$sseqid)
bblstn$chrn[bblstn$chrn == "X"] <- 10
bblstn$chrn[bblstn$chrn == "Y"] <- 11
bblstn$chrn <- as.numeric(bblstn$chrn)
```


Orient BLAST


```{r}
#hist(ablstn$sstart[ablstn$sseqid == "DPSVa.chr7"]/1e6)
```


```{r,  eval=TRUE}
# Orient

# Orient chromosomes.
my_lu <- lu[ lu$Sample == my_samp_a, ]
#chrom_lengths <- NUCSa[ , c("Id", "Length")]

for(i in 1:nrow(my_lu)){
  if( my_lu$Flip[i] == "True" ){
    my_chrom <- paste(my_lu$Sample[i], my_lu$Chromosome[i], sep = ".")
    # hist( (ablstn$sstart[ ablstn$sseqid == my_chrom ])/1e6)
    # hist( (NUCSa$Length[ NUCSa$Id == my_chrom ] - ablstn$sstart[ ablstn$sseqid == my_chrom ])/1e6)
    
    new_start <- NUCSa$Length[ NUCSa$Id == my_chrom ] - ablstn$sstart[ ablstn$sseqid == my_chrom ]
    new_stop  <- NUCSa$Length[ NUCSa$Id == my_chrom ] - ablstn$send[ ablstn$sseqid == my_chrom ]
    ablstn$sstart[ ablstn$sseqid == my_chrom ] <- new_start
    ablstn$send[ ablstn$sseqid == my_chrom ] <- new_stop
  }
}

i <- 7
my_chrom <- paste(my_lu$Sample[i], my_lu$Chromosome[i], sep = ".")
#hist( (ablstn$sstart[ ablstn$sseqid == my_chrom ])/1e6)

my_lu <- lu[ lu$Sample == my_samp_b, ]
#chrom_lengths <- NUCSb[ , c("Id", "Length")]

for(i in 1:nrow(my_lu)){
  if( my_lu$Flip[i] == "True" ){
    my_chrom <- paste(my_lu$Sample[i], my_lu$Chromosome[i], sep = ".")
    new_start <- NUCSb$Length[ NUCSb$Id == my_chrom ] - bblstn$sstart[ bblstn$sseqid == my_chrom ]
    new_stop  <- NUCSb$Length[ NUCSb$Id == my_chrom ] - bblstn$send[ bblstn$sseqid == my_chrom ]
    bblstn$sstart[ bblstn$sseqid == my_chrom ] <- new_start
    bblstn$send[ bblstn$sseqid == my_chrom ] <- new_stop
  }
}


```


```{r}
#hist(ablstn$sstart[ablstn$sseqid == "DPSVa.chr7"]/1e6)
```


## Plot


```{r}
library(ggplot2)

chrom_wid <- 0.05
p <- ggplot2::ggplot()
p <- p + theme_bw()
p <- p + ggplot2::geom_rect( 
  data = NUCSa, 
  ggplot2::aes( xmin = chrom_num - chrom_wid - 0.2,
                xmax = chrom_num + chrom_wid - 0.2,
                ymin = 1, ymax = Length),
  fill = "#DCDCDC",
  color = "#000000"
)

p <- p + ggplot2::geom_rect( 
  data = NUCSb, 
  ggplot2::aes( xmin = chrom_num - chrom_wid + 0.2,
                xmax = chrom_num + chrom_wid + 0.2,
                ymin = 1, ymax = Length),
  fill = "#DCDCDC",
  color = "#000000"
)


#my_cols <- gray.colors(n=10, start = 0.3, end = 0.9)[round(EH23a_wins$snpssc * 9 + 1)]
#my_cols <- gray.colors(n=100, start = 0.3, end = 0.9)[round(EH23a_wins$snpssc * 99 + 1)]
#p <- 
# p + ggplot2::geom_rect( 
#   data = EH23a_wins, 
#   ggplot2::aes( xmin = chrn - 0.5,
#                 xmax = chrn - 0.25,
#                 ymin = Start, ymax = End),
# #  fill = "#DCDCDC",
#   fill = my_cols,
# #  color = "#000000"
# )

# Theme
p <- p + ggplot2::scale_x_continuous( 
  breaks = NUCSa$chrom_num,
#  limits = c(0.6, 10.4),
#  limits = c(0.5, 10.5),
#  labels = EH23a$Id
  labels = sub("a.chr", ".chr", NUCSa$Id)
)
  
p <- p + scale_y_continuous(
  breaks = seq( 0, 120e6, by = 10e6), 
  labels = seq( 0, 120, by = 10)
)
  
#  p <- p + ggplot2::theme_bw() + 
p <- p + ggplot2::theme( 
      #      panel.grid.minor.x = ggplot2::element_blank(), 
    panel.grid.minor.x = ggplot2::element_line( linewidth = 0.4, color = "#C0C0C0", linetype = 3 ),
    axis.text.x = element_text(angle = 60, hjust = 1),
    axis.title.x=element_blank(),
    panel.grid.major.y = ggplot2::element_line( linewidth = 0.4, color = "#C0C0C0", linetype = 1 ),
    panel.grid.minor.y = ggplot2::element_line( linewidth = 0.4, color = "#C0C0C0", linetype = 3 )
  )
  
p <- p + ggplot2::ylab("Position (Mbp)")
#p <- p + ggtitle( "EH23" )
p <- p + ggtitle(sub("a$", "", my_samp_a))
p



#p <- p + theme(legend.position='none')
p <- p + geom_polygon( 
  data = poly_coords, 
  aes(x = poly_xs, y = poly_ys, fill = ID, group = ID), #, fill = col),
  alpha = 2/5,
  show.legend = FALSE,
  #color = NA,
#  color = "#80808022",
#  color = "#000000",
  color = "#696969",
#  color = "#808080",
#  color = "#A9A9A9",
#  color = "#C0C0C0",
  linewidth = 0.4,
#            fill = ID
#  fill = "#C0C0C044"
  fill = "#808080"
  )

# p

if( nrow(poly_coords_inv) > 0 && !is.null(nrow(poly_coords_inv)) ){
#if( poly_coords_inv != "No rows" ){
  p <- p + geom_polygon( 
    data = poly_coords_inv, 
    aes(x = poly_xs, y = poly_ys, fill = ID, group = ID),
    alpha = 0.9,
    show.legend = FALSE,
    fill = "#FFA500"
#    fill = "#808080"
    )
}
p

# p + geom_polygon( 
#   data = poly_coords_dup, 
#   aes(x = poly_xs, y = poly_ys, fill = ID, group = ID),
#   alpha = 0.9,
#   show.legend = FALSE,
#   fill = "#1E90FF"
#   )

# p + geom_polygon( 
#   data = poly_coords_trans, 
#   aes(x = poly_xs, y = poly_ys, fill = ID, group = ID),
#   alpha = 0.9,
#   show.legend = FALSE,
#   fill = "#1E90FF"
#   )


thinw <- 0.28
p <- p + ggplot2::geom_rect( 
  data = my_samp_a_wins, 
  ggplot2::aes(
    xmin = chrn - iCGs * thinw - 0.2,
    #xmax = chrn + iCGs * thinw,
    xmax = chrn - 0.2,
    ymin = Start, 
    ymax = End),
  fill = my_samp_a_wins$gcol,
  color = NA
)

p <- p + ggplot2::geom_rect( 
  data = my_samp_b_wins, 
  ggplot2::aes(
    #xmin = chrn - iCGs * thinw - 0.2,
    xmin = chrn + 0.2,
    xmax = chrn + iCGs * thinw + 0.2,
    #xmax = chrn - 0.2,
    ymin = Start, 
    ymax = End),
  fill = my_samp_b_wins$gcol,
  color = NA
)
```


```{r, fig.width=10}
p


cat("Maximum genes per window:", as.character(max(c(my_samp_a_wins$gcnt, my_samp_b_wins$gcnt))))

cat("Maximum genes per window, hap A:", max(c(my_samp_a_wins$gcnt)))
cat("Maximum genes per window, hap B:", max(c(my_samp_b_wins$gcnt)))

# hist(my_samp_b_wins$gcnt)
# my_samp_b_wins[ my_samp_b_wins$gcnt > 200, ]
# plot(my_samp_b_wins$gcnt, my_samp_b_wins$CGs)


#p + xlim(3.5, 4.5) #+ ylim(1e6, 2e6)

# p + xlim(5.5, 8.5) + ylim(10e6, 40e6)

#pEH23 <- p
#save(pEH23, file = "ideo.RData")
#load("ideo.RData")
#pEH23
```


```{r}
#hist(ablstn$sstart[ablstn$sseqid == "DPSVa.chr7"]/1e6)
```


```{r}
# Cannabinoids
my_rows <- grep("AB2|LY|NC_044376.1:c427", ablstn$qseqid)
#hist(ablstn[my_rows, ]$sstart/1e6)

blstw <- 0.48
p <- p + ggplot2::geom_rect( 
    data = ablstn[my_rows, ], 
    ggplot2::aes( 
      xmin = chrn - blstw,
      xmax = chrn + 0 - 0.14,
      ymin = sstart, 
      ymax = send),
    fill = "#228B22",
    color = '#228B22'
  )
my_rows <- grep("AB2|LY|NC_044376.1:c427", bblstn$qseqid)
blstw <- 0.48
p <- p + ggplot2::geom_rect( 
    data = bblstn[my_rows, ], 
    ggplot2::aes( 
      xmin = chrn + 0.14,
      xmax = chrn + blstw,
      ymin = sstart, 
      ymax = send),
    fill = "#228B22",
    color = '#228B22'
  )

# 5S
my_rows <- grep("5S_", ablstn$qseqid)
blstw <- 0.48
p <- p + ggplot2::geom_rect( 
  data = ablstn[my_rows, ], 
  ggplot2::aes( 
    xmin = chrn - blstw,
    xmax = chrn - 0.1,
    ymin = sstart, 
    ymax = send),
  fill = "#000000",
  color = '#000000'
)
my_rows <- grep("5S_", bblstn$qseqid)
blstw <- 0.48
p <- p + ggplot2::geom_rect( 
  data = bblstn[my_rows, ], 
  ggplot2::aes( 
    xmin = chrn + 0.1,
    xmax = chrn + blstw,
    ymin = sstart, 
    ymax = send),
  fill = "#000000",
  color = '#000000'
)

# 45S
my_rows <- grep("45s_|26s_|5.8s_|18s_", ablstn$qseqid)
blstw <- 0.48
# p <- p + ggplot2::geom_rect( 
#   data = ablstn[my_rows, ], 
#   ggplot2::aes( 
#     xmin = chrn - blstw,
#     xmax = chrn - 0.1,
#     ymin = sstart, 
#     ymax = send),
#   fill = "#B22222",
#   linewidth = 2,
#   #    linewidth = 4,
#   color = '#B22222'
# )
my_rows <- grep("45s_|26s_|5.8s_|18s_", bblstn$qseqid)
blstw <- 0.48
# p <- p + ggplot2::geom_rect( 
#   data = bblstn[my_rows, ], 
#   ggplot2::aes( 
#     xmin = chrn + 0.1,
#     xmax = chrn + blstw,
#     ymin = sstart, 
#     ymax = send),
#   fill = "#B22222",
#   linewidth = 2,
#   #    linewidth = 4,
#   color = '#B22222'
# )
```


```{r, eval=FALSE}
# Centromeres, subtelomeric repeats
my_rows <- grep("AH3Ma_CS-237_satelite|CsatSD_centromere_237bp|CsatSD_centromere_370bp", ablstn$qseqid)
#  blstw <- 0.48
p <- p + ggplot2::geom_rect( 
  data = ablstn[my_rows, ], 
  ggplot2::aes( 
    xmin = chrn - 0.5,
    xmax = chrn - 0.2,
    ymin = sstart, 
    ymax = send),
  fill = "#1E90FF",
  color = '#1E90FF'
)
my_rows <- grep("AH3Ma_CS-237_satelite|CsatSD_centromere_237bp|CsatSD_centromere_370bp", bblstn$qseqid)
p <- p + ggplot2::geom_rect( 
  data = bblstn[my_rows, ], 
  ggplot2::aes( 
    xmin = chrn + 0.2,
    xmax = chrn + 0.5,
    ymin = sstart, 
    ymax = send),
  fill = "#1E90FF",
  color = '#1E90FF'
)
```


```{r, fig.width=10}
p


# ggsave(filename = paste(my_samp, "_ideo.png", sep = ""), device = "png", 
#        width = 6.5, height = 3.25, units = "in", dpi = 300)


#pEH23 <- p
#pEH23

#save(pEH23, file = "EH23ideo.RData")
#saveRDS(pEH23, file = "EH23ideoRDS.RData")
```


```{r, fig.width=10}
# p
# p + xlim(5.5, 8.5) + ylim(10e6, 40e6)
p + xlim(5.0, 9.0) + ylim(10e6, 40e6)

# ggsave(filename = paste(my_samp, "_5-9MB_ideo.png", sep = ""), device = "png",
#        width = 6.5, height = 3.25, units = "in", dpi = 300)

```



```{r}
t99 <- Sys.time()
t99 - t1
```


