---
title: "varinEpistasis"
author: "Brian J. Knaus"
date: "`r format(Sys.time(), '%Y, %B %d')`"
output: 
  html_document:
    toc: true
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
#knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(fig.align = "center")
knitr::opts_chunk$set(fig.width = 12)
# knitr::opts_chunk$set(results = "hide")
knitr::opts_chunk$set(results = "markup")
```


## Data


```{r}
library(vcfR)
```


```{r, results='hide'}
vcf <- read.vcfR("varin_Blink.vcf.gz")
```


```{r}
vcf
gt <- extract.gt(vcf)
#is_het(gt)
gt2 <- matrix( nrow = nrow(gt), ncol = ncol(gt), dimnames = dimnames(gt) )

gt2[ gt == "0/0" ] <- 0
gt2[ gt == "0/1" ] <- 1
gt2[ gt == "1/1" ] <- 2
#gt <- apply(gt, MARGIN = 2, as.numeric)

```



```{r}
pheno <- read.csv('ERBxHO40_Phenotypes.csv')
pheno[1:3,]
```


### Synchronize


```{r, results='asis'}
pheno <- pheno[colnames(gt), ]
all(pheno$ID == colnames(gt))
```


### Phenotype

```{r, fig.width=6}
hist(pheno$Varin_Total, breaks = seq(0, 65, by = 2))
```


```{r, fig.width=6}
qqnorm(pheno$Varin_Total)
qqline(pheno$Varin_Total, col = 2, lwd = 4 )
```


```{r, fig.width=6, eval=FALSE}
# pheno$Varin_Total2 <- pheno$Varin_Total - lm11$fitted.values
# pheno$Varin_Total2 <- lm11$residuals
# pheno$Varin_Total <- lm11$residuals
hist(pheno$Varin_Total2)
qqnorm(pheno$Varin_Total2)
qqline(pheno$Varin_Total2, col = 2, lwd = 4 )
```


## Single, independent loci


### locus 4_68382150


```{r}
my_locus <- "4_68382150"
# my_locus <- "4_36256098"
# my_locus <- "7_290913"
# my_locus <- "X_12488664"
# my_locus <- "X_44482355"
chr_pos <- paste(getCHROM(vcf), getPOS(vcf), sep = "_")
#my_alleles <- paste(getREF(vcf)[my_locus == chr_pos], getALT(vcf)[my_locus == chr_pos], sep = "/")
my_alleles <- paste( "0 = ", getREF(vcf)[my_locus == chr_pos], "; 1 = ", getALT(vcf)[my_locus == chr_pos], sep = "")

my_dat <- data.frame(
  Varin_Total = pheno$Varin_Total,
  gt = gt2[my_locus, ],
  pcol = "#000000"
)
my_dat$pcol[ my_dat$gt == 2 ] <- "#B22222"

lm1 <- lm( Varin_Total ~ gt, data = my_dat )
lm11 <- lm( Varin_Total ~ gt + I(gt^2), data = my_dat )

my_vals <- data.frame(
  x = seq(0, 2, length.out=10)
)
my_vals$y <- lm11$coefficients[1] + my_vals$x * lm11$coefficients[2] + my_vals$x^2 * lm11$coefficients[3]
```


```{r, results='markup'}
library(pander)
#pander(lm1)
pander(summary(lm1))
pander(summary(lm11))
#hist(lm11$fitted.values)
```


```{r, results='markup'}
summary(lm1)
```


```{r, fig.width=6, fig.cap="**Figure X.** Ability of the genotype to predict the phenotype."}
library(ggplot2)
#my_dat$gt <- as.factor(my_dat$gt)
p <- ggplot( data = my_dat, mapping = aes( x = gt, y = Varin_Total, fill = gt, group = gt ))
p <- p + theme_bw()
p <- p + geom_violin()
p <- p + scale_x_continuous( 
  breaks = c(0, 1, 2),
  labels = c("0/0", "0/1", "1/1")
)
p <- p + theme(legend.position="none")
p <- p + scale_fill_continuous(type = "viridis", begin = 0.2)
p <- p + geom_jitter(shape=16, position=position_jitter(0.15), col = my_dat$pcol, size = 2)
p <- p + ylab("Varin Total (units???)")
p <- p + xlab("Genotype")
p <- p + ggtitle( paste(my_locus, my_alleles, sep = ": ") )
#
p <- p + geom_abline( slope = lm1$coefficients[2], intercept = lm1$coefficients[1], linewidth = 2, col = "#808080")
p <- p + geom_line( data = my_vals, mapping = aes( x = x, y = y ), inherit.aes = F, linewidth = 2, col = "#808080" )
p
```



### locus 7_290913


```{r}
# my_locus <- "4_68382150"
# my_locus <- "4_36256098"
# 
my_locus <- "7_290913"
# my_locus <- "X_12488664"
# my_locus <- "X_44482355"
chr_pos <- paste(getCHROM(vcf), getPOS(vcf), sep = "_")
#my_alleles <- paste(getREF(vcf)[my_locus == chr_pos], getALT(vcf)[my_locus == chr_pos], sep = "/")
my_alleles <- paste( "0 = ", getREF(vcf)[my_locus == chr_pos], "; 1 = ", getALT(vcf)[my_locus == chr_pos], sep = "")

my_dat <- data.frame(
  Varin_Total = pheno$Varin_Total,
  gt = gt2[my_locus, ],
  pcol = "#000000"
)
my_dat$pcol[ my_dat$gt == 2 ] <- "#B22222"

lm1 <- lm( Varin_Total ~ gt, data = my_dat )
lm11 <- lm( Varin_Total ~ gt + I(gt^2), data = my_dat )

my_vals <- data.frame(
  x = seq(0, 2, length.out=10)
)
my_vals$y <- lm11$coefficients[1] + my_vals$x * lm11$coefficients[2] + my_vals$x^2 * lm11$coefficients[3]
```



```{r, results='markup'}
library(pander)
#pander(lm1)
pander(summary(lm1))
pander(summary(lm11))
```


```{r, fig.width=6, fig.cap="**Figure X.** Ability of the genotype to predict the phenotype."}
library(ggplot2)
#my_dat$gt <- as.factor(my_dat$gt)
p <- ggplot( data = my_dat, mapping = aes( x = gt, y = Varin_Total, fill = gt, group = gt ))
p <- p + theme_bw()
p <- p + geom_violin()
p <- p + scale_x_continuous( 
  breaks = c(0, 1, 2),
  labels = c("0/0", "0/1", "1/1")
)
p <- p + theme(legend.position="none")
p <- p + scale_fill_continuous(type = "viridis", begin = 0.2)
p <- p + geom_jitter(shape=16, position=position_jitter(0.15), col = my_dat$pcol, size = 2)
p <- p + ylab("Varin Total (units???)")
p <- p + xlab("Genotype")
p <- p + ggtitle( paste(my_locus, my_alleles, sep = ": ") )
#
p <- p + geom_abline( slope = lm1$coefficients[2], intercept = lm1$coefficients[1], linewidth = 2, col = "#808080")
p <- p + geom_line( data = my_vals, mapping = aes( x = x, y = y ), inherit.aes = F, linewidth = 2, col = "#808080" )
p
```


## Two loci, additive effects


```{r, fig.width=6, fig.cap="**Figure X.** Ability of the genotype to predict the phenotype. Points colored according to locus 4_68382150."}
library(ggplot2)

my_dat$chr7col <- gt2["4_68382150", ]
# my_dat$chr7col[ my_dat$chr7col == 0 ] <- "#1b9e77"
# my_dat$chr7col[ my_dat$chr7col == 0 ] <- "#e7298a"
# my_dat$chr7col[ my_dat$chr7col == 1 ] <- "#d95f02"
# my_dat$chr7col[ my_dat$chr7col == 2 ] <- "#7570b3"
# my_dat$chr7col[ my_dat$chr7col == 0 ] <- "#7fc97f"
# my_dat$chr7col[ my_dat$chr7col == 1 ] <- "#beaed4"
# my_dat$chr7col[ my_dat$chr7col == 2 ] <- "#fdc086"

#viridisLite::magma(n=3, begin = 0.1, end = 0.9)

my_dat$chr7col[ my_dat$chr7col == 0 ] <- "#150E37FF"
my_dat$chr7col[ my_dat$chr7col == 1 ] <- "#B63679FF"
my_dat$chr7col[ my_dat$chr7col == 2 ] <- "#FECE91FF"

#my_dat$gt2 <- as.factor(my_dat$gt)
p <- ggplot( data = my_dat, mapping = aes( x = gt, y = Varin_Total, 
                                           group = gt, fill = gt), 
                                           #fill = "#C0C0C0")#,
                                           #group = gt )
                                           )
p <- p + geom_violin()
#p <- p + scale_fill_manual(values=c("#999999", "#E69F00", "#56B4E9"))
#p <- p + scale_fill_manual(values= viridisLite::viridis(n=3, alpha = 0.2) )
p <- p + scale_fill_viridis_c( alpha = 0.2)

# p <- ggplot( data = my_dat, mapping = aes( x = gt, y = Varin_Total), 
#                                            #fill = gt, 
#  #                                          fill = "#C0C0C0"#,
#                                            #group = gt )
#                                            )

p <- p + theme_bw()
p <- p + theme(legend.position="none")

#p <- p + geom_violin(fill='#A4A4A4')
#p <- p + geom_violin(fill='#C0C0C0')

p <- p + scale_x_continuous(
  breaks = c(0, 1, 2),
  labels = c("0/0", "0/1", "1/1")
)

#p <- p + scale_fill_continuous(type = "viridis", begin = 0.2)
#p <- p + geom_jitter(shape=16, position=position_jitter(0.15), col = my_dat$pcol, size = 2)
p <- p + geom_jitter(position=position_jitter(0.15), shape=21, col = "#000000",
                     bg = my_dat$chr7col, size = 2.8)

p <- p + ylab("Varin Total (units???)")
p <- p + xlab("Genotype")
p <- p + ggtitle( paste(my_locus, my_alleles, sep = ": ") )
#p <- p + geom_abline( slope = lm1$coefficients[2], intercept = lm1$coefficients[1], linewidth = 2, col = "#808080")
#p <- p + geom_line( data = my_vals, mapping = aes( x = x, y = y ), inherit.aes = F, linewidth = 2, col = "#808080" )
p
```



## Two loci, epistasis


```{r, results='markup'}
lm1 <- lm( pheno$Varin_Total ~ 
             gt2["4_68382150", ] + 
             gt2["7_290913", ] +
             I(gt2["4_68382150", ]^2) +
             I(gt2["7_290913", ]^2) +
             gt2["4_68382150", ]:gt2["7_290913", ]
             )
# pander(summary(lm1))

lm1s <- summary(lm1)


my_caption <- "Table X. Coefficients from a linear model fitted to 2 loci including additive, dominance, and epistatic effects. Dominance effects are colored in gray and include ^2 in the name. Epistatic effects are named by a pair of variants delimited by a colon (:). Effects that were statistically significant are in bold italics."



library(kableExtra)
my_dom <- grep("\\^2", rownames(lm1s$coefficients))
my_sig <- lm1s$coefficients[ , "Pr(>|t|)"] <= 0.05
my_sig <- (1:nrow(lm1s$coefficients))[my_sig]

knitr::kable(lm1s$coefficients, caption = my_caption) %>%
  kable_styling( full_width = FALSE ) %>%
  #row_spec(c(7:11), background = "#f5f5f5") %>%
  #row_spec(c(7:11), background = "#DCDCDC") %>%
  row_spec(my_dom, background = "#DCDCDC") %>%
#  row_spec(x, c(7, 12:15), bold = TRUE, italic = TRUE)
#  row_spec(c(7, 12:15), background = "#FFA07A")
  row_spec(my_sig, bold = T, italic = T)
#  row_spec(c(7, 12:15), bold = T, italic = T)
#lm1s$adj.r.squared
```






