---
title: "Varin GWAS Sup. Fig."
author: "Brian J. Knaus"
date: "`r format(Sys.time(), '%Y, %B %d')`"
output: 
  html_document:
    toc: true
    code_folding: hide
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
#
knitr::opts_chunk$set(echo = TRUE)
#knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(fig.align = "center")
knitr::opts_chunk$set(fig.width = 6)
#
knitr::opts_chunk$set(results = "hide")
#knitr::opts_chunk$set(results = "markup")
```


```{r}
t1 <- Sys.time()
```


## Phenotypes


```{r, results='hide'}
#
pheno <- readxl::read_xlsx("ERB x HO40 23 S1 Phenotypes_original_XT.xlsx")
# pheno <- read.csv("ERBxHO40_Phenotypes_adj.csv.gz")
#pheno$Total <- pheno$Total_adj
pheno[1:3, ]
#plot(pheno$Total_adj, pheno$OrderNorm_transformed)
```


```{r}
library(ggplot2)
#p <- ggplot( data = pheno, mapping = aes( x = Total) )
p <- ggplot( data = pheno, mapping = aes( x = OrderNorm_transformed) )
#p <- p + geom_histogram( binwidth = 1.0)
p <- p + geom_histogram( binwidth = 0.2)
#p <- p + geom_histogram( binwidth = 0.15)
p <- p + theme_bw()
p <- p + ylab("Count")
p <- p + xlab("Varin Content")
#p <- p + ggtitle("Phenotype")
p

p_pheno <- p + coord_flip()
```


## Manhattan plots


```{r}
# my_mhs <- c(
#   "/media/knausb/E737-9B48/knausb/allele_hw/GAPIT3_blink_v3_pca6/GAPIT.Association.GWAS_Results.BLINK.Varin_Total.csv"
#   )
#my_mhs <- "/media/knausb/E737-9B48/knausb/allele_hw/GAPIT3_blink_v3_pca6/GAPIT.Association.GWAS_Results.BLINK.Varin_Total.csv"
#
my_mhs <- "/media/knausb/E737-9B48/knausb/allele_hw/GAPIT3_blink_v3_pca6_normP/GAPIT.Association.GWAS_Results.BLINK.OrderNorm_transformed.csv"

#  "/media/knausb/E737-9B48/knausb/allele_hw/GAPIT3_blink_v3_pca6_normP/GAPIT.Association.PVE.BLINK.OrderNorm_transformed.csv"

# my_mhs <- "GAPIT.Association.GWAS_Results.BLINK.Varin_Total.csv.gz"

#my_mhs <- "gwas_GAPIT_varin_blink_pca6.csv.gz"
my_names <- sub("/media/knausb/E737-9B48/knausb/allele_hw/GAPIT3_blink_v3_", "", my_mhs)
my_names <- sub("/GAPIT.Association.GWAS_Results.", "_", my_names)
my_names <- sub(".gz", "", my_names)
my_names <- sub(".csv", "", my_names)
my_names

length(my_mhs) == length(my_names)
```


```{r}
nucs <- read.csv("nucs.csv.gz")
# nucs[1:3, 1:5]
```


```{r}

gwas <- read.csv(my_mhs)
#names(gwas)[3] <- "Position"
gwas[1:3, 1:6]

if(length(grep("POS", names(gwas))) == 0){
  #is.na(gwas$POS) <- TRUE
  gwas$POS <- gwas$Pos
  #gwas$POS <- gwas$Position
  for(i in 2:nrow(nucs)){
      gwas$POS[ gwas$Chr == nucs$Id[i] ] <- gwas$Pos[ gwas$Chr == nucs$Id[i] ] + sum( nucs$Length[1:(i - 1)] )
#      gwas$POS[ gwas$Chr == nucs$chrom_num[i] ] <- gwas$Pos[ gwas$Chr == nucs$chrom_num[i] ] + sum( nucs$Length[1:(i - 1)] )
  }
}
gwas$log_pvalue <- -log10(gwas$P.value)
gwas$alpha <- gwas$log_pvalue / max(gwas$log_pvalue)
gwas$alpha <- round(gwas$alpha, digits = 2)
#gwas[1:3, ]

gwas$chrom_num <- sub(".+\\chr", "", gwas$Chr)
gwas$chrom_num[ gwas$chrom_num == "X" ] <- 10
gwas$chrom_num[ gwas$chrom_num == "Y" ] <- 11
gwas$chrom_num <- as.numeric( gwas$chrom_num )

my_pal <- c("#420A6866", "#D94D3D66", "#FCAF1366", "#BF395266", 
            "#81206C66", "#62146E66", "#ED692566", "#A12A6366", 
            "#F98B0B66", "#F6D64566"
)
my_pal <- substr(start = 1, stop = 7, my_pal)
gwas$col <- my_pal[gwas$chrom_num]


p <- ggplot( data = gwas, mapping = aes( x = POS, y = log_pvalue) )
#p + geom_point(  )
# p + geom_point( color = gwas$col, size = gwas$log_pvalue * 0.6, alpha = 0.2)
# p + geom_point( color = gwas$col, size = gwas$log_pvalue * 0.6, alpha = gwas$log_pvalue)

p <- p + geom_point( color = gwas$col, size = gwas$log_pvalue * 0.6, alpha = gwas$alpha)
p <- p + geom_vline(xintercept = nucs$starts[-1], color = "#80808066", linewidth = 0.6)
p <- p + theme_bw()
p <- p + ylab( expression("-Log"["10"](italic(p)-value)) )
#p <- p + annotate( geom = "text", x = sum(nucs$Length[1:6]) + 1010217 , y = 11, label = "ALT4")
p <- p + annotate( geom = "text", x = sum(nucs$Length[1:6]) + 1010217 , y = 18, label = "ALT4")
p <- p + annotate( geom = "text", x = sum(nucs$Length[1:3]) + 72420315, y = 19.0, label = "BKR")
#p <- p + annotate( geom = "text", x = sum(nucs$Length[1:3]) + 72420315, y = 9.5, label = "BKR")
p <- p + scale_x_continuous( breaks = nucs$mids, labels = nucs$chrom_num)
p <- p + theme(axis.title.x = element_blank())
p <- p + ylim(0, 25)
#p <- p + ylim(0, 22)
p

p_mh <- p
```


```{r}
gwas_top <- gwas[sort.int(gwas$log_pvalue, decreasing = TRUE, index.return = TRUE)$ix[1:10], ]
#dput(names(gwas_top))
my_cols <- c("SNP", "Chr", "Pos", "P.value", "MAF", "nobs", "H.B.P.Value", 
"Effect", "log_pvalue")

knitr::kable(gwas_top[, my_cols], row.names = FALSE)
```




## VCF


```{r}
library(vcfR)
```


```{r, results='hide'}
#vcf <- read.vcfR("varin_Blink.vcf.gz")
#vcf <- read.vcfR("gwas_blink_pca6_top101.vcf.gz")
vcf <- read.vcfR("gwas_GAPIT_varin_blink_pca6_top101.vcf.gz")
```


```{r}
vcf
vcf@fix[1:3, 1:7]
gt <- extract.gt(vcf)
#is_het(gt)
gt[1:3, 1:6]
rownames(gt) <- sub("EH23b.chr", "", rownames(gt))

gt2 <- matrix( nrow = nrow(gt), ncol = ncol(gt), dimnames = dimnames(gt) )
gt2[ gt == "0/0" ] <- 0
gt2[ gt == "0/1" ] <- 1
gt2[ gt == "1/1" ] <- 2
#gt <- apply(gt, MARGIN = 2, as.numeric)

```


### Synchronize


```{r, results='asis'}
pheno <- pheno[colnames(gt), ]
all(pheno$ID == colnames(gt))
```


## Single, independent loci


### locus chr 4


```{r}
my_locus <- "4_70745833"
#my_locus <- "4_68382150"
# my_locus <- "4_36256098"
# my_locus <- "7_290913"
# my_locus <- "X_12488664"
# my_locus <- "X_44482355"
chr_pos <- paste(getCHROM(vcf), getPOS(vcf), sep = "_")
#my_alleles <- paste(getREF(vcf)[my_locus == chr_pos], getALT(vcf)[my_locus == chr_pos], sep = "/")
my_alleles <- paste( "0 = ", getREF(vcf)[my_locus == chr_pos], "; 1 = ", getALT(vcf)[my_locus == chr_pos], sep = "")

my_dat <- data.frame(
  #Varin_Total = pheno$Varin_Total,
  #Varin_Total = pheno$Total,
  Varin_Total = pheno$OrderNorm_transformed,
  gt = gt2[my_locus, ],
  pcol = "#00000044"
)
#my_dat$pcol[ my_dat$gt == 2 ] <- "#B22222"

#
lm1 <- lm( Varin_Total ~ gt, data = my_dat )
#lm1 <- lm( Varin_Total ~ gt, data = my_dat[ my_dat$gt != 2, ] )
lm11 <- lm( Varin_Total ~ gt + I(gt^2), data = my_dat )

my_vals <- data.frame(
  x = seq(0, 2, length.out=10)
)
my_vals$y <- lm11$coefficients[1] + my_vals$x * lm11$coefficients[2] + my_vals$x^2 * lm11$coefficients[3]

#my_dat$gt <- as.factor(my_dat$gt)
my_dat$gt2 <- as.factor(my_dat$gt)
lm12 <- lm( Varin_Total ~ gt2, data = my_dat )
#my_dat$gt <- as.numeric(my_dat$gt)
summary(lm12)

lm12$coefficients
#hist(lm12$residuals)
```


```{r}
all(pheno$ID == colnames(gt))
aggregate(my_dat$Varin_Total, by = list(my_dat$gt), mean)
pheno[1:3, ]

#my_locus <- "4_68382150"
my_gt = gt2[my_locus, ]

pheno$Total_adj <- pheno$Total
pheno$Total_adj[ my_gt == 1 ] <- pheno$Total_adj[ my_gt == 1 ] - lm12$coefficients[2]
pheno$Total_adj[ my_gt == 2 ] <- pheno$Total_adj[ my_gt == 2 ] - lm12$coefficients[3]

#hist(pheno$Total_adj)
#write.csv(pheno, file = "ERBxHO40_Phenotypes_adj.csv", row.names = FALSE)

```


```{r, results='markup'}
library(pander)
#pander(lm1)
pander(summary(lm1))
pander(summary(lm11))
#hist(lm11$fitted.values)
pander(summary(lm12))
```


```{r, results='markup'}
summary(lm1)
```


```{r, fig.width=6, fig.cap="**Figure X.** Ability of the genotype to predict the phenotype."}
library(ggplot2)
#my_dat$gt <- as.factor(my_dat$gt)
#p <- ggplot( data = my_dat, mapping = aes( x = gt, y = Varin_Total, fill = gt, group = gt ))
p <- ggplot( data = my_dat, mapping = aes( x = gt, y = Varin_Total, fill = gt, group = gt ))
p <- p + theme_bw()
p <- p + geom_violin()
p <- p + scale_x_continuous( 
  breaks = c(0, 1, 2),
  labels = c("0/0", "0/1", "1/1")
)
p <- p + theme(legend.position="none")
p <- p + scale_fill_continuous(type = "viridis", begin = 0.2)
p <- p + geom_jitter(shape=16, position=position_jitter(0.15), col = my_dat$pcol, size = 1)
#p <- p + ylab("Varin Total (units???)")
#p <- p + xlab("Genotype")
#p <- p + xlab("Chr 4, 67.6 Mbp")
p <- p + xlab("Chr 4, 70.7 Mbp")
p <- p + theme(axis.title.y = element_blank())
#p <- p + ggtitle( paste(my_locus, my_alleles, sep = ": ") )
#p <- p + ggtitle( "CHROM 4, POS 68,382,150" )
#
p <- p + geom_abline( slope = lm1$coefficients[2], 
                      intercept = lm1$coefficients[1],
                      linewidth = 1.2, col = "#808080")
# p <- p + geom_line( data = my_vals, mapping = aes( x = x, y = y ), inherit.aes = F, linewidth = 2, col = "#808080" )
# p <- p + annotate( geom = "text", x = -0.4, y = 55, label = "Chromosome 4,\nposition 68,382,150", size=3.0, adj=0)
p

#p <- p + ggtitle( "CHROM 4, POS 68,382,150" )

p_locus1 <- p
```


```{r, results='markup'}
#cor(my_dat$Varin_Total, my_dat$gt)
#cor.test(my_dat$Varin_Total, my_dat$gt, alternative = "greater")
cor.test(my_dat$Varin_Total, my_dat$gt, alternative = "two.sided")
```



```{r}
library(ggplot2)
p <- ggplot( data = pheno, mapping = aes( x = Total) )
p <- ggplot( data = pheno, mapping = aes( x = OrderNorm_transformed) )
p <- p + geom_histogram( binwidth = 1.0)
p <- p + theme_bw()
p <- p + ylab("Count")
p <- p + xlab("Varin Content")
#p <- p + ggtitle("Phenotype")
#p

p + facet_grid(. ~ as.factor(my_gt))
# p_pheno <- p + coord_flip()
```


### locus chr 7


```{r}
# my_locus <- "4_68382150"
# my_locus <- "4_36256098"
# my_locus <- "7_290913"
# my_locus <- "7_290840"
my_locus <- "7_1328482"
# my_locus <- "9_16275371"
# my_locus <- "X_12488664"
# my_locus <- "X_44482355"
chr_pos <- paste(getCHROM(vcf), getPOS(vcf), sep = "_")
#my_alleles <- paste(getREF(vcf)[my_locus == chr_pos], getALT(vcf)[my_locus == chr_pos], sep = "/")
my_alleles <- paste( "0 = ", getREF(vcf)[my_locus == chr_pos], "; 1 = ", getALT(vcf)[my_locus == chr_pos], sep = "")

my_dat <- data.frame(
  Varin_Total = pheno$OrderNorm_transformed,
  #Varin_Total = pheno$Total,
  #Varin_Total = pheno$Varin_Total,
  #gt = gt[my_locus, ],
  gt = gt2[my_locus, ],
  pcol = "#00000044"
)
#my_dat$pcol[ my_dat$gt == 2 ] <- "#B22222"

#lm1 <- lm( Varin_Total ~ gt, data = my_dat )
lm1 <- lm( Varin_Total ~ gt, data = my_dat[ my_dat$gt != 2, ] )
lm11 <- lm( Varin_Total ~ gt + I(gt^2), data = my_dat )

my_vals <- data.frame(
  x = seq(0, 2, length.out=10)
)
my_vals$y <- lm11$coefficients[1] + my_vals$x * lm11$coefficients[2] + my_vals$x^2 * lm11$coefficients[3]
```


```{r, results='markup'}
library(pander)
#pander(lm1)
pander(summary(lm1))
pander(summary(lm11))
```


```{r, fig.width=6, fig.cap="**Figure X.** Ability of the genotype to predict the phenotype."}
library(ggplot2)
#my_dat$gt <- as.factor(my_dat$gt)
p <- ggplot( data = my_dat, mapping = aes( x = gt, y = Varin_Total, fill = gt, group = gt ))
p <- p + theme_bw()
p <- p + geom_violin()
p <- p + scale_x_continuous( 
  breaks = c(0, 1, 2),
  labels = c("0/0", "0/1", "1/1")
)
p <- p + theme(legend.position="none")
p <- p + scale_fill_continuous(type = "viridis", begin = 0.2)
p <- p + geom_jitter(shape=16, position=position_jitter(0.15), col = my_dat$pcol, size = 1)
#p <- p + ylab("Varin Total (units???)")
p <- p + theme(axis.title.y = element_blank())
#p <- p + xlab("Genotype")
#p <- p + xlab("Chr 7, 290.8 Kbp")
p <- p + xlab("Chr 7, 1.3 Mbp")
#p <- p + ggtitle( paste(my_locus, my_alleles, sep = ": ") )
#p <- p + ggtitle( "CHROM 7, POS 290,913" )
#p <- p + annotate( geom = "text", x = 0, y = 60, label = "Chromosome 7,\nposition 290,913", size=4.0)
#
# p <- p + annotate( geom = "text", x = -0.4, y = 55, label = "Chromosome 7,\nposition 290,913", size=3.0, adj=0)

p <- p + geom_abline( slope = lm1$coefficients[2], intercept = lm1$coefficients[1], linewidth = 1.2, col = "#808080")
#p <- p + geom_line( data = my_vals, mapping = aes( x = x, y = y ), inherit.aes = F, linewidth = 2, col = "#808080" )
p

cor.test(my_dat$Varin_Total, my_dat$gt, alternative = "greater")

p_locus2 <- p
```


```{r, results='markup'}
#cor(my_dat$Varin_Total, my_dat$gt)
#cor.test(my_dat$Varin_Total, my_dat$gt, alternative = "greater")
cor.test(my_dat$Varin_Total, my_dat$gt, alternative = "two.sided")
```



## Multipanel


```{r}
library(ggpubr)

#ggarrange( p_mh, , nrow = 2)
#ggarrange( p_pheno, p_locus1, nrow = 1, ncol = 2)

p_pheno <- p_pheno + scale_x_continuous( limits = c(-3, 3), breaks = seq(-3,3, by=1))

ga1 <- ggarrange( p_pheno, p_locus1, p_locus2, 
                  nrow = 1, ncol = 3, 
                  labels = c("B", "C", "D"))
ga1
```


### Manhattan, phenotype, loci


```{r}
ggarrange( p_mh, ga1, 
           nrow = 2, ncol = 1,
           labels = c("A", ""),
           widths = 1, heights = c(0.6, 0.4))

# ggsave( filename = "SupFig25_OrderNorm.png", device = "png",
#         width = 6.5, height = 6.5,
#         units = "in", dpi = 300)

```


> Supplemental Figure 25. Varin (pentyl) cannabinoid ratio GWAS A) Significant GWA hits, using the BLINK model and normalized varin cannabinoid ratio data. B) Histogram of propyl:pentyl cannabinoid ratios for the F2 population (left), and arcsinh transformed propyl:pentyl cannabinoid ratios (right).


**Supplemental Figure 25.** Genomic markers associated with varin (pentyl) cannabinoid production.
A) Genome wide association study, using the BLINK model from GAPIT (PCA.total = 6) to test for an association between genomic markers and varin cannabinoid ratio data.
Positions plotted with a high elevation have significant associations with the varin production phenotype.
B) Distribution of propyl:pentyl cannabinoid ratios for the F2 population.
C) Varin content associated with a marker at chromosome 4, position 67,626,398.
D) Varin content associated with a marker at chromosome 7, position 290,840.
Note the bimodality in the phenotype data when the marker at chromosome 4, position 67,626,398 is not included.


```{r}

gwas_top <- gwas[sort.int(gwas$log_pvalue, decreasing = TRUE, index.return = TRUE)$ix[1:10], ]

knitr::kable(gwas_top[, c(2:6, 8, 10)], row.names = FALSE, digits=3, format.args = list(big.mark=","))

```



## Additional loci


### Chr 9, 19.2 Mbp

```{r}
# my_locus <- "4_68382150"
# my_locus <- "4_36256098"
# my_locus <- "7_290913"
# my_locus <- "7_290840"
# my_locus <- "X_12488664"
# my_locus <- "9_16275371"
my_locus <- "9_19245258"

# my_locus <- "X_44482355"

chr_pos <- paste(getCHROM(vcf), getPOS(vcf), sep = "_")
#my_alleles <- paste(getREF(vcf)[my_locus == chr_pos], getALT(vcf)[my_locus == chr_pos], sep = "/")
my_alleles <- paste( "0 = ", getREF(vcf)[my_locus == chr_pos], "; 1 = ", getALT(vcf)[my_locus == chr_pos], sep = "")

my_dat <- data.frame(
  Varin_Total = pheno$OrderNorm_transformed,
  #Varin_Total = pheno$Total,
  #Varin_Total = pheno$Total_adj,
  #Varin_Total = pheno$Varin_Total,
  #gt = gt[my_locus, ],
  gt = gt2[my_locus, ],
  pcol = "#00000044"
)
#my_dat$pcol[ my_dat$gt == 2 ] <- "#B22222"

#lm1 <- lm( Varin_Total ~ gt, data = my_dat )
lm1 <- lm( Varin_Total ~ gt, data = my_dat[ my_dat$gt != 2, ] )
lm11 <- lm( Varin_Total ~ gt + I(gt^2), data = my_dat )

my_vals <- data.frame(
  x = seq(0, 2, length.out=10)
)
my_vals$y <- lm11$coefficients[1] + my_vals$x * lm11$coefficients[2] + my_vals$x^2 * lm11$coefficients[3]
```


```{r, results='markup'}
library(pander)
#pander(lm1)
pander(summary(lm1))
pander(summary(lm11))
```


```{r, fig.width=6, fig.cap="**Figure X.** Ability of the genotype to predict the phenotype."}
library(ggplot2)
#my_dat$gt <- as.factor(my_dat$gt)
p <- ggplot( data = my_dat, mapping = aes( x = gt, y = Varin_Total, fill = gt, group = gt ))
p <- p + theme_bw()
p <- p + geom_violin()
p <- p + scale_x_continuous( 
  breaks = c(0, 1, 2),
  labels = c("0/0", "0/1", "1/1")
)
p <- p + theme(legend.position="none")
p <- p + scale_fill_continuous(type = "viridis", begin = 0.2)
p <- p + geom_jitter(shape=16, position=position_jitter(0.15), col = my_dat$pcol, size = 1)
#p <- p + ylab("Varin Total (units???)")
p <- p + theme(axis.title.y = element_blank())
#p <- p + xlab("Genotype")
#p <- p + xlab("Chr 7, 290.8 Kbp")
#p <- p + ggtitle(my_locus)
p <- p + xlab("Chr 9, 19.2 Mbp")
#p <- p + ggtitle( paste(my_locus, my_alleles, sep = ": ") )
#p <- p + ggtitle( "CHROM 7, POS 290,913" )
#p <- p + annotate( geom = "text", x = 0, y = 60, label = "Chromosome 7,\nposition 290,913", size=4.0)
#
# p <- p + annotate( geom = "text", x = -0.4, y = 55, label = "Chromosome 7,\nposition 290,913", size=3.0, adj=0)

p <- p + geom_abline( slope = lm1$coefficients[2], intercept = lm1$coefficients[1], linewidth = 1.2, col = "#808080")
#p <- p + geom_line( data = my_vals, mapping = aes( x = x, y = y ), inherit.aes = F, linewidth = 2, col = "#808080" )
#
p

#p_locus2 <- p
p_chr9_19 <- p
```




```{r}
t99 <- Sys.time()
t99 - t1
```

