---
title: "GeneChrom"
author: "Brian J. Knaus"
output: 
  html_document:
    toc: true
    code_folding: hide
date: "2023-03-14"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
#knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(results = "hide")
knitr::opts_chunk$set(fig.align = "center")
```


## GFF to genes

Documents how summary file is created.
Should not need to run this.

```{r, eval=FALSE}
# Point this at your version of the 'releases' directory.
release_dir <- "/media/knausb/E737-9B48/releases/scaffolded"
my_releases <- list.files(release_dir)
my_releases <- my_releases[-grep("EDTAOutput", my_releases)]

my_chroms <- as.data.frame(matrix( nrow = length(my_releases), ncol = 11 ))
rownames(my_chroms) <- my_releases
colnames(my_chroms) <- paste("chr", c(1:9, "X", "Y"), sep = "")
my_chroms[1:3, 1:11]

for( i in 1:length(my_releases)){
  tmp <- read.table( file = paste(release_dir, "/", my_releases[i], "/", my_releases[i], ".primary_high_confidence.gff3.gz", sep = ""), 
                     header = FALSE, sep = "\t", quote = "\"" )
  tmp <- tmp[ tmp[, 3] == "gene", ]
  
  tmp <- tmp[ grep("chr", tmp[,1]), ]
  tmp[1:3, 1:8]
  tmp <- table(tmp[, 1])
  names(tmp) <- sub("^.+chr", "chr", names(tmp))
  my_chroms[ my_releases[i], names(tmp)] <- tmp
}

my_chroms <- cbind(row.names(my_chroms), my_chroms)
colnames(my_chroms)[1] <- "Sample"
my_chroms[1:3, ]
#write.table(my_chroms, file = "gene_counts.csv", sep = ",", row.names = FALSE)
```


## Chrom lens

Documents how summary file is created.
Should not need to run this.

```{r, eval=FALSE}
my_nucs <-list.files("../FigureSideo/", pattern = "nuccomp.csv", full.names = TRUE)
my_samps <- sub("../FigureSideo//", "", my_nucs)
my_samps <- sub(".softmasked_nuccomp.csv", "", my_samps)

my_chroms <- as.data.frame(matrix( nrow = length(my_samps), ncol = 11 ))
rownames(my_chroms) <- my_samps
colnames(my_chroms) <- paste("chr", c(1:9, "X", "Y"), sep = "")
my_chroms[1:3, 1:11]

for( i in 1:nrow(my_chroms)){
  tmp <- read.csv( file = my_nucs[i] )
  tmp <- tmp[ grep("chr", tmp[,1]), ]
  tmp[1:3, 1:8]
  tmp2 <- tmp$Length
  names(tmp2) <- tmp$Id
  names(tmp2) <- sub("^.+chr", "chr", names(tmp2))
  my_chroms[ my_samps[i], names(tmp2)] <- tmp2
}

my_chroms <- cbind(row.names(my_chroms), my_chroms)
colnames(my_chroms)[1] <- "Sample"
my_chroms[1:3, ]
#write.table(my_chroms, file = "chrom_lengths.csv", sep = ",", row.names = FALSE)
```



## Chrom by genes


```{r}
my_data <- readr::read_csv("gene_counts.csv", )
my_data
```


```{r}
gene_cnts <- data.frame(
  Sample = my_data[, 1],
  Count = rowSums(my_data[, -1], na.rm = TRUE)
)
knitr::kable(gene_cnts)
mean(gene_cnts$Count)
```




```{r}
library(tidyr)
data_long <- my_data %>%
  pivot_longer( cols = !Sample, names_to = "Chrom", values_to = "Count")
#data_long$Length <- data_long$Length / 1e6
gcount <- data_long
sum(is.na(data_long$Count))
data_long <- data_long[!is.na(data_long$Count), ]
data_long

#my_data <- readr::read_csv("../Figure1b/chrom_lengths.csv")
my_data <- readr::read_csv("chrom_lengths.csv")
my_data$Sample <- as.factor( my_data$Sample )
my_data

library(tidyr)
data_long <- my_data %>%
  pivot_longer( cols = !Sample, names_to = "Chrom", values_to = "Length")
data_long$Length <- data_long$Length / 1e6
data_long
glength <- data_long

```



```{r}
glength
gcount
data_long <- cbind(glength, gcount$Count)
names(data_long)[4] <- "Count"
data_long[1:3, ]

my_pal <- RColorBrewer::brewer.pal(n=12, name = "Paired")
#my_pal[11] <- "#B15928"
my_pal[11] <- "#C71585"
my_pal <- paste(my_pal, "88", sep = "")

library(ggplot2)
# Basic scatter plot
p <- ggplot(data_long, aes(x=Length, y=Count, color=Chrom)) 
p <- p + geom_point(size=2)
#p <- p + geom_smooth(method=lm)
p <- p + geom_smooth(method=lm, se=FALSE, linewidth = 1)

p <- p + theme_bw()
p <- p + theme(legend.title = element_blank()) 
#p <- p + theme(legend.position = "left")
p <- p + theme(legend.position = "right")
#p <- p + theme( legend.spacing.y = unit(1.0, 'mm') )
## important additional element
#p <- p + guides(fill = guide_legend(byrow = TRUE))
#p + theme(legend.spacing.y = unit(1.0, 'mm')) +
#  guides(fill = guide_legend(byrow = TRUE))
#p <- p + scale_color_brewer(palette="Dark2")
#p <- p + scale_color_brewer(palette="Paired")
p <- p + scale_color_manual(values=my_pal)
#p <- p + scale_color_brewer(palette="Set3")
p <- p + xlab("Chromosome length (Mbp)")
p <- p + ylab("Genes per chromosome")
gchrom <- p
```


```{r}
gchrom
# ggsave(filename = "gene_by_chrom.tiff",
#        device = "tiff", 
#        #width = 3.25, height = 3.25, 
#        width = 6.5, height = 3.25, 
#        units = "in", dpi = 300, compression = "lzw")
```


## Gene counts


```{r, fig.width=12}
data_long[1:3, ]

my_lvls <- sort(unique(data_long$Chrom), decreasing = TRUE)
#dput(unique(data_long$Chrom))

data_long$Chrom <- factor(data_long$Chrom, levels = my_lvls)

p <- ggplot(data_long, aes(x=Sample, y=Count, 
                           #color=Chrom, 
                           fill=Chrom)) 
p <- p + geom_bar(stat="identity")
p <- p + theme_bw()
p <- p + theme( axis.text.x = element_text(angle = 60, hjust = 1) )
#p

p <- p + scale_fill_brewer(palette="Paired")
p


```





## Lengths


```{r, results='markup'}

data_long[1:3, ]
range(data_long$Length, na.rm = TRUE)

minchr <- data_long[ data_long$Length == min(data_long$Length, na.rm = TRUE), ]
minchr <- minchr[!is.na(minchr$Length), ]
minchr

maxchr <- data_long[ data_long$Length == max(data_long$Length, na.rm = TRUE), ]
maxchr <- maxchr[!is.na(maxchr$Length), ]
maxchr

#data_long[ data_long$Length == min(data_long$Length, na.rm = TRUE), ]


dlchr1 <- data_long[ data_long$Chrom == "chr1", ]
dlchr1[1:3, ]

hist(dlchr1$Length, breaks = seq(45, 90, by = 1))
sort(na.omit(dlchr1$Length))

minchr <- dlchr1[ dlchr1$Length == min(dlchr1$Length, na.rm = TRUE), ]
minchr <- minchr[!is.na(minchr$Length), ]
minchr

maxchr <- dlchr1[ dlchr1$Length == max(dlchr1$Length, na.rm = TRUE), ]
maxchr <- maxchr[!is.na(maxchr$Length), ]
maxchr


dlchrY <- data_long[ data_long$Chrom == "chrY", ]
dlchrY[1:3, ]
dlchrY[!is.na(dlchrY$Length), ]

hist(dlchrY$Length, breaks = seq(100, 120, by = 1))

#na.omit(dlchrY$Length)
sort(na.omit(dlchrY$Length)[1:4])

```




