---
title: "Figure 1"
author: "Brian J. Knaus"
date: "`r format(Sys.time(), '%Y, %B %d')`"
output:
  html_document:
    toc: true
    code_folding: hide
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



## Abstract (prophetic)

Cell Summary (no longer than 150 words)

The *Cannabis sativa* L. genome has coevolved with human culture over thousands of years as a source of food, fiber, and therapeutic compounds.
The ~700 Mbp genome is organized into nine pairs of autosomes and as pair of sex chromosomes where genes are predominantly located near the ends of each chromosome.
The majority of each chromosome is occupied by complete and fragmented repetitive elements occupying ~70% of the genome, a level comparable to maize.
Predominant among the intact repetitive elements are Gypsy long terminal repeat elements.
Within this expanse of repetitive sequences, each chromosome includes a metacentric or acrocentric chromosome.
<span style="color: red;">Understanding the genomic landscape within *Cannabis* that has evolved along with human culture may help us to better appreciate these cultural ties as well as guide it's agronomic future</span>.

**Keywords:** *Cannabis*, Cannabinoid, hemp, Long Terminal Repeat (LTR) retrotransposon.

<br> 

---


## Brian's notes


The term 'gypsy' has been used in genetics to refer to a class of Long Terminal Repeat (LTR) retrotransposon.
The term 'gyspy' is also increasingly recognized as a pejorative term for the Romani people.
Here we retain the use of this term because the genetics community has yet to agree on an alternative.
We would also like to acknowledge that an alternate terminology may be more appropriate in the future.



- Gene density
- LTRs, gypsy
- Cannabinoids
- Sex chromosome
- Terpenes
- Disease ressistance
- Chemotype: THCAS, THCAS:CBDAS, CBDAS, CBG, cannabinoid free
- Chromosome/Genome length variability
- PanGenome
  - Diversity of gene classes (cannabinoids, terpenes, disease resistance)?



## Data management


```{r}
t1 <- Sys.time()
nucs <- read.csv("EH23a.softmasked_nuccomp.csv")

nucs$GC <- rowSums( nucs[, c("C", "c", "G", "g")] )/rowSums( nucs[, c("A", "a", "C", "c", "G", "g", "T", "t")] )
nucs$GC <- nucs$GC * 100
#nucs[1:3, ]
wins <- read.csv("EH23a.softmasked_wins1e6.csv")
#wins <- read.csv("EH23a.softmasked_wins1e7.csv")
wins$CGs <- 0
for( i in unique(wins$Id) ){
  wins$CGs[ wins$Id == i] <- wins$CG[ wins$Id == i] - min(wins$CG[ wins$Id == i], na.rm = TRUE)
  wins$CGs[ wins$Id == i] <- wins$CGs[ wins$Id == i]/max(wins$CGs[ wins$Id == i], na.rm = TRUE)
}

#wins[1:3, ]
#nrow(wins)

blst <- read.csv("EH23a_blastn.csv", header = FALSE)
colnames(blst) <- c('qseqid','qlen','sseqid','slen','qstart','qend',
                    'sstart','send','evalue','bitscore','score','length',
                    'pident','nident','mismatch','positive','gapopen',
                    'gaps','ppos','sstrand','staxids',
                    'sblastnames','salltitles','sseq')[1:21]

blst$sseqid <- factor(blst$sseqid, levels = c("EH23a.chr1", "EH23a.chr2", "EH23a.chr3", "EH23a.chr4", "EH23a.chr5", "EH23a.chr6", "EH23a.chr7", "EH23a.chr8", "EH23a.chr9", "EH23a.chrX"))
blst <- blst[blst$qseqid == "CsatSD_centromere_370bp", ]
#table(blst$qseqid)
# blst[1:3, 1:10]

cann_blst <- read.csv("cann_3dom_EH23a_tblastn.csv", header = FALSE)
colnames(cann_blst) <- c('qseqid','qlen','sseqid','slen','qstart','qend','sstart','send','evalue','bitscore','score','length','pident','nident','mismatch','positive','gapopen','gaps','ppos','sstrand','sseq')
#
cann_blst <- cann_blst[grep("SignalP|FAD|BBE", cann_blst$qseqid, invert = TRUE, value = F), ]
cann_blst$name <- sub("^.+_", "", cann_blst$qseqid)

# table(cann_blst$qlen)
cann_blst <- cann_blst[ cann_blst$gaps <= 0, ]
cann_blst <- cann_blst[ cann_blst$mismatch <= 10, ]

#table(cann_blst$qseqid)

# cann_blst[1:3, 1:10]
# cann_blst[1:3, 10:20]

knitr::kable(cann_blst[, c(1, 2, 3, 7, 13:15, 18)], caption = "**Table 1. Cannabinoid synthase genes.**")

```


```{r}
myBed <- data.frame( chrom = nucs$Id, chromStart = 1, chromEnd = nucs$Length, gc = nucs$GC)

chrm <- sub("^.+chr" ,"", myBed$chrom)
chrm[10] <- 10
myBed <- myBed[sort.int(as.numeric(chrm), index.return = TRUE)$ix, ]

#dput(myBed$chrom)
myBed$chrom <- factor(myBed$chrom, levels = c("EH23a.chr1", "EH23a.chr2", "EH23a.chr3", "EH23a.chr4", "EH23a.chr5", "EH23a.chr6", "EH23a.chr7", "EH23a.chr8", "EH23a.chr9", "EH23a.chrX"))

myBed$gcs <- myBed$gc - min(myBed$gc)
myBed$gcs <- myBed$gcs/max(myBed$gcs)
myBed$gcs <- myBed$gcs * 100
myBed$gcs <- as.integer(myBed$gcs)
myBed$gcs[ myBed$gcs == 0 ] <- 1
myBed$cols <- viridisLite::magma(n=100, alpha = 0.4, begin = 0.2, end = 1.0)[ myBed$gcs ]

#head(myBed)
```


```{r}
gff <- read.table("EH23a.primary_high_confidence.gff3.gz", sep = "\t", quote = "\"")
colnames(gff) <- c("seqid","source","type","start","end","score","strand","phase","attributes")
#gff[1:3, 1:8]
genes <- gff[grep("gene", gff$type), ]
#nrow(genes)

```

```{r}
gff2 <- read.table("EH23a.unmasked.fasta.mod.EDTA.intact.gff3", sep = "\t", quote = "\"")
colnames(gff2) <- c("seqid","source","type","start","end","score","strand","phase","attributes")
#
gff2[1:3, 1:8]
genes2 <- gff2[grep("Gypsy", gff2$type), ]
#

ltri <- as.numeric(sub("ltr_identity=", "", unlist(lapply(strsplit(genes2$attributes, split = ";"), function(x){grep("ltr_identity=", x, value = TRUE)}))))
#hist(ltri)
genes2 <- genes2[ltri >= 0.98, ]
#nrow(genes2)

```



```{r}
chrlen <- read.csv("chrom_lengths.csv", row.names = 1)
#chrlens[1:3, ]
```


## Circlize plot


```{r, fig.width=10, fig.height=10, message=FALSE, fig.cap="**Figure 1. Genomic architecture of the** ***Cannabis sativa*** **genome.** (Panel A) Circular representation of the strain ERBxHO40_23 haplotype 'a' assembly. (A) Gene density, the background of each panel is colored by GC content (legend in lower left corner). (B) Density of the motif 'CG' (also referred to as 'CpG', a motif associated with methylation). (C) *Cannabis* centromere motif density (370bp motif, blastn hits). (Panel B) Ideogram of the hemp genome. Chromosomes are named according to the CBDRx assembly (NCBI: GCF_900626175.2_cs10). Each autosomal ideogram summarizes the length of assemblies from 69 strains. The ideogram for the X chromosome summarizes the length of X chromosomal assemblies from 65 strains. The Y chromosomal ideogram summarizes the length of Y chromosomal assemblies from 4 strains. (Panel C) A Y chromosome panel? The strain ERBxHO40 resulted from a cross between strains ERB (type III or CBD dominant) and HO40 (type I or THC dominant; a genetically female plant (XX) that was chemically masculinized to produce pollen flowers to facilitate the cross)."}


suppressPackageStartupMessages(library(circlize))

#png(filename = "gdplot.png", width = 6.5, height = 6.5, units = "in", res = 300)

layout(
  matrix( c(1, 1, 2, 3), nrow = 2, ncol = 2, byrow = TRUE),
  widths = 1, heights = c(0.7, 0.3)
  )

# circos.clear()

circos.par( start.degree = 90, 
            gap.after = c(rep(0, times = 9), 8)
            #gap.after = 10
            )

circos.genomicInitialize( 
  data = myBed,
  axis.labels.cex =  0.4 * par("cex"),
  labels.cex = 0.8 * par("cex") 
)

# circos.labels(
#   sectors = as.factor(cann_blst$sseqid),
#   x = cann_blst$sstart,
#   labels = cann_blst$name,
# #  side = "outside",
#   side = "inside"
# #  track.margin = c(0.1, 0.1)
#   )


circos.trackHist( 
  sectors = as.factor(genes$seqid),
  x = genes$start,
  track.height = 0.2,
  bg.border = "#FFFFFF",
  bg.col = myBed$cols,
  col = "#808080",
  border = NA,
  breaks = seq(0, 1e9, by = 2e6) )

circos.trackHist( 
  sectors = as.factor(genes2$seqid),
  x = genes2$start,
  track.height = 0.2,
  bg.border = "#FFFFFF",
  bg.col = myBed$cols,
  col = "#808080",
  border = NA,
  breaks = seq(0, 1e9, by = 1e6)
#  breaks = seq(0, 1e9, by = 2e6) 
  )

circos.track( 
    ylim = c(0, 1),
    bg.col = NA,
    bg.border = "#00000044",
    track.height = 0.20, bg.lty = 3, bg.lwd = 1)

for( i in unique( wins$Id ) ){
  tmp <- wins[ wins$Id == i, ]
  circos.rect(
    xleft = tmp$Start,
    ybottom = 0,
    xright = tmp$End,
    ytop = tmp$CGs,
    sector.index = i,
    col = "#1E90FF66",
    border = "#1E90FF11",
    lwd = 4
  )
}

for( i in unique( cann_blst$sseqid ) ){
  tmp <- cann_blst[ cann_blst$sseqid == i, ]
  circos.rect(
    xleft = tmp$sstart,
    ybottom = 0.8,
    xright = tmp$send,
    ytop = 1.0,
    sector.index = i,
    col = "#B2222244",
    border = "#B2222244",
#    border = NA,
    lwd = 2
  )
}


circos.track(
    ylim = c(0, 1),
    bg.col = NA,
    bg.border = "#00000044",
    track.height = 0.10, bg.lty = 3, bg.lwd = 1)

for( i in unique( blst$sseqid ) ){
  tmp <- wins[ blst$sseqid == i, ]

  circos.rect(
    xleft = tmp$Start,
    ybottom = 0.2,
    xright = tmp$End,
    ytop = 0.8,
    sector.index = i,
    col = "#1E90FF22",
    border = "#FFFFFF00",
    lwd = 4
  )
}

# circos.trackHist(
#   sectors = blst$sseqid,
#   x = blst$sstart,
#   track.height = 0.2,
#   #border = "#FFFFFF",
#   #bg.border = "#FFFFFF",
#   bg.border = "#00000044",
#   bg.col = "#FFFFFF",
# #  bg.col = myBed$cols,
#   #col = c("#FF0000", "#00FF00"),
#   #col = "#C0C0C0",
#   col = "#808080",
#   border = NA,
#   breaks = seq(0, 1e9, by = 1e3)
# )




circos.yaxis( side = "left", at = seq(0, 400, by = 50), labels = c("", "", "", "", "A", "", ""), sector.index =  "EH23a.chr1", track.index = 2)
#circos.yaxis( side = "left", at = c(0, 180, 290), labels = c("", "A", ""), sector.index =  "EH23a.chr1", track.index = 2)
circos.yaxis( side = "left", at = seq(0, 20, by = 5), labels = c("", "", "B", "", ""), sector.index =  "EH23a.chr1", track.index = 3)
#circos.yaxis( side = "left", at = c(0, 0.5, 1), labels = c("", "B", ""), sector.index =  "EH23a.chr1", track.index = 3)
#circos.yaxis( side = "left", at = c(0, 1000, 2000), labels = c("", "C", ""), sector.index =  "EH23a.chr1", track.index = 4)
circos.yaxis( side = "left", at = c(0, 0.5, 1), labels = c("", "C", ""), sector.index =  "EH23a.chr1", track.index = 4)
circos.yaxis( side = "left", at = c(0, 0.5, 1), labels = c("", "D", ""), sector.index =  "EH23a.chr1", track.index = 5)

text(-1, 1, "A", font = 2, cex = 2)
myLeg <- rep("", times = 10)
myLeg[c(2, 9)] <- c("High GC%", "Low GC%")

legend('bottomleft',
       legend = myLeg,
       fill = viridisLite::magma(n=10, alpha = 0.4, begin = 1.0, end = 0.2),
       border = NA,
       pt.cex = 2,
       bty = "n", y.intersp = 0.5)


title( main = "EH23a", line = -1 )


circos.clear()

#####

plot(1, 1, type = "n", xlim = c(1, 11), ylim = c(0, 110), 
     xlab = "", ylab = "", axes = FALSE, frame.plot = FALSE)
abline( h = seq( 0, 200, by = 10), lty = 3, col = "#808080")

for( i in 1:nrow(chrlen) ){
  rect( xleft  = 1:11 - 0.2, ybottom = -chrlen[i, ]/2/1e6 + 60, 
        xright = 1:11 + 0.2, ytop    =  chrlen[i, ]/2/1e6 + 60, 
        density = NULL, angle = 45,
        col = "#1E90FF20",
#        col = "#1E90FF22",
#        col = "#1E90FF66",
        border = NA #, lty = par("lty"), lwd = par("lwd")
  )
}

axis( side = 1, at = 1:11, labels = colnames(chrlen), lty = 0, las = 3 )
axis( side = 2 )
title( ylab = "Chromosome\nlength (Mbp)", line = 2 )
mtext(text = "B", side = 3, line = -1.2, adj = -0.18, cex = 1.4, font = 2)

#####

plot(1, 1, type = "n", xlim = c(1, 11), ylim = c(0, 110), 
     xlab = "", ylab = "", axes = FALSE, 
     frame.plot = TRUE)

text(5, 50, "Y chromosome plot?")
mtext(text = "C", side = 3, line = 0, adj = 0, cex = 1.4, font = 2)

#par( mfrow = c(1, 1) )

#dev.off()
```


```{r}
t99 <- Sys.time()
t99 - t1
```

