---
title: "Figure1"
output: 
  html_document:
    code_folding: hide
    toc: true
date: "2023-02-12"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.align = "center")
knitr::opts_chunk$set(warning = FALSE)
```


## Data input




```{r}
# nucs <- read.csv("EH23a.softmasked_nuccomp.csv")
# nucs$GC <- rowSums( nucs[, c("C", "c", "G", "g")] )/rowSums( nucs[, c("A", "a", "C", "c", "G", "g", "T", "t")] )
# nucs$GC <- nucs$GC * 100
```


```{r}
gff <- read.table("../Figure1/EH23a.primary_high_confidence.gff3.gz", sep = "\t", quote = "\"")
#
gff[1:3, 1:8]
genes <- gff[ gff[, 3] == "gene", ]
# genes[1:3, 1:8]
```

```{r}
gff <- read.table("../Figure1/EH23a.unmasked.fasta.mod.EDTA.intact.gff3", sep = "\t", quote = "\"")
names(gff) <- c("seqid","source","type","start","end","score","strand","phase","attributes")

#gff$chrn <- sub("^.+chr", "", gff$V1)
gff$chrn <- sub("^.+chr", "", gff$seqid)
gff$chrn[ gff$chrn == "X" ] <- 10
gff$chrn <- as.numeric(gff$chrn)

gff$classification <- unlist(lapply(strsplit(gff[,9], split = ";"), function(x){ grep("Classification=", x, value = TRUE) }))
gff$classification <- sub("^Classification=", "", gff$classification)
gff$classification <- sub("^LTR/Copia", "Ty1", gff$classification)
gff$classification <- sub("^LTR/Gypsy", "Ty3", gff$classification)
#
gff[1:8, c(1:8, 10)]

# Ty3
Ty3 <- gff[ gff$classification == "Ty3", ]
Ty3 <- Ty3[ Ty3$type == "Gypsy_LTR_retrotransposon", ]
#Ty3 <- Ty3[ Ty3$V3 == "Gypsy_LTR_retrotransposon", ]
Ty3[1:8, c(1:8, 10:11)]

#table(gff$classification, gff$V1)
#tetbl <- table(gff$V1, gff$classification)
tetbl <- table(gff$seqid, gff$classification)
tetbl <- rbind(tetbl, colMeans(tetbl))
#tetbl

Ty3$motif <- sub("^motif=", "", unlist(lapply(strsplit(Ty3[, 9], split = ";"), function(x){ grep("motif", x, value = TRUE) })))

my_tbl <- table(Ty3$motif, Ty3$chrn)
#my_tbl[, 1:10]
my_mat <- as.matrix(my_tbl)
#my_mat[, 1:10]

my_mat2 <- my_mat[apply(my_mat, MARGIN = 1, function(x){ sum( x != 0 ) }) == 1, ]

my_mat2 <- my_mat2[sort.int(as.numeric(apply(my_mat2, MARGIN = 1, function(x){ paste(x, collapse = "") })), 
         index.return = T)$ix, ]
my_mat2

```


```{r, eval=FALSE}
# my_cmd <- c("~/gits/hempy/bin/fast_win.py --win_size 1000000 /media/knausb/E737-9B48/releases/EH23a/EH23a.softmasked.fasta")
# my_cmd
# system(my_cmd)
# x <- read.table("EH23a.softmasked_wins.csv", header = T, sep = ",")
# #x[1:80, ]

```


```{r}
#wins <- read.csv("EH23a.softmasked_wins1e5.csv")
wins <- read.csv("EH23a.softmasked_wins1e6.csv")
#wins <- read.csv("EH23a.softmasked_wins1e7.csv")
# wins <- read.csv("EH23a.softmasked_wins.csv")
wins$CGs <- 0
for( i in unique(wins$Id) ){
  wins$CGs[ wins$Id == i] <- wins$CG[ wins$Id == i] - min(wins$CG[ wins$Id == i], na.rm = TRUE)
  wins$CGs[ wins$Id == i] <- wins$CGs[ wins$Id == i]/max(wins$CGs[ wins$Id == i], na.rm = TRUE)
}
wins$chrn <- sub(".+chr", "", wins$Id)
wins$chrn[ wins$chrn == "X" ] <- 10
wins$chrn <- as.numeric(wins$chrn)
wins$ATs <- 1 - wins$CGs
#wins[1:3, ]

# 
cwins <- wins[wins$CGs == 1 & !is.na(wins$CGs) , ]
# cwins <- cwins[ !duplicated( cwins$Id ), ]
# cwins$cent <- cwins$Start + cwins$Win_length/2

```


Centromere genes

```{r}
gff <- read.table("../Figure1/EH23a.primary_high_confidence.gff3.gz", sep = "\t", quote = "\"")
#gff[gff$V3 == "remark", ]
#genes[ genes$V1 == cwins$Id[i] & genes$V4 >= cwins$Start[i] & genes$V5 <= cwins$End[i], ]
i <- 1
gff2 <- gff[ gff$V1 == cwins$Id[i] & gff$V4 >= cwins$Start[i] & gff$V5 <= cwins$End[i], ]
for( i in 2:nrow(cwins) ){
#  gff2 <- gff[ gff$V1 == cwins$Id[i] & gff$V4 >= cwins$Start[i] & gff$V5 <= cwins$End[i], ]
  gff2 <- rbind(gff2, gff[ gff$V1 == cwins$Id[i] & gff$V4 >= cwins$Start[i] & gff$V5 <= cwins$End[i], ])
}

#cbind(1:nrow(gff2), gff2[, 1:8])
#table(gff2$V3)

#gff2[gff2$V3 == "CDS", ]
#gff2[gff2$V3 == "gene", ]
gff2 <- gff2[gff2$V3 == "mRNA", ]

#unlist(lapply(strsplit(gff2$V9, split = ";"), function(x){ grep("desc", x, value = TRUE) }))

```



```{r}
# lndmrk <- read.csv("csat_landmark_EH23a_blastn.csv", header = FALSE)
# names(lndmrk) <- c("qseqid","qlen","sseqid","slen","qstart","qend","sstart","send","evalue","bitscore","score","length","pident","nident","mismatch","positive","gapopen","gaps","ppos","sstrand")
# lndmrk$chrn <- sub("^.+chr", "", lndmrk$sseqid)
# lndmrk$chrn[ lndmrk$chrn == "X" ] <- 10
# lndmrk$chrn[ lndmrk$chrn == "Y" ] <- 11
# lndmrk$chrn <- as.numeric(lndmrk$chrn)
# 
# lndmrk[1:3, ]
```



### Variant data


```{r, include=FALSE}
# my_files <- list.files("/media/knausb/Vining_lab/knausb/GENOMES_NRQ_2267/ERBxHO40_hapERB/het_wins/", full.names = TRUE, pattern = "csv.gz$")
# 
# variants <- read.table(my_files[1], sep = ",", header = TRUE)
# names(variants) <- c("CHROM", "POS", "n", "Allele_counts", "He", "Ne", "0/0", "0/1", "1/1", NA)
# for(i in 2:length(my_files)){
#   tmp <- read.table(my_files[i], sep = ",", header = TRUE)
#   names(tmp) <- c("CHROM", "POS", "n", "Allele_counts", "He", "Ne", "0/0", "0/1", 
# "1/1", NA)
#   variants <- rbind(variants, tmp)
# }
# names(variants) <- c("CHROM", "POS", "n", "Allele_counts", "He", "Ne", "REFhom", "HET", "ALThom", "Missing")
# variants <- variants[variants$Missing <= 10, ]
# variants[1:3, ]
# nrow(variants)
# #write.table(x = variants, file = "ERBxHO40vars.csv", quote = TRUE, sep = ",")
```


```{r}
vars <- read.table("ERBxHO40vars.csv.gz", header = TRUE, sep = ",")
nrow(vars)
vars <- vars[vars$Missing == 0, ]
nrow(vars)
vars[1:3, ]

vars$ERBallele <- as.numeric(unlist(lapply(strsplit(vars$Allele_counts, split = ","), function(x){x[1]})))
vars$HO40allele <- as.numeric(unlist(lapply(strsplit(vars$Allele_counts, split = ","), function(x){x[2]})))

vars$ERBallele <- vars$ERBallele / (vars$n * 2)
vars$HO40allele <- vars$HO40allele / (vars$n * 2)


my_chrom <- "chr_1e"
plot( vars$POS[ vars$CHROM == my_chrom ], vars$ERBallele[ vars$CHROM == my_chrom ],
      pch = 20, col = "#C0C0C066", ylim = c(0, 1))

points( vars$POS[ vars$CHROM == my_chrom ], vars$HO40allele[ vars$CHROM == my_chrom ],
      pch = 20, col = "#4682B4")
```


```{r}
vars$chrn <- sub("chr_", "", vars$CHROM)
vars$chrn <- sub("e$", "", vars$chrn)
vars$chrn[ vars$chrn == "X" ] <- 10
vars$chrn <- as.numeric(vars$chrn)
table(vars$chrn)

library(ggplot2)
ggplot(vars, aes(x=chrn, y=POS, group=CHROM)) +
  geom_point()
  

my_cols1 <- paste(RColorBrewer::brewer.pal( n = 12, name = "Set3"), "08", sep="")
#my_cols1 <- paste(RColorBrewer::brewer.pal( n = 12, name = "Paired"), "08", sep="")

hist(vars$ERBallele[vars$chrn == 2], xlim = c(0, 1))

hist(vars$HO40allele[vars$chrn == 2], xlim = c(0, 1))

p <- ggplot(vars, aes(x=chrn+ERBallele-0.5, y=POS, group=CHROM)) +
  geom_point( aes( color = CHROM ) )
p <- p + scale_color_manual(values=my_cols1)
#p <- p + geom_point( data = vars, aes(x = chrn-HO40allele, y=POS), color = vars$chrn)
#
p <- p + geom_point( data = vars, aes(x = chrn+HO40allele-0.5, y=POS), color = my_cols1[vars$chrn])
#p <- p + geom_point( data = vars, aes(x = chrn-0, y=POS), color = my_cols1[vars$chrn])
#p <- p + scale_color_manual(values=my_cols1)
p <- p + theme_bw()
p <- p + theme(legend.position='none')
p <- p + scale_x_continuous(breaks=seq(1, 10, 1))
p <- p + xlab("Chromosome")
p




#p + scale_color_brewer(palette="Dark2")
#p + scale_color_brewer(palette="Set3")
#  geom_point( aes( color = CHROM, alpha = 1/10 ) )
#  geom_point( aes( alpha = 1/10 ) )


```




### Windowize


```{r}
# wins[1:3, ]
# genes[1:3, 1:8]

wins$gcnt <- 0
wins$ty1cnt <- 0
wins$ty3cnt <- 0
for(i in 1:nrow(wins)){
   tmp <- genes[genes$V1 == wins$Id[i] & genes$V4 >= wins$Start[i] & genes$V5 < wins$End[i], ]
   wins$gcnt[i] <- nrow(tmp)

   tmp <- gff[gff$V1 == wins$Id[i] & gff$V4 >= wins$Start[i] & gff$V5 < wins$End[i], ]

   wins$ty1cnt[i] <- sum(tmp$classification == "Ty1")
   wins$ty3cnt[i] <- sum(tmp$classification == "Ty3")
}

wins$gcntsc <- wins$gcnt - min(wins$gcnt)
wins$gcntsc <- wins$gcntsc / max(wins$gcntsc)
#range(round( wins$gcntsc * 100 ))
my_index <- round( wins$gcntsc * 100 )
#my_index <- 100 - my_index
my_index[ my_index <= 0] <- 1
# 
# #wins$gcol <- heat.colors(n=100)[ my_index ]
# #wins$gcol <- colorRampPalette(c("yellow", "orange", "red"))( 100 )[ my_index ]
# #wins$gcol <- colorRampPalette(c("red", "orange", "yellow"))( 100 )[ my_index ]
# #wins$gcol <- colorRampPalette(c("#0000FF", "#228B22", "#A0522D"))( 100 )[ my_index ]
# #wins$gcol <- colorRampPalette(c("#87CEEB", "#3CB371", "#228B22", "#A0522D"))( 100 )[ my_index ]
# 
# 
# #wins$gcol <- viridisLite::plasma(n = 100, alpha = 1, begin = 0.1, end = 1)[ my_index ]
# wins$gcol <- viridisLite::magma(n = 100, alpha = 1, begin = 0.2, end = 1.00)[ my_index ]
# 
# wins$chr <- sub("^.+chr", "chr", wins$Id)
# #
# wins[1:3, ]

```


```{r, include=TRUE}
# Cannabinoid BLAST results.

# cann_blst <- read.csv("cann_3dom_EH23a_tblastn.csv", header = FALSE)
# colnames(cann_blst) <- c('qseqid','qlen','sseqid','slen','qstart',
#                          'qend','sstart','send','evalue','bitscore',
#                          'score','length','pident','nident','mismatch',
#                          'positive','gapopen','gaps','ppos','sstrand',
#                          'sseq')
# #
# cann_blst <- cann_blst[grep("SignalP|FAD|BBE", cann_blst$qseqid, invert = TRUE, value = F), ]
# cann_blst$name <- sub("^.+_", "", cann_blst$qseqid)
# 
# # table(cann_blst$qlen)
# cann_blst <- cann_blst[ cann_blst$gaps <= 0, ]
# cann_blst <- cann_blst[ cann_blst$mismatch <= 10, ]
# 
# # write.table(cann_blst, file = "cann_3dom_EH23a_tblastn_filter.csv",
# #             sep = ",", row.names = FALSE, col.names = FALSE, quote = FALSE)
# # system("~/gits/hempy/bin/blast_to_gff.py cann_3dom_EH23a_tblastn_filter.csv")
# 
# cann_blst$chrn <- as.numeric(sub(".+chr", "", cann_blst$sseqid))
# 
# #table(cann_blst$qseqid)
# 
# # cann_blst[1:3, 1:10]
# # cann_blst[1:3, 10:20]
# 
# knitr::kable(cann_blst[, c(1, 2, 3, 7, 13:15, 18, 23)], caption = "**Table 1. Cannabinoid synthase genes.**")
# 
#nrow(cann_blst)

```




```{r}
# blst <- read.csv("EH23a_blastn.csv", header = FALSE)
# colnames(blst) <- c('qseqid','qlen','sseqid','slen','qstart',
#                     'qend','sstart','send','evalue','bitscore',
#                     'score','length','pident','nident','mismatch',
#                     'positive','gapopen','gaps','ppos','sstrand',
#                     #'staxids','sblastnames','salltitles',
#                     'sseq')#[1:21]
# 
# #blst$sseqid <- factor(blst$sseqid, levels = c("EH23a.chr1", "EH23a.chr2", "EH23a.chr3", "EH23a.chr4", "EH23a.chr5", "EH23a.chr6", "EH23a.chr7", "EH23a.chr8", "EH23a.chr9", "EH23a.chrX"))
# #blst <- blst[blst$qseqid == "CsatSD_centromere_370bp", ]
# #table(blst$qseqid)
# # blst[1:3, 1:10]
# 
# blst$chrn <- sub(".+chr", "", blst$sseqid)
# blst$chrn[ blst$chrn == "X" ] <- 10
# blst$chrn <- as.numeric(blst$chrn)
# 
# subt <- blst[grep("CsatSD_centromere_370bp", blst$qseqid), ]
# cent <- blst[grep("CsatSD_centromere_237bp", blst$qseqid), ]
# 
# #blst
# #knitr::kable(blst[1:3, c(1, 2, 3, 7, 13:15, 18, 22)], caption = "**Table X. Blast.**")
# knitr::kable(subt[1:3, c(1, 2, 3, 7, 13:15, 18, 22)], caption = "**Table X. Blast.**")

```







## Ideo function


```{r}
# plot_ideo <- function() {
#   suppressPackageStartupMessages(require(ggplot2))
# 
#   # marker_df <- data.frame(
#   #   chrom = rep(names(map), times = lapply(map, length)),
#   #   pos = unlist(map),
#   #   marker = names(unlist(map))
#   # )
#   # marker_df$chromf <- factor( marker_df$chrom, levels = names(map) )
#   # marker_df$chromn <- as.numeric( marker_df$chromf )
# 
#   # chr_df <- data.frame(
#   #   start = unlist(lapply(map, min)),
#   #   end = unlist(lapply(map, max))
#   # )
#   # chr_df$chr <- names(map)
#   # chr_df$chrf <- factor( chr_df$chr, levels = names(map))
#   # chr_df$chrn <- as.numeric( chr_df$chrf )
# 
#   chr_df <- data.frame(
#     start = 1,
#     end = nucs$Length,
#     chr = nucs$Id
#   )
#   chr_df$chrn <- sub(".+chr", "", nucs$Id)
#   chr_df$chrn[chr_df$chrn == "X"] <- 10
#   chr_df$chrn <- as.numeric(chr_df$chrn)
#   
#   
#   chrom_wid <- 0.02
#   p <- ggplot2::ggplot()
#   p <- p + ggplot2::geom_rect( data = chr_df, 
#                                ggplot2::aes( xmin = chrn - chrom_wid,
#                                   xmax = chrn + chrom_wid,
#                                   #xmin = as.numeric(as.factor(chr)) - chrom_wid,
#                                   #xmax = as.numeric(as.factor(chr)) + chrom_wid,
#                                   ymin = end, ymax = start), 
#                       #fill = "#C0C0C0",
#                       fill = "#DCDCDC",
#                       #fill = "#F5F5F5",
#                       color = "#000000"
#                       )
#   #p <- p + scale_y_reverse(limits = c(max_gd, 0))
#   
#   #wins$Id
#   thinw <- 0.28
#   p <- p + ggplot2::geom_rect( 
#     data = wins, 
#     ggplot2::aes( 
#       # xmin = chrn - CGs,
#       # xmax = chrn + CGs,
#       xmin = chrn - ATs * thinw,
#       xmax = chrn + ATs * thinw,
#       ymin = Start, 
#       ymax = End),
#     fill = wins$gcol,
#     #fill = "#0000CD",
#     #fill = "#A9A9A9",
#     #fill = "#C0C0C0",
#     #fill = "#DCDCDC",
#     #fill = "#F5F5F5",
# #                  color = "#000000"
#     color = NA
#     )
#   #p
# 
#   cmwidth <- 0.4
#   p <- p + ggplot2::geom_rect( 
#     data = cann_blst, 
#     ggplot2::aes(
#       xmin = chrn - cmwidth,
#       xmax = chrn + cmwidth,
#       ymin = sstart, 
#       ymax = send),
#     #fill = "#0000CD",
#     #fill = "#A9A9A9",
#     fill = "#C0C0C0",
#     #fill = "#DCDCDC",
#     #fill = "#F5F5F5",
#     #color = "#000000"
#     color = "#228B22"
#     #color = NA
#     )
#   #p
#   
# 
#   # p <- p + annotate(geom="text", x=7.5, y=11.3e6, label="THCAS1",
#   #             color="#228B22", size = 3)
#   # p <- p + annotate(geom="text", x=7.5, y=12.1e6, label="CBDAS2",
#   #             color="#0000FF", size = 3)
#   # p <- p + annotate(geom="text", x=7.5, y=30.7e6, label="CBCAS",
#   #             color="#B22222", bg = "#ffffff", size = 3)
#   #p
#   
#   stwidth <- 0.40
#   p <- p + ggplot2::geom_rect( 
#     data = subt, 
#     ggplot2::aes(
#       xmin = chrn - stwidth,
#       xmax = chrn + stwidth,
#       #xmax = chrn,
#       ymin = sstart, 
#       ymax = send),
#     #fill = "#0000CD",
#     #fill = "#A9A9A9",
#     fill = "#C0C0C0",
#     #fill = "#DCDCDC",
#     #fill = "#F5F5F5",
#     #color = "#000000"
#     color = "#0000FF"
#     #color = "#228B22"
#     #color = NA
#     )
#   #p
#   
#   mwidth <- 0.6
#   p <- p + ggplot2::geom_rect( 
#     data = cent, 
#     ggplot2::aes(
#       xmin = chrn - mwidth,
# #      xmax = chrn + mwidth,
#       xmax = chrn - 0.2,
#       ymin = sstart, 
#       ymax = send),
#     #fill = "#0000CD",
#     #fill = "#A9A9A9",
#     fill = "#C0C0C0",
#     #fill = "#DCDCDC",
#     #fill = "#F5F5F5",
#     color = "#000000"
#     #color = "#0000FF"
#     #color = "#8A2BE2"
#     #color = "#228B22"
#     #color = NA
#     )
#   #p
#   
# 
# #   mwidth <- 0.4
# #   p <- p + ggplot2::geom_rect( 
# #     data = Ty3, 
# #     ggplot2::aes(
# #       xmin = chrn - 0,
# # #      xmin = chrn - mwidth,
# #       xmax = chrn + mwidth,
# # #      xmax = chrn,
# #       ymin = V4, 
# #       ymax = V5),
# #     fill = "#0000FF",
# #     color = NA
# #     )
# #   #p
#   
#   
#   # marker_wid <- 0.1
#   # marker_high <- 0.4
#   # p <- p + ggplot2::geom_rect( data = marker_df, 
#   #                     ggplot2::aes( xmin = chromn - marker_wid,
#   #                          xmax = chromn + marker_wid, 
#   #                          #xmin = as.numeric(as.factor(chrom)) - marker_wid,
#   #                          #xmax = as.numeric(as.factor(chrom)) + marker_wid,
#   #                          ymin = pos - marker_high, ymax = pos + marker_high),
#   #                     fill = "#228B22", color = "#228B22"
#   # )
#   # 
#   # p <- p + scale_y_reverse( breaks = seq(0, 2e3, by = 100) )
#   #p <- p + ggplot2::scale_y_reverse( minor_breaks = seq(0, 2e3, by = 20), breaks = seq(0, 2e3, by = 100) )
#   #p <- p + scale_x_continuous( breaks = as.numeric(as.factor(chr_df$chr) ) )
#   p <- p + ggplot2::scale_x_continuous( breaks = chr_df$chrn, labels = chr_df$chr )
#   # p <- p + scale_y_reverse(limits = c(max_gd, 0))
#   # p <- p + scale_y_reverse(limits = c(0, max_gd))
#   #p <- p + theme_bw() + theme( panel.grid.minor = element_blank() )
#   p <- p + ggplot2::theme_bw() + 
#     ggplot2::theme( panel.grid.minor.x = ggplot2::element_blank(), 
#            axis.text.x = element_text(angle = 60, hjust = 1),
#            axis.title.x=element_blank(),
#            panel.grid.major.y = ggplot2::element_line( linewidth = 0.4, color = "#C0C0C0", linetype = 1 ),
#            panel.grid.minor.y = ggplot2::element_line( linewidth = 0.4, color = "#C0C0C0", linetype = 3 )
#            #panel.grid.major.y = ggplot2::element_line( size = 0.4, color = "#C0C0C0", linetype = 1 ),
#            #panel.grid.minor.y = ggplot2::element_line( size = 0.4, color = "#C0C0C0", linetype = 3 )
#           )
# #  p <- p + ggplot2::xlab("Chromosome")
# 
#   #p <- p + ylab("Distance (cM)")
#   #p <- p + ggplot2::ylab("Location (cM)")
#   p <- p + ggplot2::ylab("Position (bp)")
#   #p
#   
# 
#   #p
# 
#   #return( invisible( NULL ) )
#   p
# }
```



```{r}
# p1 <- plot_ideo()
# p1
```


## Add 


```{r}
#lndmrk[1:3, ]
#nrow(lndmrk)
#table(lndmrk$qseqid)

# pmildew <- "PNW39_LH3804_locus"
# lgrp <- system( 'grep "^>" ../csat_landmark_dna.fa', intern = TRUE )
# lgrp <- sub("^>", "", lgrp)
# #lgrp
# 
# # Cann
# my_regex <- sub("[[:space:]].+", "", lgrp[1:6])
# my_regex <- paste(my_regex, collapse="|")
# lndmrk2 <- lndmrk[grep(my_regex, lndmrk$qseqid), ]
```


```{r}
# Cent, subTelo
#lgrp[7:9]
# my_regex <- sub("[[:space:]].+", "", lgrp[7:9])
# my_regex <- paste(my_regex, collapse="|")
# lndmrk2 <- lndmrk[grep(my_regex, lndmrk$qseqid), ]
# table(lndmrk2$qseqid)
# 
# chrom_wid <- 0.46
# tmp <- data.frame( 
#   xmin = lndmrk2$chrn - chrom_wid, 
#   xmax = lndmrk2$chrn + chrom_wid, 
#   ymin = lndmrk2$sstart, 
#   ymax = lndmrk2$send
#   )
# 
# p1 + ggplot2::geom_rect( 
#   data = tmp, 
#       ggplot2::aes( 
#         xmin = xmin,
#         xmax = xmax,
#         ymin = ymin, 
#         ymax = ymax
#         ),
#       fill = "#DCDCDC",
# #      color = "#CD853F",
#       color = "#F08080",
# #      color = "#000000"
#       linewidth = 2
#       ) + ggtitle("Sub-telo, centromeric motifs")


```


```{r}
# P mildew
#lgrp[10:10]
# my_regex <- sub("[[:space:]].+", "", lgrp[10:10])
# my_regex <- paste(my_regex, collapse="|")
# lndmrk2 <- lndmrk[grep(my_regex, lndmrk$qseqid), ]
# 
# chrom_wid <- 0.46
# tmp <- data.frame( 
#   xmin = lndmrk2$chrn - chrom_wid, 
#   xmax = lndmrk2$chrn + chrom_wid, 
#   ymin = lndmrk2$sstart, 
#   ymax = lndmrk2$send
#   )
# 
# p1 + ggplot2::geom_rect( 
#   data = tmp, 
#       ggplot2::aes( 
#         xmin = xmin,
#         xmax = xmax,
#         ymin = ymin, 
#         ymax = ymax
#         ),
#       fill = "#DCDCDC",
# #      color = "#CD853F",
#       color = "#F08080",
# #      color = "#000000"
#       linewidth = 2
#       ) + ggtitle("Powdery mildew")

```


```{r}
# # CENP
# grep("NC_044371.1:c59909808-59908167 Cannabis sativa chromosome 1, cs10, whole genome shotgun sequence", lgrp):grep("NC_044371.1:c74285628-74284242 Cannabis sativa chromosome 1, cs10, whole genome shotgun sequence", lgrp)
# #lgrp[11:26]
# my_regex <- sub("[[:space:]].+", "", lgrp[11:26])
# my_regex <- paste(my_regex, collapse="|")
# lndmrk2 <- lndmrk[grep(my_regex, lndmrk$qseqid), ]
# 
# chrom_wid <- 0.46
# tmp <- data.frame( 
#   xmin = lndmrk2$chrn - chrom_wid, 
#   xmax = lndmrk2$chrn + chrom_wid, 
#   ymin = lndmrk2$sstart, 
#   ymax = lndmrk2$send
#   )
# 
# p1 + ggplot2::geom_rect( 
#   data = tmp, 
#       ggplot2::aes( 
#         xmin = xmin,
#         xmax = xmax,
#         ymin = ymin, 
#         ymax = ymax
#         ),
#       fill = "#DCDCDC",
# #      color = "#CD853F",
#       color = "#F08080",
# #      color = "#000000"
#       linewidth = 2
#       ) + ggtitle("CENP")

```


```{r}
# # rRNA
# grep("NC_044375.1:87745002-87745120 Cannabis sativa chromosome 2, cs10, whole genome shotgun sequence", lgrp):grep("NC_026562.1:100368-101858 Cannabis sativa cultivar Carmagnola chloroplast, complete genome", lgrp)
# #lgrp[27:55]
# my_regex <- sub("[[:space:]].+", "", lgrp[27:55])
# my_regex <- paste(my_regex, collapse="|")
# lndmrk2 <- lndmrk[grep(my_regex, lndmrk$qseqid), ]
# 
# lndmrk2 <- lndmrk2[ lndmrk2$gaps == 0, ]
# lndmrk2 <- lndmrk2[ lndmrk2$pident >= 95, ]
# 
# #lndmrk2 <- lndmrk2[grep("^NC_044", lndmrk2$qseqid), ]
# 
# tbl2 <- table(lndmrk2$chrn, lndmrk2$qseqid)
# tbl2[, colSums(as.matrix(tbl2)) > 1e3]
# #tbl2['8',]
# 
# #tbl2[, grep("^NW", colnames(tbl2))]
# 
# table(lndmrk2$chrn)
# 
# # OAS
# #lgrp[56:57]
# #my_regex <- sub("[[:space:]].+", "", lgrp[56:57])
# #my_regex <- paste(my_regex, collapse="|")
# #lndmrk2 <- lndmrk[grep(my_regex, lndmrk$qseqid), ]
# 
# 
# #tmp <- data.frame( xmin = 1:4, xmax = 2:5, ymin = rep(10, times = 4), ymax = rep(10000000, times = 4))
# 
# chrom_wid <- 0.46
# tmp <- data.frame( 
#   xmin = lndmrk2$chrn - chrom_wid, 
#   xmax = lndmrk2$chrn + chrom_wid, 
#   ymin = lndmrk2$sstart, 
#   ymax = lndmrk2$send
#   )
# 
# p1 + ggplot2::geom_rect( 
#   data = tmp, 
#       ggplot2::aes( 
#         xmin = xmin,
#         xmax = xmax,
#         ymin = ymin, 
#         ymax = ymax
#         ),
#       #fill = "#C0C0C0",
#       fill = "#DCDCDC",
#       #fill = "#F5F5F5",
# #      color = "#FFD700"
# #      color = "#DAA520",
#       color = "#CD853F",
# #      color = "#000000"
#       linewidth = 2
#       ) + ggtitle("rRNA")
# 
# #ggsave(filename = "EH23a_ideo.png", device = "png", width = 6.5, height = 6.5, units = "in", dpi = 300)

```


```{r}
# 
# # OAS
# #lgrp[56:57]
# my_regex <- sub("[[:space:]].+", "", lgrp[56:57])
# my_regex <- paste(my_regex, collapse="|")
# lndmrk2 <- lndmrk[grep(my_regex, lndmrk$qseqid), ]
# 
# chrom_wid <- 0.46
# tmp <- data.frame( 
#   xmin = lndmrk2$chrn - chrom_wid, 
#   xmax = lndmrk2$chrn + chrom_wid, 
#   ymin = lndmrk2$sstart, 
#   ymax = lndmrk2$send
#   )
# 
# p1 + ggplot2::geom_rect( 
#   data = tmp, 
#       ggplot2::aes( 
#         xmin = xmin,
#         xmax = xmax,
#         ymin = ymin, 
#         ymax = ymax
#         ),
#       fill = "#DCDCDC",
# #      color = "#CD853F",
#       color = "#F08080",
# #      color = "#000000"
#       linewidth = 2
#       ) + ggtitle("OAS")

```



## Chrom by genes


```{r}

my_data <- readr::read_csv("gene_counts.csv", )
# names(my_data)[1] <- "Sample"
# my_data$Sample <- as.factor( my_data$Sample )
my_data

library(tidyr)
data_long <- my_data %>%
  pivot_longer( cols = !Sample, names_to = "Chrom", values_to = "Count")
#data_long$Length <- data_long$Length / 1e6
gcount <- data_long
sum(is.na(data_long$Count))
data_long <- data_long[!is.na(data_long$Count), ]
data_long


my_data <- readr::read_csv("chrom_lengths.csv")
names(my_data)[1] <- "Sample"
my_data$Sample <- as.factor( my_data$Sample )
my_data

library(tidyr)
data_long <- my_data %>%
  pivot_longer( cols = !Sample, names_to = "Chrom", values_to = "Length")
data_long$Length <- data_long$Length / 1e6
data_long
glength <- data_long


```




```{r}
glength
gcount
data_long <- cbind(glength, gcount$Count)
names(data_long)[4] <- "Count"
data_long[1:3, ]

my_pal <- RColorBrewer::brewer.pal(n=12, name = "Paired")
#my_pal[11] <- "#B15928"
my_pal[11] <- "#C71585"
my_pal <- paste(my_pal, "88", sep = "")

library(ggplot2)
# Basic scatter plot
p <- ggplot(data_long, aes(x=Length, y=Count, color=Chrom)) 
p <- p + geom_point(size=2)
#p <- p + geom_smooth(method=lm)
p <- p + geom_smooth(method=lm, se=FALSE, linewidth = 1)

p <- p + theme_bw()
p <- p + theme(legend.title = element_blank()) 
#p <- p + theme(legend.position = "left")
p <- p + theme(legend.position = "right")
#p <- p + theme( legend.spacing.y = unit(1.0, 'mm') )
## important additional element
#p <- p + guides(fill = guide_legend(byrow = TRUE))
p + theme(legend.spacing.y = unit(1.0, 'mm')) +
  guides(fill = guide_legend(byrow = TRUE))
#p <- p + scale_color_brewer(palette="Dark2")
#p <- p + scale_color_brewer(palette="Paired")
p <- p + scale_color_manual(values=my_pal)
#p <- p + scale_color_brewer(palette="Set3")
p <- p + xlab("Chromosome length (Mbp)")
p <- p + ylab("Genes per chromosome")



p2 <- p
p2

```


## TE plot


```{r}
wins <- read.csv("EH23a.softmasked_wins1e6.csv")

gff <- read.table("../Figure1/EH23a.primary_high_confidence.gff3.gz", sep = "\t", quote = "\"")
#gff[1:3, 1:8]
genes <- gff[ gff[, 3] == "gene", ]

```



```{r}
gff <- read.table("../Figure1/EH23a.unmasked.fasta.mod.EDTA.intact.gff3", sep = "\t", quote = "\"")
names(gff) <- c("seqid","source","type","start","end","score","strand","phase","attributes")

#gff$chrn <- sub("^.+chr", "", gff$V1)
gff$chrn <- sub("^.+chr", "", gff$seqid)
gff$chrn[ gff$chrn == "X" ] <- 10
gff$chrn <- as.numeric(gff$chrn)

gff$classification <- unlist(lapply(strsplit(gff[,9], split = ";"), function(x){ grep("Classification=", x, value = TRUE) }))
gff$classification <- sub("^Classification=", "", gff$classification)
gff$classification <- sub("^LTR/Copia", "Ty1", gff$classification)
gff$classification <- sub("^LTR/Gypsy", "Ty3", gff$classification)
#
gff[1:8, c(1:8, 10)]

# Ty3
Ty3 <- gff[ gff$classification == "Ty3", ]
Ty3 <- Ty3[ Ty3$type == "Gypsy_LTR_retrotransposon", ]
#Ty3 <- Ty3[ Ty3$V3 == "Gypsy_LTR_retrotransposon", ]
Ty3[1:8, c(1:8, 10:11)]
```



```{r}
wins$gcnt <- 0
#wins$ty1cnt <- 0
wins$ty3cnt <- 0
for(i in 1:nrow(wins)){
   tmp <- genes[genes$V1 == wins$Id[i] & genes$V4 >= wins$Start[i] & genes$V5 < wins$End[i], ]
   wins$gcnt[i] <- nrow(tmp)
   #tmp <- gff[gff$V1 == wins$Id[i] & gff$V4 >= wins$Start[i] & gff$V5 < wins$End[i], ]
   tmp <- Ty3[Ty3$seqid == wins$Id[i] & Ty3$start >= wins$Start[i] & Ty3$end < wins$End[i], ]
   wins$ty3cnt[i] <- nrow(tmp)
#   wins$ty1cnt[i] <- sum(tmp$classification == "Ty1")
}
wins[1:3, ]
hist(wins$ty3cnt)
```




```{r}
#plot(wins$gcnt, wins$ty3cnt)
#plot(wins$gcnt, wins$ty1cnt)
# wins[1:3, ]
# table(wins$Id)

# my_pal <- RColorBrewer::brewer.pal(n=12, name = "Paired")
# #my_pal[11] <- "#B15928"
# my_pal[11] <- "#C71585"
# #my_pal <- paste(my_pal, "66", sep = "")

my_pal <- RColorBrewer::brewer.pal(n=12, name = "Paired")
#my_pal[11] <- "#B15928"
my_pal[11] <- "#C71585"
my_pal <- paste(my_pal, "88", sep = "")

lm1 <- lm( wins$ty3cnt ~ wins$gcnt )
summary(lm1)
my_coefs <- round(lm1$coefficients, digits = 3)

#wins[1:3, ]

library(ggplot2)
# Basic scatter plot
#p <- ggplot(wins, aes(x=gcnt, y=ty3cnt, color=Id))
#p <- ggplot(wins, aes(x=gcnt, y=ty3cnt, color=chr))
p <- ggplot( data = wins, aes(x=gcnt, y=ty3cnt, color = Id))
#p <- p + geom_point(size=2, color = "#C0C0C0")
#p <- p + geom_point(size=2, color = "#77889966")
# p <- p + geom_point(size=2, color = "#70809044")
#
p <- p + geom_point(size=2)

#p <- p + geom_smooth(method=lm)
#p <- p + geom_smooth(method=lm, se=FALSE, linewidth = 2)
p <- p + geom_smooth(method=lm, se=FALSE, linewidth = 1)

#p <- p + geom_text(x=100, y=90, label="y = (-0.4)x + 56.7", size = 4)
p <- p + geom_text(x=100, y=15, label= paste("y = ", my_coefs[2], "x + ", my_coefs[1], sep = ""), color = "#000000", size = 4, parse = F)

p <- p + theme_bw()
p <- p + theme(legend.position = "none")
#p <- p + theme(legend.title = element_blank()) 
#p <- p + scale_color_brewer(palette="Dark2")
#p <- p + scale_color_brewer(palette="Paired")
p <- p + scale_color_manual(values=my_pal)
#p <- p + scale_color_brewer(palette="Set3")
# p <- p + xlab("Genes per 1 Mbp window")
# p <- p + ylab("Ty3 elements per 1 Mbp window")
p <- p + xlab("Genes per window")
p <- p + ylab("Ty3 elements window")
# p

p3 <- p
p3


# lm2 <- lm( wins$ty1cnt ~ wins$gcnt )
# summary(lm2)

```




## Graphic


```{r, results='hide'}
source("../FigureSideo/Rfunctions.R")
p1 <- plot_ideo("/media/knausb/E737-9B48/releases/scaffolded//EH23a")
```


```{r}
#library(magick)
#ggdraw() +
#  cowplot::draw_image("https://i.stack.imgur.com/WDOo4.jpg?s=328&g=1") #+
#  draw_plot(my_plot)
```


---


## Figure 1 (Composite plot)


```{r, fig.height=8, fig.width=8, message=FALSE, fig.cap="**Figure 1. Genomic architecture of the** ***Cannabis sativa*** **genome.**"}

library("ggpubr")

ggarrange(
  plotlist = list(p1, 
                  ggarrange(p3, p2, 
                            ncol = 2, nrow = 1, 
                            widths = c(2, 3),
                            labels = c("B", "C"))
                  ),
  labels = c("A", ""),
  ncol = 1, nrow = 2, 
  widths = 1, heights = c(2, 1))
    
```


**Figure 1. Genomic architecture of the** ***Cannabis sativa*** **genome.** 
**(A)** Ideogram of the ERBxHO40_23 haplotype A genome.
Each chromosome is divided into 1 Mbp windows where each window's width is proportional to the inverse abundance of the motif 'CG' and each window is colored according to gene density.
Windows with high 'CG' abundance are narrow and windows with high gene counts are yellow.
The plant ERBxHO40_23 resulted from a cross between strains ERB (chemotype III or CBD dominant) and HO40 (chemotype I or THC dominant). The genetically female plant HO40 (XX) was chemically masculinized to produce pollen flowers to facilitate the cross.
Subtelomeric repeat motifs are marked with blue horizontal lines.
Cannabinoid synthase genes are marked with green horizontal lines on chromosome seven.
**(B)** The abundances of genes and Ty3 LTRs are inversely related.
Each dot represents a 1 Mbp window from the ERBxHO40_23 haplotype A assembly, organized by the number of genes and Ty3 LTR elements contained in each window.
Windows with high quantities of genes have low Ty3 LTR counts.
The genes in ERBxHO40 are predominantly near the ends of each chromosome, while the central portion of each chromosome appears populated by Ty3 LTR elements.
**(C)** Long chromosomes have more genes, however some chromosomes demonstrate exceptional gene content.
Chromosome Y is the longest chromosome but includes a number of genes that is comparable to other chromosomes.
Each dot represents a chromosome from phased and assembled haplotypes (n = 69; X = 65; Y = 4).
Chromosome 1 is of average length but contains more genes than other chromosomes of a similar length.
Chromosomes X and 8 also have more genes than chromosomes of similar length.
Chromosome 7, where the cannabinoid synthases (CBCAS, CBDAS, THCAS) are found, has the lowest number of genes.



